<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Donghyung Lee &amp; Duygu Ucar" />

<meta name="date" content="2018-08-03" />

<title>Performance assessment of IA-SVA using Brain scRNA-Seq data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">iasvaExamples</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/dleelab/iasvaExamples">GitHub</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Performance assessment of IA-SVA using Brain scRNA-Seq data</h1>
<h4 class="author"><em>Donghyung Lee &amp; Duygu Ucar</em></h4>
<h4 class="date"><em>2018-08-03</em></h4>

</div>

<div id="TOC">
<ul>
<li><a href="#load-packages">Load packages</a></li>
<li><a href="#load-the-brain-single-cell-rna-seq-data">Load the brain single cell RNA-Seq data</a></li>
<li><a href="#calculate-geometric-library-size-i.e.-library-size-of-log-transfromed-read-counts">Calculate geometric library size, i.e., library size of log-transfromed read counts</a></li>
<li><a href="#run-analyses">run analyses</a></li>
<li><a href="#compare-results">Compare results</a></li>
<li><a href="#session-information">Session information</a></li>
</ul>
</div>

<p><strong>Last updated:</strong> 2018-08-03</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(20180731)</code> </summary></p>
<p>The command <code>set.seed(20180731)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/dleelab/iasvaExamples/tree/268983524d3aee232c1bf032e8dd8993bcfb4fc3" target="_blank">2689835</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    inst/.DS_Store
    Ignored:    inst/doc/.DS_Store
    Ignored:    vignettes/.DS_Store

Untracked files:
    Untracked:  Clustering_analyses_figure4_islets_sv1_3.pdf
    Untracked:  docs/figure/hidden_heterogeneity_glioblastoma.Rmd/
    Untracked:  docs/figure/tSNE_post_IA-SVA_3celltypes.Rmd/
    Untracked:  docs/figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/
    Untracked:  output/CC_genes.long.txt
    Untracked:  output/CC_genes.short.txt
    Untracked:  output/Clustering_analyses_figure3_sv1.pdf
    Untracked:  output/Clustering_analyses_figure4_Xin.pdf
    Untracked:  output/Clustering_analyses_figure4_Xin_sv1.pdf
    Untracked:  output/FigureS11_Xin_Islets_AllCells_IASVA_Markers_pheatmap.pdf
    Untracked:  output/Lawlor_Islets_3Cells_CellView_Seurat_FigS.pdf
    Untracked:  output/Lawlor_Islets_3Cells_IASVA_SV1SV3_rsqcutoff0.3_pheatmap_iasvaV0.95_Figure4_C.pdf
    Untracked:  output/Lawlor_Islets_3Cells_IASVA_SV4_rsqcutoff0.3_pheatmap_iasvaV0.95.pdf
    Untracked:  output/Lawlor_Islets_3Cells_IASVA_pairs4SVs_iasvaV0.95_black_FigS6.pdf
    Untracked:  output/Lawlor_Islets_3Cells_IASVA_pairs4SVs_iasvaV0.95_color_FigS6.pdf
    Untracked:  output/Lawlor_Islets_3Cells_SV1_SV3_Cell_Type_Genes_rsqcutoff0.3.txt
    Untracked:  output/Lawlor_Islets_3Cells_SV4_Genes_rsqcutoff0.3.txt
    Untracked:  output/Lawlor_Islets_3Cells_tSNE_IA-SVA_Fig4AB.pdf
    Untracked:  output/Patel_Glioblastoma_MGH30_CellCycle_Figure3ABCD.pdf
    Untracked:  output/Patel_Glioblastoma_MGH30_Cellcycle_SV1_Genes_rsqcutoff0.3.txt
    Untracked:  output/Patel_Glioblastoma_MGH30_Cellcycle_SV1_Genes_rsqcutoff0.4.txt
    Untracked:  output/Patel_Glioblastoma_MGH30_iasva_SV1_genes_rsqcutoff0.3_pheatmap_iasvaV0.95_Figure3F.pdf
    Untracked:  output/Xin_Islets_AllCells_IASVA.pdf
    Untracked:  output/Xin_Islets_AllCells_IASVA_nocolor.pdf
    Untracked:  output/Xin_Islets_AllCells_PCA.pdf
    Untracked:  output/Xin_Islets_AllCells_SV1_Genes_rsqcutoff0.2.txt
    Untracked:  output/Xin_Islets_AllCells_SV1_Genes_rsqcutoff0.3.txt
    Untracked:  output/Xin_Islets_AllCells_SV3_Genes_rsqcutoff0.2.txt
    Untracked:  output/Xin_Islets_AllCells_SV3_Genes_rsqcutoff0.3.txt
    Untracked:  output/Xin_Islets_AllCells_USVA.pdf
    Untracked:  output/Xin_Islets_AllCells_tSNEByKnownFactors_FigureS9.pdf
    Untracked:  output/Xin_Islets_All_demensionality_reduction_Figure4DEFG.pdf

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/c241b04dac899407045051a04b7539237f4d183f/analysis/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/dleelab/iasvaExamples/c241b04dac899407045051a04b7539237f4d183f/docs/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.html" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<p>Here, we becnhmark several factor analyses methods on single cell RNA sequencing (scRNA-Seq) data obtained from human human brain samples (<a href="http://www.pnas.org/content/112/23/7285.long">Darmanis et. al., 2015</a>). All of these methods are used to uncover variation associated with different cell types and compared against each other. This dataset is included in an R data package (“iasvaExamples”) containing data examples for IA-SVA (<a href="https://github.com/dleelab/iasvaExamples" class="uri">https://github.com/dleelab/iasvaExamples</a>). To install the ‘iasvaExamples’ package, follow the instructions provided in the GitHub page.</p>
<div id="load-packages" class="section level2">
<h2>Load packages</h2>
<pre class="r"><code>rm(list=ls())
library(iasva)
library(iasvaExamples)
library(sva)
library(RUVSeq)
library(Rtsne)
library(pheatmap)
library(corrplot)
library(DescTools) #pcc i.e., Pearson&#39;s contingency coefficient
library(RColorBrewer)
library(zinbwave)
library(nnet)
library(SummarizedExperiment)
color.vec &lt;- brewer.pal(8, &quot;Set1&quot;)
set.seed(3532379)
# To make the task more challenging we selected 
  #1000 genes to be used in downstream analyses. 
num.genes &lt;- 1000</code></pre>
</div>
<div id="load-the-brain-single-cell-rna-seq-data" class="section level2">
<h2>Load the brain single cell RNA-Seq data</h2>
<pre class="r"><code>data(&quot;Darmanis_Brain_scRNAseq_Read_Counts&quot;)
data(&quot;Darmanis_Brain_scRNAseq_Annotations&quot;)
ls()</code></pre>
<pre><code>[1] &quot;color.vec&quot;                          
[2] &quot;Darmanis_Brain_scRNAseq_Annotations&quot;
[3] &quot;Darmanis_Brain_scRNAseq_Read_Counts&quot;
[4] &quot;num.genes&quot;                          </code></pre>
<pre class="r"><code>counts.total &lt;- Darmanis_Brain_scRNAseq_Read_Counts
anns.total &lt;- Darmanis_Brain_scRNAseq_Annotations
dim(anns.total)</code></pre>
<pre><code>[1] 461   6</code></pre>
<pre class="r"><code>dim(counts.total)</code></pre>
<pre><code>[1] 23267   461</code></pre>
<pre class="r"><code>summary(anns.total)</code></pre>
<pre><code>         run              tissue                cell_type  
 SRR1974543:  1   cortex     :426   neurons          :129  
 SRR1974544:  1   hippocampus: 35   fetal_quiescent  :109  
 SRR1974545:  1                     astrocytes       : 62  
 SRR1974546:  1                     hybrid           : 46  
 SRR1974547:  1                     oligodendrocytes : 37  
 SRR1974548:  1                     fetal_replicating: 25  
 (Other)   :455                     (Other)          : 53  
                 age           c1_chip_id   sample_name 
 prenatal 16-18 W  :133   nochipID2 : 77   AB_S4  : 77  
 postnatal 50 years: 77   nochipID8 : 55   AB_S11 : 62  
 postnatal 37 years: 62   nochipID11: 45   AB_S8  : 57  
 postnatal 54 years: 57   1772078217: 33   AB_S7  : 55  
 postnatal 21 years: 55   1772078236: 33   FB_S2  : 45  
 postnatal 63 years: 49   nochipID13: 33   AB_S5  : 44  
 (Other)           : 28   (Other)   :185   (Other):121  </code></pre>
<pre class="r"><code>table(anns.total$tissue, anns.total$cell_type)</code></pre>
<pre><code>             
              astrocytes endothelial fetal_quiescent fetal_replicating
  cortex              62          17             109                25
  hippocampus          0           3               0                 0
             
              hybrid microglia neurons oligodendrocytes OPC
  cortex          44         8     128               31   2
  hippocampus      2         7       1                6  16</code></pre>
<pre class="r"><code>ContCoef(table(anns.total$tissue, anns.total$cell_type))</code></pre>
<pre><code>[1] 0.5794563</code></pre>
<pre class="r"><code>table(anns.total$age, anns.total$cell_type)</code></pre>
<pre><code>                    
                     astrocytes endothelial fetal_quiescent
  postnatal 21 years          3           0               0
  postnatal 22 years          0           1               1
  postnatal 37 years          7           1               0
  postnatal 47 years         11           2               0
  postnatal 50 years         38           0               0
  postnatal 54 years          1           0               0
  postnatal 63 years          2          16               0
  prenatal 16-18 W            0           0             108
                    
                     fetal_replicating hybrid microglia neurons
  postnatal 21 years                 0      7         0      34
  postnatal 22 years                 0      0         0       2
  postnatal 37 years                 0      3         1      49
  postnatal 47 years                 0      0         5       6
  postnatal 50 years                 0     19         0      19
  postnatal 54 years                 0     17         6       0
  postnatal 63 years                 0      0         3      19
  prenatal 16-18 W                  25      0         0       0
                    
                     oligodendrocytes OPC
  postnatal 21 years               11   0
  postnatal 22 years                0   0
  postnatal 37 years                1   0
  postnatal 47 years                0   0
  postnatal 50 years                1   0
  postnatal 54 years               15  18
  postnatal 63 years                9   0
  prenatal 16-18 W                  0   0</code></pre>
<pre class="r"><code>ContCoef(table(anns.total$age, anns.total$cell_type))</code></pre>
<pre><code>[1] 0.8256324</code></pre>
<pre class="r"><code>table(anns.total$sample_name, anns.total$cell_type)</code></pre>
<pre><code>        
         astrocytes endothelial fetal_quiescent fetal_replicating hybrid
  AB_S1          11           2               0                 0      0
  AB_S11          7           1               0                 0      3
  AB_S2           2           0               0                 0      0
  AB_S3           0           1               1                 0      0
  AB_S4          38           0               0                 0     19
  AB_S5           0          16               0                 0      0
  AB_S7           3           0               0                 0      7
  AB_S8           1           0               0                 0     17
  FB_S1           0           0               5                21      0
  FB_S2           0           0              43                 2      0
  FB_S3           0           0              33                 0      0
  FB_S6           0           0              27                 2      0
        
         microglia neurons oligodendrocytes OPC
  AB_S1          5       6                0   0
  AB_S11         1      49                1   0
  AB_S2          2       1                0   0
  AB_S3          0       2                0   0
  AB_S4          0      19                1   0
  AB_S5          1      18                9   0
  AB_S7          0      34               11   0
  AB_S8          6       0               15  18
  FB_S1          0       0                0   0
  FB_S2          0       0                0   0
  FB_S3          0       0                0   0
  FB_S6          0       0                0   0</code></pre>
<pre class="r"><code>ContCoef(table(anns.total$sample_name, anns.total$cell_type))</code></pre>
<pre><code>[1] 0.8602264</code></pre>
<pre class="r"><code>table(anns.total$c1_chip_id, anns.total$cell_type)</code></pre>
<pre><code>            
             astrocytes endothelial fetal_quiescent fetal_replicating
  1772078217          1           0               0                 0
  1772078218          0           0               0                 0
  1772078236          2           0               0                 0
  1772078237          5           1               0                 0
  nochipID10          0           0               5                21
  nochipID11          0           0              43                 2
  nochipID12          1           2               0                 0
  nochipID13          0           0              33                 0
  nochipID14          2           0               0                 0
  nochipID15          0           1               1                 0
  nochipID2          38           0               0                 0
  nochipID3           0          13               0                 0
  nochipID4           0           0              27                 2
  nochipID5           0           3               0                 0
  nochipID8           3           0               0                 0
  nochipID9          10           0               0                 0
            
             hybrid microglia neurons oligodendrocytes OPC
  1772078217     15         0       0               15   2
  1772078218      2         6       0                0  16
  1772078236      3         1      26                1   0
  1772078237      0         0      23                0   0
  nochipID10      0         0       0                0   0
  nochipID11      0         0       0                0   0
  nochipID12      0         2       0                0   0
  nochipID13      0         0       0                0   0
  nochipID14      0         2       1                0   0
  nochipID15      0         0       2                0   0
  nochipID2      19         0      19                1   0
  nochipID3       0         0      17                3   0
  nochipID4       0         0       0                0   0
  nochipID5       0         1       1                6   0
  nochipID8       7         0      34               11   0
  nochipID9       0         3       6                0   0</code></pre>
<pre class="r"><code>ContCoef(table(anns.total$c1_chip_id, anns.total$cell_type))</code></pre>
<pre><code>[1] 0.8802176</code></pre>
<p>The annotations describing samples and experimental settings are stored in “anns” and the read counts information is stored in “counts”.</p>
<p>For the downstream analyses, select neurons, oligodendrcytes and astrocytes from 2 cell types and 8 individuals.</p>
<pre class="r"><code># Select neurons and astrocytes for downstream analyses
selected.cells &lt;- intersect(which(anns.total$tissue==&quot;cortex&quot;),
                            which(anns.total$cell_type==&quot;neurons&quot;|anns.total$cell_type==&quot;astrocytes&quot;))
counts &lt;- counts.total[, selected.cells]
anns &lt;- droplevels(anns.total[selected.cells,])
table(anns$tissue, anns$cell_type)</code></pre>
<pre><code>     astrocytes neurons</code></pre>
<p>cortex 62 128</p>
<pre class="r"><code>table(anns$tissue, anns$age)</code></pre>
<pre><code>     postnatal 21 years postnatal 22 years postnatal 37 years</code></pre>
<p>cortex 37 2 56</p>
<pre><code>     postnatal 47 years postnatal 50 years postnatal 54 years</code></pre>
<p>cortex 17 57 1</p>
<pre><code>     postnatal 63 years</code></pre>
<p>cortex 20</p>
<pre class="r"><code>table(anns$age, anns$sample_name)</code></pre>
<pre><code>                 AB_S1 AB_S11 AB_S2 AB_S3 AB_S4 AB_S5 AB_S7 AB_S8</code></pre>
<p>postnatal 21 years 0 0 0 0 0 0 37 0 postnatal 22 years 0 0 0 2 0 0 0 0 postnatal 37 years 0 56 0 0 0 0 0 0 postnatal 47 years 17 0 0 0 0 0 0 0 postnatal 50 years 0 0 0 0 57 0 0 0 postnatal 54 years 0 0 0 0 0 0 0 1 postnatal 63 years 0 0 3 0 0 17 0 0</p>
<pre class="r"><code>table(anns$age, anns$c1_chip_id)</code></pre>
<pre><code>                 1772078217 1772078236 1772078237 nochipID12</code></pre>
<p>postnatal 21 years 0 0 0 0 postnatal 22 years 0 0 0 0 postnatal 37 years 0 28 28 0 postnatal 47 years 0 0 0 1 postnatal 50 years 0 0 0 0 postnatal 54 years 1 0 0 0 postnatal 63 years 0 0 0 0</p>
<pre><code>                 nochipID14 nochipID15 nochipID2 nochipID3 nochipID8</code></pre>
<p>postnatal 21 years 0 0 0 0 37 postnatal 22 years 0 2 0 0 0 postnatal 37 years 0 0 0 0 0 postnatal 47 years 0 0 0 0 0 postnatal 50 years 0 0 57 0 0 postnatal 54 years 0 0 0 0 0 postnatal 63 years 3 0 0 17 0</p>
<pre><code>                 nochipID9</code></pre>
<p>postnatal 21 years 0 postnatal 22 years 0 postnatal 37 years 0 postnatal 47 years 16 postnatal 50 years 0 postnatal 54 years 0 postnatal 63 years 0</p>
<pre class="r"><code>table(anns$tissue, anns$c1_chip_id)</code></pre>
<pre><code>     1772078217 1772078236 1772078237 nochipID12 nochipID14 nochipID15</code></pre>
<p>cortex 1 28 28 1 3 2</p>
<pre><code>     nochipID2 nochipID3 nochipID8 nochipID9</code></pre>
<p>cortex 57 17 37 16</p>
<pre class="r"><code>table(anns$cell_type, anns$sample_name)</code></pre>
<pre><code>         AB_S1 AB_S11 AB_S2 AB_S3 AB_S4 AB_S5 AB_S7 AB_S8</code></pre>
<p>astrocytes 11 7 2 0 38 0 3 1 neurons 6 49 1 2 19 17 34 0</p>
<pre class="r"><code>table(anns$cell_type, anns$tissue)</code></pre>
<pre><code>         cortex</code></pre>
<p>astrocytes 62 neurons 128</p>
<pre class="r"><code>dim(counts)</code></pre>
<p>[1] 23267 190</p>
<pre class="r"><code>dim(anns)</code></pre>
<p>[1] 190 6</p>
<pre class="r"><code>dropout.rate &lt;- sum(counts==0)/(dim(counts)[1]*dim(counts)[2])*100
dropout.rate</code></pre>
<p>[1] 66.32653</p>
<pre class="r"><code>## Filter out lowly-expressed genes 
filter = apply(counts, 1, function(x) length(x[x&gt;5])&gt;=3)
counts = as.matrix(counts[filter,])
dim(counts)</code></pre>
<p>[1] 21806 190</p>
<pre class="r"><code>dropout.rate &lt;- sum(counts==0)/(dim(counts)[1]*dim(counts)[2])*100
dropout.rate</code></pre>
<p>[1] 64.13206</p>
<pre class="r"><code># To make this task more challenging for benchamarking purposes
# we only used randomly selected 1000 genes in downstream analyses
rand.gene.list &lt;- sample(1:nrow(counts), num.genes)
counts &lt;- counts[rand.gene.list,]

Cell_Type &lt;- anns$cell_type
Tissue &lt;- anns$tissue
Cell_Type_Num &lt;- as.integer(Cell_Type)
Samples &lt;- as.integer(anns$sample_name)</code></pre>
</div>
<div id="calculate-geometric-library-size-i.e.-library-size-of-log-transfromed-read-counts" class="section level2">
<h2>Calculate geometric library size, i.e., library size of log-transfromed read counts</h2>
<p>It is well known that the geometric library size (i.e., library size of log-transfromed read counts) or proportion of expressed genes in each cell explains very large portion of variability of scRNA-Seq data (<a href="http://biorxiv.org/content/early/2015/08/25/025528">Hicks et. al. 2015 BioRxiv</a>, <a href="http://www.nature.com/nbt/journal/v34/n6/full/nbt.3498.html">McDavid et. al. 2016 Nature Biotechnology</a>). Frequently, the first principal component of log-transformed scRNA-Seq read counts is highly correlated with the geometric library size. (r &gt; 0.9). Here, we calculate it and show the high correlation. Later this vector will be used as an known factor in IA-SVA algorithm.</p>
<pre class="r"><code>Geo_Lib_Size &lt;- colSums(log(counts+1))
barplot(Geo_Lib_Size, xlab=&quot;Cell&quot;)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/geno_lib_size-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>lcounts &lt;- log(counts + 1)
pca.res = svd(lcounts - rowMeans(lcounts))$v
cor(Geo_Lib_Size, pca.res[,1])</code></pre>
<pre><code>[1] 0.9435376</code></pre>
<pre class="r"><code>cor(Geo_Lib_Size, Cell_Type_Num)</code></pre>
<pre><code>[1] 0.1815641</code></pre>
</div>
<div id="run-analyses" class="section level2">
<h2>run analyses</h2>
<p>Library size, tissue, and sample ID are used as covariates in all factor analyses except for PCA. 3 SVs are captured with each method and the SV that has the maximum correlation with the cell type assignments is selected among these 3. This is especially important for a fair comparison of PCA.</p>
<pre class="r"><code>mod1 = model.matrix(~Geo_Lib_Size+anns$sample_name)
mod0 = cbind(mod1[,1])
summ_exp &lt;- SummarizedExperiment(assays = counts)
### IA-SVA 
start_time &lt;- Sys.time()
iasva.res&lt;- iasva(summ_exp, mod1[,-1],verbose=FALSE, permute=FALSE, num.sv=3)</code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
SV 2 Detected!</code></pre>
<pre><code>
SV 3 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 3</code></pre>
<pre class="r"><code>end_time &lt;- Sys.time()
iasva.time &lt;- start_time - end_time
iasva.sv &lt;- iasva.res$sv
plot(iasva.sv[,1], iasva.sv[,2],col=Cell_Type, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pairs(iasva.sv, col=Cell_Type, main=&quot;IA-SVA&quot;)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-2.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>max.cor.iasva &lt;- which.is.max(c(abs(cor(iasva.sv[,1], Cell_Type_Num)),
                abs(cor(iasva.sv[,2], Cell_Type_Num)),
                abs(cor(iasva.sv[,3], Cell_Type_Num))))
max.cor.iasva                </code></pre>
<pre><code>[1] 1</code></pre>
<pre class="r"><code>## PCA
ldat0 = log(counts + 1)
start_time &lt;- Sys.time()
pca.sv = svd(ldat0 - rowMeans(ldat0))$v
end_time &lt;- Sys.time()
pca.time &lt;- start_time - end_time
plot(pca.sv[,1], pca.sv[,2],col=Cell_Type, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-3.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pairs(pca.sv[,1:3], col=Cell_Type, main=&quot;PCA&quot;)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-4.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>max.cor.pca &lt;- which.is.max(c(abs(cor(pca.sv[,1], Cell_Type_Num)),
                abs(cor(pca.sv[,2], Cell_Type_Num)),
                abs(cor(pca.sv[,3], Cell_Type_Num))))
max.cor.pca</code></pre>
<pre><code>[1] 2</code></pre>
<pre class="r"><code>## USVA
start_time &lt;- Sys.time()
usva.sv = svaseq(counts,mod1,mod0,n.sv=3)$sv</code></pre>
<pre><code>Number of significant surrogate variables is:  3 
Iteration (out of 5 ):1  2  3  4  5  </code></pre>
<pre class="r"><code>end_time &lt;- Sys.time()
usva.time &lt;- start_time - end_time
plot(usva.sv[,1], usva.sv[,2],col=Cell_Type, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-5.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pairs(usva.sv[,1:3], col=Cell_Type, main=&quot;USVA&quot;)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-6.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>max.cor.usva &lt;- which.is.max(c (abs(cor(usva.sv[,1], Cell_Type_Num)),
                abs(cor(usva.sv[,2], Cell_Type_Num)),
                abs(cor(usva.sv[,3], Cell_Type_Num))))
max.cor.usva</code></pre>
<pre><code>[1] 2</code></pre>
<pre class="r"><code>### RUVres
design &lt;- model.matrix(~Geo_Lib_Size+anns$sample_name)
start_time &lt;- Sys.time()
y &lt;- DGEList(counts=counts)
y &lt;- calcNormFactors(y, method=&quot;TMM&quot;)
y &lt;- estimateGLMCommonDisp(y, design)
y &lt;- estimateGLMTagwiseDisp(y, design)
fit &lt;- glmFit(y, design)
res &lt;- residuals(fit, type=&quot;deviance&quot;)
set &lt;- betweenLaneNormalization(counts, which=&quot;upper&quot;)
genes = rep(TRUE,dim(counts)[1])
ruvres.sv = RUVr(set,genes,k=3,res,isLog = T)$W
end_time &lt;- Sys.time()
ruvres.time &lt;- start_time - end_time
plot(ruvres.sv[,1], ruvres.sv[,2],col=Cell_Type, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-7.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pairs(ruvres.sv[,1:3], col=Cell_Type)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-8.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>max.cor.ruvres &lt;- which.is.max (c (abs(cor(ruvres.sv[,1], Cell_Type_Num)),
                abs(cor(ruvres.sv[,2], Cell_Type_Num)),
                abs(cor(ruvres.sv[,3], Cell_Type_Num))))
max.cor.ruvres</code></pre>
<pre><code>[1] 1</code></pre>
<pre class="r"><code>### RUVemp
start_time &lt;- Sys.time()
y &lt;- DGEList(counts=counts)
y &lt;- calcNormFactors(y, method=&quot;TMM&quot;)
y &lt;- estimateGLMCommonDisp(y, design)
y &lt;- estimateGLMTagwiseDisp(y, design)
fit &lt;- glmFit(y, design)
lrt &lt;- glmLRT(fit, coef=2)
emp.genes =rank(lrt$table$LR) &gt; 400
ruvemp.sv &lt;- RUVg(counts, emp.genes, k=3)$W
end_time &lt;- Sys.time()
ruvemp.time &lt;- start_time - end_time
plot(ruvemp.sv[,1], ruvemp.sv[,2],col=Cell_Type, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-9.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pairs(ruvemp.sv[,1:3], col=Cell_Type)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-10.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>max.cor.ruvemp &lt;- which.is.max(c (abs(cor(ruvemp.sv[,1], Cell_Type_Num)),
                abs(cor(ruvemp.sv[,2], Cell_Type_Num)),
                abs(cor(ruvemp.sv[,3], Cell_Type_Num))))
max.cor.ruvemp</code></pre>
<pre><code>[1] 2</code></pre>
<pre class="r"><code>### ZINB WAVE
se.data &lt;- SummarizedExperiment(assays=list(counts=counts))
se.data$LS &lt;- Geo_Lib_Size
se.data$SN &lt;- anns$sample_name
se.data$Tissue &lt;- anns$tissue
# use Geometric library size + Tissue + sample is as covariates
# suggested epsilon parameter is used for the analyses
start_time &lt;- Sys.time()
zinb_cov &lt;- zinbFit(se.data, K=3, X=&quot;~LS+SN&quot;, epsilon=1000)</code></pre>
<pre><code>Warning in simpute.als(x, J, thresh, lambda, maxit, trace.it, warm.start, :
Convergence not achieved by 100 iterations</code></pre>
<pre class="r"><code>W &lt;- getW(zinb_cov)
end_time &lt;- Sys.time()
zinb.time &lt;- start_time - end_time
colnames(W) &lt;- paste0(&quot;W&quot;, 1:3)
pairs(W,col=Cell_Type)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/run_analyses-11.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>max.cor.zinb &lt;- which.is.max (c (abs(cor(W[,1], Cell_Type_Num)),
                abs(cor(W[,2], Cell_Type_Num)),
                abs(cor(W[,3], Cell_Type_Num))))

max.cor.zinb</code></pre>
<pre><code>[1] 1</code></pre>
</div>
<div id="compare-results" class="section level2">
<h2>Compare results</h2>
<p>Compare results from different methods to the true cell type assignments using correlation plots.</p>
<pre class="r"><code>## Plot correlation between true cell_type and estimates (SV1)
celltype.est = cbind(Cell_Type_Num, iasva.sv[,max.cor.iasva],
                     pca.sv[,max.cor.pca],usva.sv[,max.cor.usva], 
                     ruvres.sv[,max.cor.ruvres], 
                     ruvemp.sv[,max.cor.ruvemp], W[,max.cor.zinb])

colnames(celltype.est) = c(&quot;Cell Type&quot;,&quot;IA-SVA&quot;,&quot;PCA&quot;,&quot;USVA&quot;,
                           &quot;RUVres&quot;,&quot;RUVemp&quot;, &quot;ZINB-WAVE&quot;)
corr = abs(cor(celltype.est))
print(corr)</code></pre>
<pre><code>          Cell Type    IA-SVA       PCA      USVA    RUVres    RUVemp
Cell Type 1.0000000 0.9364864 0.8412577 0.9217416 0.7647545 0.8358981
IA-SVA    0.9364864 1.0000000 0.9241849 0.9441659 0.7767147 0.9186404
PCA       0.8412577 0.9241849 1.0000000 0.8876566 0.5462038 0.9959706
USVA      0.9217416 0.9441659 0.8876566 1.0000000 0.7562104 0.8756892
RUVres    0.7647545 0.7767147 0.5462038 0.7562104 1.0000000 0.5373816
RUVemp    0.8358981 0.9186404 0.9959706 0.8756892 0.5373816 1.0000000
ZINB-WAVE 0.8255676 0.8472393 0.8045799 0.8615873 0.6648695 0.7891535
          ZINB-WAVE
Cell Type 0.8255676
IA-SVA    0.8472393
PCA       0.8045799
USVA      0.8615873
RUVres    0.6648695
RUVemp    0.7891535
ZINB-WAVE 1.0000000</code></pre>
<pre class="r"><code>cols = colorRampPalette(c(color.vec[2],&quot;white&quot;,color.vec[1]))

corrplot.mixed(corr, lower=&quot;ellipse&quot;, upper=&quot;number&quot;, mar=c(0,0,1,0))</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/plot_cor-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>num.cells &lt;- dim(anns)[1]
pdf(paste0(&quot;output/Brain_scRNASeq_neuron_astro_&quot;,num.cells,&quot;_cells.pdf&quot;), height=6, width=6)
corrplot.mixed(corr, lower=&quot;ellipse&quot;, upper=&quot;number&quot;, mar=c(0,0,1,0))
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>time.seconds &lt;- c(iasva.time,pca.time,usva.time,ruvres.time,ruvemp.time,zinb.time)
names(time.seconds) &lt;- c(&quot;IA-SVA&quot;,&quot;PCA&quot;,&quot;USVA&quot;,&quot;RUVres&quot;,&quot;RUVemp&quot;, &quot;ZINB-WAVE&quot;)
time.seconds</code></pre>
<pre><code>Time differences in secs
     IA-SVA         PCA        USVA      RUVres      RUVemp   ZINB-WAVE 
 -0.2537930  -0.1267381  -1.1145370 -73.1166251 -72.4422059 -52.8407030 </code></pre>
<pre class="r"><code>barplot(t(as.matrix(abs(time.seconds))), las =2, ylab = &quot;Run time (seconds)&quot;)</code></pre>
<p><img src="figure/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd/plot_cor-2.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pdf(paste0(&quot;output/Brain_scRNASeq_neuron_astro_&quot;,num.cells,&quot;_cells_run_time.pdf&quot;), height=6, width=6)
barplot(t(as.matrix(abs(time.seconds))), las =2, ylab = &quot;Run time (seconds)&quot;)
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.0 (2018-04-23)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] nnet_7.3-12                 zinbwave_1.2.0             
 [3] SingleCellExperiment_1.2.0  RColorBrewer_1.1-2         
 [5] DescTools_0.99.24           corrplot_0.84              
 [7] pheatmap_1.0.10             Rtsne_0.13                 
 [9] RUVSeq_1.14.0               edgeR_3.22.3               
[11] limma_3.36.2                EDASeq_2.14.1              
[13] ShortRead_1.38.0            GenomicAlignments_1.16.0   
[15] SummarizedExperiment_1.10.1 DelayedArray_0.6.1         
[17] matrixStats_0.53.1          Rsamtools_1.32.2           
[19] GenomicRanges_1.32.3        GenomeInfoDb_1.16.0        
[21] Biostrings_2.48.0           XVector_0.20.0             
[23] IRanges_2.14.10             S4Vectors_0.18.3           
[25] Biobase_2.40.0              BiocGenerics_0.26.0        
[27] sva_3.28.0                  BiocParallel_1.14.2        
[29] genefilter_1.62.0           mgcv_1.8-23                
[31] nlme_3.1-137                iasvaExamples_1.0.0        
[33] iasva_0.99.3               

loaded via a namespace (and not attached):
 [1] bitops_1.0-6           bit64_0.9-7            progress_1.2.0        
 [4] httr_1.3.1             rprojroot_1.3-2        numDeriv_2016.8-1     
 [7] tools_3.5.0            backports_1.1.2        R6_2.2.2              
[10] irlba_2.3.2            DBI_1.0.0              colorspace_1.3-2      
[13] prettyunits_1.0.2      bit_1.1-14             compiler_3.5.0        
[16] git2r_0.21.0           glmnet_2.0-16          pspline_1.0-18        
[19] expm_0.999-2           rtracklayer_1.40.3     scales_0.5.0          
[22] mvtnorm_1.0-8          DESeq_1.32.0           stringr_1.3.1         
[25] digest_0.6.15          foreign_0.8-70         rmarkdown_1.9         
[28] R.utils_2.6.0          pkgconfig_2.0.1        htmltools_0.3.6       
[31] manipulate_1.0.1       stabledist_0.7-1       ADGofTest_0.3         
[34] rlang_0.2.1            RSQLite_2.1.1          hwriter_1.3.2         
[37] R.oo_1.22.0            RCurl_1.95-4.10        magrittr_1.5          
[40] GenomeInfoDbData_1.1.0 Matrix_1.2-14          Rcpp_0.12.17          
[43] munsell_0.4.3          R.methodsS3_1.7.1      stringi_1.2.2         
[46] whisker_0.3-2          yaml_2.1.19            MASS_7.3-50           
[49] zlibbioc_1.26.0        plyr_1.8.4             grid_3.5.0            
[52] blob_1.1.1             crayon_1.3.4           lattice_0.20-35       
[55] splines_3.5.0          GenomicFeatures_1.32.0 annotate_1.58.0       
[58] hms_0.4.2              locfit_1.5-9.1         knitr_1.20            
[61] boot_1.3-20            softImpute_1.4         codetools_0.2-15      
[64] geneplotter_1.58.0     biomaRt_2.36.1         XML_3.98-1.11         
[67] evaluate_0.10.1        latticeExtra_0.6-28    foreach_1.4.4         
[70] gtable_0.2.0           assertthat_0.2.0       aroma.light_3.10.0    
[73] xtable_1.8-2           pcaPP_1.9-73           survival_2.42-3       
[76] gsl_1.9-10.3           copula_0.999-18        iterators_1.0.9       
[79] AnnotationDbi_1.42.1   memoise_1.1.0          workflowr_1.0.1       
[82] cluster_2.0.7-1       </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1
</p>
<hr>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
