<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Donghyung Lee" />

<meta name="date" content="2018-08-03" />

<title>Detecting hidden heterogeneity in single cell RNA-Seq data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">iasvaExamples</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/dleelab/iasvaExamples">GitHub</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Detecting hidden heterogeneity in single cell RNA-Seq data</h1>
<h4 class="author"><em>Donghyung Lee</em></h4>
<h4 class="date"><em>2018-08-03</em></h4>

</div>

<div id="TOC">
<ul>
<li><a href="#install-packages">Install packages</a></li>
<li><a href="#load-packages">Load packages</a></li>
<li><a href="#load-the-islet-single-cell-rna-seq-data">Load the islet single cell RNA-Seq data</a></li>
<li><a href="#extract-alpha-cells-gcg-expressed-cells-from-non-diabetics">Extract alpha cells (GCG expressed cells) from non-diabetics</a></li>
<li><a href="#normalization-using-scnorm">Normalization using SCnorm</a></li>
<li><a href="#calculate-the-number-of-detected-genes">Calculate the number of detected genes</a></li>
<li><a href="#run-ia-sva">Run IA-SVA</a></li>
<li><a href="#find-marker-genes-for-the-detected-heterogeneity-sv2.">Find marker genes for the detected heterogeneity (SV2).</a></li>
<li><a href="#run-tsne-to-detect-the-hidden-heterogeneity.">Run tSNE to detect the hidden heterogeneity.</a></li>
<li><a href="#run-principal-component-analysis-pca-to-detect-the-hidden-heterogeneity-sv2.">Run principal component analysis (PCA) to detect the hidden heterogeneity (SV2).</a></li>
<li><a href="#run-surrogate-variable-analysis-sva-to-detect-the-hidden-heterogeneity-sv2.">Run surrogate variable analysis (SVA) to detect the hidden heterogeneity (SV2).</a></li>
<li><a href="#correlation-between-sv2-and-the-geometric-library-size">Correlation between SV2 and the geometric library size</a></li>
<li><a href="#session-information">Session information</a></li>
</ul>
</div>

<p><strong>Last updated:</strong> 2018-08-03</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(20180731)</code> </summary></p>
<p>The command <code>set.seed(20180731)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/dleelab/iasvaExamples/tree/6e130ac4b06b66f39579264f0513cc9a690335a4" target="_blank">6e130ac</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    inst/.DS_Store
    Ignored:    inst/doc/.DS_Store
    Ignored:    vignettes/.DS_Store

Untracked files:
    Untracked:  analysis/figure/
    Untracked:  analysis/hidden_heterogeneity_glioblastoma.Rmd
    Untracked:  output/Lawlor_Islets_Alpha_Doublets_Figure2_ABCD.pdf
    Untracked:  output/Lawlor_Islets_Alpha_Doublets_SV2_Genes_rsqcutoff0.3.txt
    Untracked:  output/Lawlor_Islets_Alpha_Doublets_SV2_Genes_rsqcutoff0.6.txt
    Untracked:  output/Lawlor_Islets_Alpha_iasva_SV2Markers_rsqcutoff0.3_pheatmap_iasvaV0.95.pdf
    Untracked:  output/Lawlor_Islets_Alpha_iasva_SV2Markers_rsqcutoff0.6_pheatmap_iasvaV0.95.pdf

Unstaged changes:
    Deleted:    analysis/about.Rmd
    Modified:   analysis/license.Rmd
    Modified:   output/Clustering_analyses_figure2.pdf

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/6e130ac4b06b66f39579264f0513cc9a690335a4/analysis/detecting_hidden_heterogeneity.Rmd" target="_blank">6e130ac</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
SCnorm codes updated
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/13e8676267f6c0206e1b63bb7cbb5ce0aae6f611/analysis/detecting_hidden_heterogeneity.Rmd" target="_blank">13e8676</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
figure size updated
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/dleelab/iasvaExamples/c241b04dac899407045051a04b7539237f4d183f/docs/detecting_hidden_heterogeneity.html" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/bae72f794f1550d5582823fe3636d0a5fad0a8fe/analysis/detecting_hidden_heterogeneity.Rmd" target="_blank">bae72f7</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-01
</td>
<td style="text-align:left;">
figure updated
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<p>The iasva package can be used to detect hidden heterogenity within bulk or single cell sequencing data. To illustrate how to use the iasva package for heterogenity detection, we use real-world single cell RNA sequencing (scRNA-Seq) data obtained from human pancreatic islet samples (<a href="http://genome.cshlp.org/content/early/2017/01/16/gr.212720.116">Lawlor et. al., 2016</a>). This dataset is included in a R data package (“iasvaExamples”) containing data examples for IA-SVA (<a href="https://github.com/dleelab/iasvaExamples" class="uri">https://github.com/dleelab/iasvaExamples</a>). To install the package, follow the instruction provided in the GitHub page.</p>
<div id="install-packages" class="section level2">
<h2>Install packages</h2>
<pre class="r"><code>#devtools
library(devtools)
#iasva
devtools::install_github(&quot;UcarLab/iasva&quot;)
#iasvaExamples  
devtools::install_github(&quot;dleelab/iasvaExamples&quot;)</code></pre>
</div>
<div id="load-packages" class="section level2">
<h2>Load packages</h2>
<pre class="r"><code>rm(list=ls())
library(irlba) # partial SVD, the augmented implicitly restarted Lanczos bidiagonalization algorithm
library(iasva)
library(iasvaExamples)
library(sva)
library(SCnorm)</code></pre>
<pre><code>Warning: package &#39;SCnorm&#39; was built under R version 3.5.1</code></pre>
<pre class="r"><code>library(Rtsne)
library(pheatmap)
library(corrplot)
library(DescTools) #pcc i.e., Pearson&#39;s contingency coefficient
library(RColorBrewer)
library(SummarizedExperiment)

color.vec &lt;- brewer.pal(9, &quot;Set1&quot;)[c(1,9)]

# Normalization.
normalize &lt;- function(counts) 
{
    normfactor &lt;- colSums(counts)
    return(t(t(counts)/normfactor)*median(normfactor))
}</code></pre>
</div>
<div id="load-the-islet-single-cell-rna-seq-data" class="section level2">
<h2>Load the islet single cell RNA-Seq data</h2>
<pre class="r"><code>data(&quot;Lawlor_Islet_scRNAseq_Read_Counts&quot;)
data(&quot;Lawlor_Islet_scRNAseq_Annotations&quot;)
ls()</code></pre>
<pre><code>[1] &quot;color.vec&quot;                         &quot;Lawlor_Islet_scRNAseq_Annotations&quot;
[3] &quot;Lawlor_Islet_scRNAseq_Read_Counts&quot; &quot;normalize&quot;                        </code></pre>
<pre class="r"><code>counts &lt;- Lawlor_Islet_scRNAseq_Read_Counts
anns &lt;- Lawlor_Islet_scRNAseq_Annotations
dim(anns)</code></pre>
<pre><code>[1] 638  26</code></pre>
<pre class="r"><code>dim(counts)</code></pre>
<pre><code>[1] 26542   638</code></pre>
<pre class="r"><code>summary(anns)</code></pre>
<pre><code>     run             cell.type             COL1A1          INS       
 Length:638         Length:638         Min.   :1.00   Min.   :1.000  
 Class :character   Class :character   1st Qu.:1.00   1st Qu.:1.000  
 Mode  :character   Mode  :character   Median :1.00   Median :1.000  
                                       Mean   :1.03   Mean   :1.414  
                                       3rd Qu.:1.00   3rd Qu.:2.000  
                                       Max.   :2.00   Max.   :2.000  
                                                                     
     PRSS1            SST             GCG            KRT19      
 Min.   :1.000   Min.   :1.000   Min.   :1.000   Min.   :1.000  
 1st Qu.:1.000   1st Qu.:1.000   1st Qu.:1.000   1st Qu.:1.000  
 Median :1.000   Median :1.000   Median :1.000   Median :1.000  
 Mean   :1.038   Mean   :1.039   Mean   :1.375   Mean   :1.044  
 3rd Qu.:1.000   3rd Qu.:1.000   3rd Qu.:2.000   3rd Qu.:1.000  
 Max.   :2.000   Max.   :2.000   Max.   :2.000   Max.   :2.000  
                                                                
      PPY          num.genes             Cell_ID        UNOS_ID   
 Min.   :1.000   Min.   :3529   10th_C1_S59  :  1   ACCG268 :136  
 1st Qu.:1.000   1st Qu.:5270   10th_C10_S104:  1   ACJV399 :108  
 Median :1.000   Median :6009   10th_C11_S96 :  1   ACEL337 :103  
 Mean   :1.028   Mean   :5981   10th_C13_S61 :  1   ACIW009 : 93  
 3rd Qu.:1.000   3rd Qu.:6676   10th_C14_S53 :  1   ACCR015A: 57  
 Max.   :2.000   Max.   :8451   10th_C16_S105:  1   ACIB065 : 57  
                                (Other)      :632   (Other) : 84  
      Age        Biomaterial_Provider    Gender              Phenotype  
 Min.   :22.00   IIDP      : 45       Female:303   Non-Diabetic   :380  
 1st Qu.:29.00   Prodo Labs:593       Male  :335   Type 2 Diabetic:258  
 Median :42.00                                                          
 Mean   :39.63                                                          
 3rd Qu.:53.00                                                          
 Max.   :56.00                                                          
                                                                        
               Race          BMI          Cell_Type     Patient_ID 
 African American:175   Min.   :22.00   INS    :264   P1     :136  
 Hispanic        :165   1st Qu.:26.60   GCG    :239   P8     :108  
 White           :298   Median :32.95   KRT19  : 28   P3     :103  
                        Mean   :32.85   SST    : 25   P7     : 93  
                        3rd Qu.:35.80   PRSS1  : 24   P5     : 57  
                        Max.   :55.00   none   : 21   P6     : 57  
                                        (Other): 37   (Other): 84  
 Sequencing_Run Batch       Coverage       Percent_Aligned 
 12t    : 57    B1:193   Min.   :1206135   Min.   :0.8416  
 4th    : 57    B2:148   1st Qu.:2431604   1st Qu.:0.8769  
 9th    : 57    B3:297   Median :3042800   Median :0.8898  
 10t    : 56             Mean   :3160508   Mean   :0.8933  
 7th    : 55             3rd Qu.:3871697   3rd Qu.:0.9067  
 3rd    : 53             Max.   :5931638   Max.   :0.9604  
 (Other):303                                               
 Mitochondrial_Fraction Num_Expressed_Genes
 Min.   :0.003873       Min.   :3529       
 1st Qu.:0.050238       1st Qu.:5270       
 Median :0.091907       Median :6009       
 Mean   :0.108467       Mean   :5981       
 3rd Qu.:0.142791       3rd Qu.:6676       
 Max.   :0.722345       Max.   :8451       
                                           </code></pre>
<pre class="r"><code>ContCoef(table(anns$Gender, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.225969</code></pre>
<pre class="r"><code>ContCoef(table(anns$Phenotype, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.1145096</code></pre>
<pre class="r"><code>ContCoef(table(anns$Race, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.3084146</code></pre>
<pre class="r"><code>ContCoef(table(anns$Patient_ID, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.5232058</code></pre>
<pre class="r"><code>ContCoef(table(anns$Batch, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.3295619</code></pre>
<p>The annotations describing the islet samples and experimental settings are stored in “anns” and the read counts information is stored in “counts”.</p>
</div>
<div id="extract-alpha-cells-gcg-expressed-cells-from-non-diabetics" class="section level2">
<h2>Extract alpha cells (GCG expressed cells) from non-diabetics</h2>
<p>To illustrate how IA-SVA can be used to detect hidden heterogeneity within a homogenous cell population (i.e., alpha cells), we use read counts of alpha cells from healthy (non-diabetic) subjects (n = 101).</p>
<pre class="r"><code># Selected T2D patients/GCG cell type
counts &lt;- counts[, (anns$Phenotype!=&quot;Non-Diabetic&quot;)&amp;(anns$Cell_Type==&quot;GCG&quot;)]
anns &lt;- subset(anns, (Phenotype!=&quot;Non-Diabetic&quot;)&amp;(Cell_Type==&quot;GCG&quot;))
dim(counts)</code></pre>
<p>[1] 26542 101</p>
<pre class="r"><code>dim(anns)</code></pre>
<p>[1] 101 26</p>
<pre class="r"><code>anns &lt;- droplevels(anns)
prop.zeros &lt;- sum(counts==0)/length(counts)
prop.zeros</code></pre>
<p>[1] 0.6952497</p>
<pre class="r"><code># filter out genes that are sparsely and lowly expressed
filter = apply(counts, 1, function(x) length(x[x&gt;5])&gt;=3)
counts = counts[filter,]
dim(counts)</code></pre>
<p>[1] 14384 101</p>
<pre class="r"><code>prop.zeros &lt;- sum(counts==0)/length(counts)
prop.zeros</code></pre>
<p>[1] 0.4520968</p>
</div>
<div id="normalization-using-scnorm" class="section level2">
<h2>Normalization using SCnorm</h2>
<pre class="r"><code>## count-depth relationship for all genes
Conditions = rep(c(1), each=101)
countDeptEst &lt;- plotCountDepth(Data = counts, Conditions = Conditions,
                               FilterCellProportion = .1, NCores=8)</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/normalization-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>DataNorm &lt;- SCnorm(Data = counts, Conditions = Conditions,
                   PrintProgressPlots = FALSE,
                   FilterCellNum = 10,
                   NCores=8)</code></pre>
<pre><code>Setting up parallel computation using 8 cores</code></pre>
<pre><code>Gene filter is applied within each condition.</code></pre>
<pre><code>1223 genes in condition 1 will not be included in the normalization due to 
             the specified filter criteria.</code></pre>
<pre><code>A list of these genes can be accessed in output, 
    see vignette for example.</code></pre>
<pre><code>Finding K for Condition 1</code></pre>
<pre><code>Trying K = 1</code></pre>
<pre><code>Trying K = 2</code></pre>
<pre><code>Trying K = 3</code></pre>
<pre><code>Trying K = 4</code></pre>
<pre><code>Trying K = 5</code></pre>
<pre><code>Trying K = 6</code></pre>
<pre><code>Trying K = 7</code></pre>
<pre><code>Trying K = 8</code></pre>
<pre><code>Done!</code></pre>
<pre class="r"><code>counts &lt;- SingleCellExperiment::normcounts(DataNorm)
summary(colSums(counts))</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
1274488 1315313 1332682 1358957 1386293 1650415 </code></pre>
</div>
<div id="calculate-the-number-of-detected-genes" class="section level2">
<h2>Calculate the number of detected genes</h2>
<p>It is well known that the number of detected genes in each cell explains a very large portion of variability in scRNA-Seq data (<a href="http://biorxiv.org/content/early/2015/08/25/025528">Hicks et. al. 2015 BioRxiv</a>, <a href="http://www.nature.com/nbt/journal/v34/n6/full/nbt.3498.html">McDavid et. al. 2016 Nature Biotechnology</a>). Frequently, the first principal component of log-transformed scRNA-Seq read counts is highly correlated with the number of detected genes (e.g., r &gt; 0.9). Here, we calculate the number of detected genes for islet cells, which will be used as an known factor in the IA-SVA analyses.</p>
<pre class="r"><code>Num_Detected_Genes &lt;- colSums(counts&gt;0)
Geo_Lib &lt;- colSums(log(counts+1))
summary(Num_Detected_Genes)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   5787    7043    7925    7881    8652   10188 </code></pre>
<pre class="r"><code>summary(Geo_Lib)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  16571   20527   24507   24349   27475   33830 </code></pre>
<pre class="r"><code>barplot(Num_Detected_Genes, xlab=&quot;Cell&quot;, las=2,
        ylab = &quot;Number of detected genes&quot;)</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/num_detected_genes-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>lcounts &lt;- log(counts + 1)

# PC1 and Geometric library size correlation
pc1 = irlba(lcounts - rowMeans(lcounts), 1)$v[,1] ## partial SVD
cor(Num_Detected_Genes, pc1)</code></pre>
<pre><code>[1] 0.9778019</code></pre>
<pre class="r"><code>cor(Geo_Lib, pc1)</code></pre>
<pre><code>[1] 0.99461</code></pre>
</div>
<div id="run-ia-sva" class="section level2">
<h2>Run IA-SVA</h2>
<p>Here, we run IA-SVA using Patient_ID and Geo_Lib_Size as known factors and identify five hidden factors. SVs are plotted in a pairwise fashion to uncover which SVs can seperate cell types.</p>
<pre class="r"><code>set.seed(4543535)
Patient_ID &lt;- anns$Patient_ID
mod &lt;- model.matrix(~Patient_ID+Geo_Lib)
summ_exp &lt;- SummarizedExperiment(assays = counts)
iasva.res&lt;- iasva(summ_exp, mod[,-1],verbose=FALSE, permute=FALSE, num.sv=5) ##irlba</code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
SV 2 Detected!</code></pre>
<pre><code>
SV 3 Detected!</code></pre>
<pre><code>
SV 4 Detected!</code></pre>
<pre><code>
SV 5 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 5</code></pre>
<pre class="r"><code>iasva.sv &lt;- iasva.res$sv
plot(iasva.sv[,1], iasva.sv[,2], xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;)</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/run_iasva-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>Cluster &lt;- as.factor(iasva.sv[,2] &lt; 0.1) 
levels(Cluster)=c(&quot;Cell1&quot;,&quot;Cell2&quot;)
table(Cluster)</code></pre>
<pre><code>Cluster
Cell1 Cell2 
    6    95 </code></pre>
<pre class="r"><code># We identified 6 outlier cells based on SV2 that are marked in red
pairs(iasva.sv, main=&quot;IA-SVA&quot;, pch=21, col=color.vec[Cluster],
      bg=color.vec[Cluster], oma=c(4,4,6,12)) #4,4,6,12
legend(&quot;right&quot;, levels(Cluster), fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/run_iasva-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(iasva.sv[,1:2], main=&quot;IA-SVA&quot;, pch=21, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;,
     col=color.vec[Cluster], bg=color.vec[Cluster])</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/run_iasva-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>cor(Num_Detected_Genes, iasva.sv[,2])</code></pre>
<pre><code>[1] 0.1571675</code></pre>
<pre class="r"><code>cor(Geo_Lib, iasva.sv[,2])</code></pre>
<pre><code>[1] 0.2009796</code></pre>
<pre class="r"><code>corrplot(cor(iasva.sv))</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/run_iasva-4.png" width="672" style="display: block; margin: auto;" /></p>
<p>As shown in the above figure, SV2 clearly separates alpha cells into two groups: 6 outlier cells (marked in red) and the rest of the alpha cells (marked in green). SV3 and SV4 also capture outlier cells. However, we will focus on SV2 in the rest of the analyses.</p>
</div>
<div id="find-marker-genes-for-the-detected-heterogeneity-sv2." class="section level2">
<h2>Find marker genes for the detected heterogeneity (SV2).</h2>
<p>Here, using the find_markers() function we find marker genes (n=105 genes) that are significantly associated with SV2 (multiple testing adjusted p-value &lt; 0.05, default significance cutoff, and R-squared value &gt; 0.3, default R-squared cutoff).</p>
<pre class="r"><code># try different R2 thresholds
pdf(&quot;output/Clustering_analyses_figure2.pdf&quot;)
r2.results &lt;- study_R2(summ_exp, iasva.sv, selected.svs=2, no.clusters=2) </code></pre>
<pre><code># of markers (): 396</code></pre>
<pre><code>total # of unique markers: 396</code></pre>
<pre><code># of markers (): 237</code></pre>
<pre><code>total # of unique markers: 237</code></pre>
<pre><code># of markers (): 177</code></pre>
<pre><code>total # of unique markers: 177</code></pre>
<pre><code># of markers (): 143</code></pre>
<pre><code>total # of unique markers: 143</code></pre>
<pre><code># of markers (): 108</code></pre>
<pre><code>total # of unique markers: 108</code></pre>
<pre><code># of markers (): 93</code></pre>
<pre><code>total # of unique markers: 93</code></pre>
<pre><code># of markers (): 72</code></pre>
<pre><code>total # of unique markers: 72</code></pre>
<pre><code># of markers (): 58</code></pre>
<pre><code>total # of unique markers: 58</code></pre>
<pre><code># of markers (): 47</code></pre>
<pre><code>total # of unique markers: 47</code></pre>
<pre><code># of markers (): 35</code></pre>
<pre><code>total # of unique markers: 35</code></pre>
<pre><code># of markers (): 27</code></pre>
<pre><code>total # of unique markers: 27</code></pre>
<pre><code># of markers (): 22</code></pre>
<pre><code>total # of unique markers: 22</code></pre>
<pre><code># of markers (): 12</code></pre>
<pre><code>total # of unique markers: 12</code></pre>
<pre><code># of markers (): 8</code></pre>
<pre><code>total # of unique markers: 8</code></pre>
<pre><code># of markers (): 4</code></pre>
<pre><code>total # of unique markers: 4</code></pre>
<pre><code># of markers (): 2</code></pre>
<pre><code>total # of unique markers: 2</code></pre>
<pre><code># of markers (): 1</code></pre>
<pre><code>total # of unique markers: 1</code></pre>
<pre class="r"><code>dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>marker.counts &lt;- find_markers(summ_exp, as.matrix(iasva.sv[,2]), rsq.cutoff = 0.6)</code></pre>
<pre><code># of markers (): 27</code></pre>
<pre><code>total # of unique markers: 27</code></pre>
<pre class="r"><code>marker.counts.long &lt;- find_markers(summ_exp, as.matrix(iasva.sv[,2]), rsq.cutoff = 0.3)</code></pre>
<pre><code># of markers (): 108</code></pre>
<pre><code>total # of unique markers: 108</code></pre>
<pre class="r"><code>nrow(marker.counts)</code></pre>
<pre><code>[1] 27</code></pre>
<pre class="r"><code>rownames(marker.counts)</code></pre>
<pre><code> [1] &quot;PMEPA1&quot;      &quot;LINC00152&quot;   &quot;MIR4435-1HG&quot; &quot;ENG&quot;         &quot;ITGA5&quot;      
 [6] &quot;TMEM233&quot;     &quot;PRDM1&quot;       &quot;C8orf4&quot;      &quot;ERG&quot;         &quot;THBS1&quot;      
[11] &quot;MEF2C&quot;       &quot;LGALS1&quot;      &quot;CD93&quot;        &quot;ELTD1&quot;       &quot;COL4A1&quot;     
[16] &quot;COL4A2&quot;      &quot;HBEGF&quot;       &quot;SPARC&quot;       &quot;SPARCL1&quot;     &quot;RAPGEF5&quot;    
[21] &quot;KDR&quot;         &quot;GNG11&quot;       &quot;CD9&quot;         &quot;PODXL&quot;       &quot;PLVAP&quot;      
[26] &quot;IFI16&quot;       &quot;RHOJ&quot;       </code></pre>
<pre class="r"><code>nrow(marker.counts.long)</code></pre>
<pre><code>[1] 108</code></pre>
<pre class="r"><code>anno.col &lt;- data.frame(Cluster=Cluster, SV2=iasva.sv[,2])
rownames(anno.col) &lt;- colnames(marker.counts)
head(anno.col)</code></pre>
<pre><code>            Cluster         SV2
4th-C63_S30   Cell2 -0.02724532
4th-C66_S36   Cell2 -0.02091476
4th-C18_S31   Cell2 -0.01190674
4th-C57_S18   Cell1  0.23502845
4th-C56_S17   Cell2 -0.03704850
4th-C68_S41   Cell2 -0.03570599</code></pre>
<pre class="r"><code>cluster.col &lt;- color.vec[1:2]
names(cluster.col) &lt;- as.vector(levels(Cluster))
anno.colors &lt;- list(Cluster=cluster.col)
anno.colors</code></pre>
<pre><code>$Cluster
    Cell1     Cell2 
&quot;#E41A1C&quot; &quot;#999999&quot; </code></pre>
<pre class="r"><code>pheatmap(log(marker.counts+1), show_colnames =FALSE, 
         clustering_method = &quot;ward.D2&quot;,cutree_cols = 2,annotation_col = anno.col,
         annotation_colors = anno.colors)</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/find_markers-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="run-tsne-to-detect-the-hidden-heterogeneity." class="section level2">
<h2>Run tSNE to detect the hidden heterogeneity.</h2>
<p>For comparison purposes, we applied tSNE on read counts of all genes to identify the hidden heterogeneity. We used the Rtsne R package with default settings.</p>
<pre class="r"><code>set.seed(323542534)
tsne.res &lt;- Rtsne(t(lcounts), dims = 2)
plot(tsne.res$Y, main=&quot;tSNE&quot;, xlab=&quot;tSNE Dim1&quot;, ylab=&quot;tSNE Dim2&quot;, pch=21, 
     col=color.vec[Cluster], bg=color.vec[Cluster], oma=c(4,4,6,12))
legend(&quot;bottomright&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/run_tsne-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>As shown above, tSNE fails to detect the outlier cells that are identified by IA-SVA when all genes are used. Same color coding is used as above.</p>
</div>
<div id="run-principal-component-analysis-pca-to-detect-the-hidden-heterogeneity-sv2." class="section level2">
<h2>Run principal component analysis (PCA) to detect the hidden heterogeneity (SV2).</h2>
<p>Here, we use PCA to detect the hidden heterogeneity (SV2) detected by IA-SVA.</p>
<pre class="r"><code>set.seed(345233)
pca.res = irlba(lcounts - rowMeans(lcounts), 5)$v ## partial SVD

pairs(pca.res, main=&quot;PCA&quot;, pch=21, col=color.vec[Cluster],
      bg=color.vec[Cluster], oma=c(4,4,6,12)) #4,4,6,12
legend(&quot;right&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/run_pca-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(pca.res[,2:3], main=&quot;PCA&quot;, xlab=&quot;PC2&quot;, ylab=&quot;PC3&quot;, pch=21,
     col=color.vec[Cluster], bg=color.vec[Cluster], oma=c(4,4,6,12))
legend(&quot;bottomright&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/run_pca-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>PC3 somewhat captures the six outlier cells, however this seperation is not as clear as the IA-SVA results.</p>
</div>
<div id="run-surrogate-variable-analysis-sva-to-detect-the-hidden-heterogeneity-sv2." class="section level2">
<h2>Run surrogate variable analysis (SVA) to detect the hidden heterogeneity (SV2).</h2>
<p>Here, for comparison purposes we use SVA (using thre SVs) to detect the hidden heterogeneity in our example data.</p>
<pre class="r"><code>mod1 &lt;- model.matrix(~Patient_ID+Geo_Lib)
mod0 &lt;- cbind(mod1[,1])
sva.res = svaseq(counts,mod1,mod0, n.sv=5)$sv</code></pre>
<pre><code>Number of significant surrogate variables is:  5 
Iteration (out of 5 ):1  2  3  4  5  </code></pre>
<pre class="r"><code>pairs(sva.res, main=&quot;SVA&quot;, pch=21, col=color.vec[Cluster], bg=color.vec[Cluster], oma=c(4,4,6,12)) #4,4,6,12
legend(&quot;right&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/run_sva-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(sva.res[,1:2], main=&quot;SVA&quot;, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;, pch=21, col=color.vec[Cluster], bg=color.vec[Cluster])
legend(&quot;topleft&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/detecting_hidden_heterogeneity.Rmd/run_sva-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>SV2 is associated with the six outlier samples, however the seperation of these cells is not as clear as the IA-SVA results.</p>
</div>
<div id="correlation-between-sv2-and-the-geometric-library-size" class="section level2">
<h2>Correlation between SV2 and the geometric library size</h2>
<pre class="r"><code>cor(Num_Detected_Genes, iasva.sv[,2])</code></pre>
<pre><code>[1] 0.1571675</code></pre>
<pre class="r"><code>cor(Geo_Lib, iasva.sv[,2])</code></pre>
<pre><code>[1] 0.2009796</code></pre>
<pre class="r"><code>pdf(file=&quot;output/Lawlor_Islets_Alpha_Doublets_Figure2_ABCD.pdf&quot;, width=5, height=6)
layout(matrix(c(1,2,3,4), nrow=2, ncol=2, byrow=TRUE))
plot(iasva.sv[,1:2], main=&quot;IA-SVA&quot;, pch=21, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;, col=color.vec[Cluster], bg=color.vec[Cluster])
legend(&quot;topright&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)
plot(pca.res[,2:3], main=&quot;PCA&quot;, pch=21, xlab=&quot;PC2&quot;, ylab=&quot;PC3&quot;, col=color.vec[Cluster], bg=color.vec[Cluster])
plot(sva.res[,1:2], main=&quot;USVA&quot;, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;, pch=21, col=color.vec[Cluster], bg=color.vec[Cluster])
plot(tsne.res$Y, main=&quot;tSNE&quot;, xlab=&quot;Dimension 1&quot;, ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Cluster], bg=color.vec[Cluster])
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>anno.col &lt;- data.frame(Cluster=Cluster)
rownames(anno.col) &lt;- colnames(marker.counts)
head(anno.col)</code></pre>
<pre><code>            Cluster
4th-C63_S30   Cell2
4th-C66_S36   Cell2
4th-C18_S31   Cell2
4th-C57_S18   Cell1
4th-C56_S17   Cell2
4th-C68_S41   Cell2</code></pre>
<pre class="r"><code>cluster.col &lt;- color.vec
names(cluster.col) &lt;- as.vector(levels(Cluster))
anno.colors &lt;- list(Cluster=cluster.col)
anno.colors</code></pre>
<pre><code>$Cluster
    Cell1     Cell2 
&quot;#E41A1C&quot; &quot;#999999&quot; </code></pre>
<pre class="r"><code>pheatmap(log(marker.counts+1), show_colnames =FALSE, 
         clustering_method = &quot;ward.D2&quot;,cutree_cols = 2,annotation_col = anno.col,
         annotation_colors = anno.colors,
         filename=&quot;output/Lawlor_Islets_Alpha_iasva_SV2Markers_rsqcutoff0.6_pheatmap_iasvaV0.95.pdf&quot;,
         width=6, height=5)

pheatmap(log(marker.counts.long+1), show_colnames =FALSE,
         clustering_method = &quot;ward.D2&quot;,cutree_cols = 2,annotation_col = anno.col,
         annotation_colors = anno.colors,
         filename=&quot;output/Lawlor_Islets_Alpha_iasva_SV2Markers_rsqcutoff0.3_pheatmap_iasvaV0.95.pdf&quot;,
         width=6, height=14)</code></pre>
<pre class="r"><code>write.table(as.data.frame(rownames(marker.counts)),
            file=&quot;output/Lawlor_Islets_Alpha_Doublets_SV2_Genes_rsqcutoff0.6.txt&quot;, quote=F,
              row.names=F, col.names=F, sep=&quot; &quot;)

write.table(as.data.frame(rownames(marker.counts.long)),
            file=&quot;output/Lawlor_Islets_Alpha_Doublets_SV2_Genes_rsqcutoff0.3.txt&quot;, quote=F,
              row.names=F, col.names=F, sep=&quot; &quot;)</code></pre>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.0 (2018-04-23)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] SummarizedExperiment_1.10.1 DelayedArray_0.6.1         
 [3] matrixStats_0.53.1          Biobase_2.40.0             
 [5] GenomicRanges_1.32.3        GenomeInfoDb_1.16.0        
 [7] IRanges_2.14.10             S4Vectors_0.18.3           
 [9] BiocGenerics_0.26.0         RColorBrewer_1.1-2         
[11] DescTools_0.99.24           corrplot_0.84              
[13] pheatmap_1.0.10             Rtsne_0.13                 
[15] SCnorm_1.2.1                sva_3.28.0                 
[17] BiocParallel_1.14.2         genefilter_1.62.0          
[19] mgcv_1.8-23                 nlme_3.1-137               
[21] iasvaExamples_1.0.0         iasva_0.99.3               
[23] irlba_2.3.2                 Matrix_1.2-14              
[25] workflowr_1.0.1             rmarkdown_1.9              

loaded via a namespace (and not attached):
 [1] bitops_1.0-6               bit64_0.9-7               
 [3] rprojroot_1.3-2            tools_3.5.0               
 [5] backports_1.1.2            R6_2.2.2                  
 [7] DBI_1.0.0                  lazyeval_0.2.1            
 [9] colorspace_1.3-2           tidyselect_0.2.4          
[11] moments_0.14               bit_1.1-14                
[13] compiler_3.5.0             git2r_0.21.0              
[15] quantreg_5.36              expm_0.999-2              
[17] SparseM_1.77               labeling_0.3              
[19] scales_0.5.0               mvtnorm_1.0-8             
[21] stringr_1.3.1              digest_0.6.15             
[23] foreign_0.8-70             R.utils_2.6.0             
[25] XVector_0.20.0             pkgconfig_2.0.1           
[27] htmltools_0.3.6            manipulate_1.0.1          
[29] limma_3.36.2               rlang_0.2.1               
[31] RSQLite_2.1.1              bindr_0.1.1               
[33] dplyr_0.7.5                R.oo_1.22.0               
[35] RCurl_1.95-4.10            magrittr_1.5              
[37] GenomeInfoDbData_1.1.0     Rcpp_0.12.17              
[39] munsell_0.4.3              R.methodsS3_1.7.1         
[41] stringi_1.2.2              whisker_0.3-2             
[43] yaml_2.1.19                MASS_7.3-50               
[45] zlibbioc_1.26.0            plyr_1.8.4                
[47] grid_3.5.0                 blob_1.1.1                
[49] lattice_0.20-35            splines_3.5.0             
[51] annotate_1.58.0            knitr_1.20                
[53] pillar_1.2.3               boot_1.3-20               
[55] reshape2_1.4.3             XML_3.98-1.11             
[57] glue_1.2.0                 evaluate_0.10.1           
[59] data.table_1.11.4          MatrixModels_0.4-1        
[61] gtable_0.2.0               purrr_0.2.5               
[63] assertthat_0.2.0           ggplot2_2.2.1.9000        
[65] xtable_1.8-2               survival_2.42-3           
[67] SingleCellExperiment_1.2.0 tibble_1.4.2              
[69] AnnotationDbi_1.42.1       memoise_1.1.0             
[71] bindrcpp_0.2.2             cluster_2.0.7-1           </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1
</p>
<hr>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
