<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Donghyung Lee" />

<meta name="date" content="2018-08-03" />

<title>Detecting cell-cycle stage difference in glioblastoma cells</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">iasvaExamples</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/dleelab/iasvaExamples">GitHub</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Detecting cell-cycle stage difference in glioblastoma cells</h1>
<h4 class="author"><em>Donghyung Lee</em></h4>
<h4 class="date"><em>2018-08-03</em></h4>

</div>

<div id="TOC">
<ul>
<li><a href="#install-packages">Install packages</a></li>
<li><a href="#load-packages">Load packages</a></li>
<li><a href="#load-the-glioblastoma-single-cell-rna-seq-data">Load the glioblastoma single cell RNA-Seq data</a></li>
<li><a href="#extract-glioblastoma-cells-from-patient-mgh30">Extract glioblastoma cells from Patient MGH30</a></li>
<li><a href="#normalization-using-scnorm">Normalization using SCnorm</a></li>
<li><a href="#calculate-the-number-of-detected-genes">Calculate the number of detected genes</a></li>
<li><a href="#run-ia-sva">Run IA-SVA</a></li>
<li><a href="#find-marker-genes-for-the-detected-heterogeneity-sv1.">Find marker genes for the detected heterogeneity (SV1).</a></li>
<li><a href="#cell-type-assignment-using-scran-r-package">Cell type assignment using scran R package</a></li>
<li><a href="#run-tsne-post-ia-sva-analyses-i.e.-run-tsne-on-marker-genes-associated-with-sv1-as-detected-by-ia-sva.">Run tSNE post IA-SVA analyses, i.e., run tSNE on marker genes associated with SV1 as detected by IA-SVA.</a></li>
<li><a href="#run-principal-component-analysis-pca-to-detect-the-hidden-heterogeneity-sv1.">Run principal component analysis (PCA) to detect the hidden heterogeneity (SV1).</a></li>
<li><a href="#run-surrogate-variable-analysis-sva-to-detect-the-hidden-heterogeneity-sv1.">Run surrogate variable analysis (SVA) to detect the hidden heterogeneity (SV1).</a></li>
<li><a href="#correlation-between-sv1-and-the-geometric-library-size">Correlation between SV1 and the geometric library size</a></li>
<li><a href="#session-information">Session information</a></li>
</ul>
</div>

<p><strong>Last updated:</strong> 2018-08-03</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(20180731)</code> </summary></p>
<p>The command <code>set.seed(20180731)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/dleelab/iasvaExamples/tree/268983524d3aee232c1bf032e8dd8993bcfb4fc3" target="_blank">2689835</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    inst/.DS_Store
    Ignored:    inst/doc/.DS_Store
    Ignored:    vignettes/.DS_Store

Untracked files:
    Untracked:  analysis/figure/
    Untracked:  output/CC_genes.long.txt
    Untracked:  output/CC_genes.short.txt
    Untracked:  output/Clustering_analyses_figure3_sv1.pdf
    Untracked:  output/Patel_Glioblastoma_MGH30_CellCycle_Figure3ABCD.pdf
    Untracked:  output/Patel_Glioblastoma_MGH30_Cellcycle_SV1_Genes_rsqcutoff0.3.txt
    Untracked:  output/Patel_Glioblastoma_MGH30_Cellcycle_SV1_Genes_rsqcutoff0.4.txt
    Untracked:  output/Patel_Glioblastoma_MGH30_iasva_SV1_genes_rsqcutoff0.3_pheatmap_iasvaV0.95_Figure3F.pdf

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/268983524d3aee232c1bf032e8dd8993bcfb4fc3/analysis/hidden_heterogeneity_glioblastoma.Rmd" target="_blank">2689835</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
error corrected
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/a414c7cd1a6c7fd4f5db3c298a98e124608d4091/analysis/hidden_heterogeneity_glioblastoma.Rmd" target="_blank">a414c7c</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/dleelab/iasvaExamples/c241b04dac899407045051a04b7539237f4d183f/docs/hidden_heterogeneity_glioblastoma.html" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<p>Here, we illustrate how to use the iasva package to detect cell cycle stage difference within single cell RNA sequencing data. We use single cell RNA sequencing (scRNA-Seq) data obtained from human glioblastoma samples (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4123637/">Petel et. al., 2014</a>). This dataset is included in a R data package (“iasvaExamples”) containing data examples for IA-SVA (<a href="https://github.com/dleelab/iasvaExamples" class="uri">https://github.com/dleelab/iasvaExamples</a>). To install the package, follow the instruction provided in the GitHub page.</p>
<div id="install-packages" class="section level2">
<h2>Install packages</h2>
<pre class="r"><code>#devtools
library(devtools)
#iasva
devtools::install_github(&quot;UcarLab/iasva&quot;)
#iasvaExamples  
devtools::install_github(&quot;dleelab/iasvaExamples&quot;)</code></pre>
</div>
<div id="load-packages" class="section level2">
<h2>Load packages</h2>
<pre class="r"><code>rm(list=ls())
library(irlba) # partial SVD, the augmented implicitly restarted Lanczos bidiagonalization algorithm
library(iasva)
library(iasvaExamples)
library(sva)
library(SCnorm)</code></pre>
<pre><code>Warning: package &#39;SCnorm&#39; was built under R version 3.5.1</code></pre>
<pre class="r"><code>library(scran)
library(scater)</code></pre>
<pre><code>Warning: package &#39;scater&#39; was built under R version 3.5.1</code></pre>
<pre class="r"><code>library(Rtsne)
library(pheatmap)
library(corrplot)
library(DescTools) #pcc i.e., Pearson&#39;s contingency coefficient
library(RColorBrewer)
library(SummarizedExperiment)
library(vioplot)
color.vec &lt;- brewer.pal(3, &quot;Set1&quot;)

# Normalization.
normalize &lt;- function(counts) 
{
    normfactor &lt;- colSums(counts)
    return(t(t(counts)/normfactor)*median(normfactor))
}</code></pre>
</div>
<div id="load-the-glioblastoma-single-cell-rna-seq-data" class="section level2">
<h2>Load the glioblastoma single cell RNA-Seq data</h2>
<pre class="r"><code>data(&quot;Patel_Glioblastoma_scRNAseq_Read_Counts&quot;)
data(&quot;Patel_Glioblastoma_scRNAseq_Annotations&quot;)
data(&quot;Patel_Glioblastoma_scRNAseq_ENSG_ID&quot;)
ls()</code></pre>
<pre><code>[1] &quot;color.vec&quot;                              
[2] &quot;normalize&quot;                              
[3] &quot;Patel_Glioblastoma_scRNAseq_Annotations&quot;
[4] &quot;Patel_Glioblastoma_scRNAseq_ENSG_ID&quot;    
[5] &quot;Patel_Glioblastoma_scRNAseq_Read_Counts&quot;</code></pre>
<pre class="r"><code>counts &lt;- Patel_Glioblastoma_scRNAseq_Read_Counts
anns &lt;- Patel_Glioblastoma_scRNAseq_Annotations
ENSG.ID &lt;- Patel_Glioblastoma_scRNAseq_ENSG_ID
dim(anns)</code></pre>
<pre><code>[1] 434   3</code></pre>
<pre class="r"><code>dim(counts)</code></pre>
<pre><code>[1] 25353   434</code></pre>
<pre class="r"><code>length(ENSG.ID)</code></pre>
<pre><code>[1] 25353</code></pre>
<pre class="r"><code>summary(anns)</code></pre>
<pre><code>         run      patient_id     subtype   
 SRR1294493:  1   MGH26:118   None   :120  
 SRR1294494:  1   MGH28: 95   Mes    :103  
 SRR1294496:  1   MGH29: 76   Pro    : 89  
 SRR1294498:  1   MGH30: 74   Cla    : 46  
 SRR1294499:  1   MGH31: 71   Neu    : 24  
 SRR1294500:  1               Pro+Cla: 20  
 (Other)   :428               (Other): 32  </code></pre>
<pre class="r"><code>table(anns$patient_id, anns$subtype)</code></pre>
<pre><code>       
        Cla Cla+Mes Mes Neu Neu+Cla Neu+Mes None Pro Pro+Cla Pro+Neu
  MGH26  10       0   0   1       1       0   19  71      14       2
  MGH28   1       5  56   0       0       7   21   5       0       0
  MGH29   0       0  28  12       0      12   19   4       0       1
  MGH30  33       1   8   1       1       0   16   6       6       2
  MGH31   2       0  11  10       0       0   45   3       0       0</code></pre>
<pre class="r"><code>ContCoef(table(anns$patient_id, anns$subtype))</code></pre>
<pre><code>[1] 0.723431</code></pre>
<p>The annotations describing the glioblastoma samples and experimental settings are stored in “anns” and the read counts information is stored in “counts”.</p>
</div>
<div id="extract-glioblastoma-cells-from-patient-mgh30" class="section level2">
<h2>Extract glioblastoma cells from Patient MGH30</h2>
<p>We use read counts of glioblastoma cells from Patient MGH30 (n = 74).</p>
<pre class="r"><code>counts &lt;- counts[, (anns$patient_id==&quot;MGH30&quot;)] 
anns &lt;- subset(anns, (patient_id==&quot;MGH30&quot;))
dim(counts)</code></pre>
<p>[1] 25353 74</p>
<pre class="r"><code>dim(anns)</code></pre>
<p>[1] 74 3</p>
<pre class="r"><code>anns &lt;- droplevels(anns)
prop.zeros &lt;- sum(counts==0)/length(counts)
prop.zeros</code></pre>
<p>[1] 0.6290769</p>
<pre class="r"><code># filter out genes that are sparsely and lowly expressed
filter = apply(counts, 1, function(x) length(x[x&gt;5])&gt;=3)
counts = counts[filter,]
dim(counts)</code></pre>
<p>[1] 21907 74</p>
<pre class="r"><code>ENSG.ID &lt;- ENSG.ID[filter]
length(ENSG.ID)</code></pre>
<p>[1] 21907</p>
<pre class="r"><code>prop.zeros &lt;- sum(counts==0)/length(counts)
prop.zeros</code></pre>
<p>[1] 0.5731686</p>
<pre class="r"><code>Subtype &lt;- anns$subtype
Patient_ID &lt;- anns$patient_id
mito.genes &lt;- grep(pattern = &quot;^MT-&quot;, x = rownames(x = counts), value = TRUE)
Percent_Mito &lt;- colSums(counts[mito.genes, ])/colSums(counts)</code></pre>
</div>
<div id="normalization-using-scnorm" class="section level2">
<h2>Normalization using SCnorm</h2>
<pre class="r"><code>## count-depth relationship for all genes
Conditions = rep(c(1), each=74)
countDeptEst &lt;- plotCountDepth(Data = counts, Conditions = Conditions,
                               FilterCellProportion = .1, NCores=8)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/normalization-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>DataNorm &lt;- SCnorm(Data = counts, Conditions = Conditions,
                   PrintProgressPlots = FALSE,
                   FilterCellNum = 10,
                   NCores=8)</code></pre>
<pre><code>Setting up parallel computation using 8 cores</code></pre>
<pre><code>Gene filter is applied within each condition.</code></pre>
<pre><code>5275 genes in condition 1 will not be included in the normalization due to 
             the specified filter criteria.</code></pre>
<pre><code>A list of these genes can be accessed in output, 
    see vignette for example.</code></pre>
<pre><code>Finding K for Condition 1</code></pre>
<pre><code>Trying K = 1</code></pre>
<pre><code>Trying K = 2</code></pre>
<pre><code>Trying K = 3</code></pre>
<pre><code>Trying K = 4</code></pre>
<pre><code>Trying K = 5</code></pre>
<pre><code>Trying K = 6</code></pre>
<pre><code>Trying K = 7</code></pre>
<pre><code>Done!</code></pre>
<pre class="r"><code>counts &lt;- SingleCellExperiment::normcounts(DataNorm)
summary(colSums(counts))</code></pre>
<pre><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
 54112972  56272666  56519431  58933473  57764297 116031129 </code></pre>
<pre class="r"><code>dim(counts)</code></pre>
<pre><code>[1] 21907    74</code></pre>
</div>
<div id="calculate-the-number-of-detected-genes" class="section level2">
<h2>Calculate the number of detected genes</h2>
<p>It is well known that the number of detected genes in each cell explains a very large portion of variability in scRNA-Seq data (<a href="http://biorxiv.org/content/early/2015/08/25/025528">Hicks et. al. 2015 BioRxiv</a>, <a href="http://www.nature.com/nbt/journal/v34/n6/full/nbt.3498.html">McDavid et. al. 2016 Nature Biotechnology</a>). Frequently, the first principal component of log-transformed scRNA-Seq read counts is highly correlated with the number of detected genes (e.g., r &gt; 0.9). Here, we calculate the number of detected genes for glioblastoma cells, which will be used as an known factor in the IA-SVA analyses.</p>
<pre class="r"><code>Num_Detected_Genes &lt;- colSums(counts&gt;0)
Geo_Lib &lt;- colSums(log(counts+1))
summary(Geo_Lib)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  48594   55663   60062   60197   62882   99461 </code></pre>
<pre class="r"><code>barplot(Geo_Lib, xlab=&quot;Cell&quot;, las=2, ylab = &quot;Geometric library size&quot;)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/num_detected_genes-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>lcounts &lt;- log(counts + 1)
# PC1 and Geometric library size correlation
pc1 = irlba(lcounts - rowMeans(lcounts), 1)$v[,1] ## partial SVD
cor(Num_Detected_Genes, pc1)</code></pre>
<pre><code>[1] 0.8357149</code></pre>
<pre class="r"><code>cor(Geo_Lib, pc1)</code></pre>
<pre><code>[1] 0.9299861</code></pre>
</div>
<div id="run-ia-sva" class="section level2">
<h2>Run IA-SVA</h2>
<p>Here, we run IA-SVA using Geo_Lib_Size as a known factor and identify five hidden factors. SVs are plotted in a pairwise fashion to uncover which SVs can seperate cells.</p>
<pre class="r"><code>set.seed(3445)
mod &lt;- model.matrix(~Geo_Lib)
summ_exp &lt;- SummarizedExperiment(assays = counts)
iasva.res&lt;- iasva(summ_exp, as.matrix(mod[,-1]),verbose=FALSE, permute=FALSE, num.sv=5)</code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
SV 2 Detected!</code></pre>
<pre><code>
SV 3 Detected!</code></pre>
<pre><code>
SV 4 Detected!</code></pre>
<pre><code>
SV 5 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 5</code></pre>
<pre class="r"><code>iasva.sv &lt;- iasva.res$sv

Cluster &lt;- as.factor(iasva.sv[,1] &lt; 0.1) 
levels(Cluster)=c(&quot;Cluster1&quot;,&quot;Cluster2&quot;)
table(Cluster)</code></pre>
<pre><code>Cluster
Cluster1 Cluster2 
      13       61 </code></pre>
<pre class="r"><code>pairs(iasva.sv[,1:5], main=&quot;IA-SVA&quot;, pch=21, col=color.vec[Cluster],
      bg=color.vec[Cluster], oma=c(4,4,6,14))
legend(&quot;right&quot;, levels(Cluster), fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/run_iasva-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(iasva.sv[,1:2], main=&quot;IA-SVA&quot;, pch=21, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;,
     col=color.vec[Cluster], bg=color.vec[Cluster])</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/run_iasva-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>cor(Num_Detected_Genes, iasva.sv[,1])</code></pre>
<pre><code>[1] 0.3477167</code></pre>
<pre class="r"><code>cor(Geo_Lib, iasva.sv[,1])</code></pre>
<pre><code>[1] 0.4360903</code></pre>
<pre class="r"><code>corrplot(cor(iasva.sv))</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/run_iasva-3.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="find-marker-genes-for-the-detected-heterogeneity-sv1." class="section level2">
<h2>Find marker genes for the detected heterogeneity (SV1).</h2>
<p>Here, using the find_markers() function we find marker genes that are significantly associated with SV1 (multiple testing adjusted p-value &lt; 0.05, default significance cutoff, and R-squared value &gt; 0.3).</p>
<pre class="r"><code># try different R2 thresholds
pdf(paste0(&quot;output/Clustering_analyses_figure3_sv1.pdf&quot;))
r2.results &lt;- study_R2(summ_exp, iasva.sv,selected.svs=1, no.clusters=2)</code></pre>
<pre><code># of markers (): 466</code></pre>
<pre><code>total # of unique markers: 466</code></pre>
<pre><code># of markers (): 361</code></pre>
<pre><code>total # of unique markers: 361</code></pre>
<pre><code># of markers (): 232</code></pre>
<pre><code>total # of unique markers: 232</code></pre>
<pre><code># of markers (): 165</code></pre>
<pre><code>total # of unique markers: 165</code></pre>
<pre><code># of markers (): 119</code></pre>
<pre><code>total # of unique markers: 119</code></pre>
<pre><code># of markers (): 89</code></pre>
<pre><code>total # of unique markers: 89</code></pre>
<pre><code># of markers (): 62</code></pre>
<pre><code>total # of unique markers: 62</code></pre>
<pre><code># of markers (): 47</code></pre>
<pre><code>total # of unique markers: 47</code></pre>
<pre><code># of markers (): 33</code></pre>
<pre><code>total # of unique markers: 33</code></pre>
<pre><code># of markers (): 21</code></pre>
<pre><code>total # of unique markers: 21</code></pre>
<pre><code># of markers (): 14</code></pre>
<pre><code>total # of unique markers: 14</code></pre>
<pre><code># of markers (): 5</code></pre>
<pre><code>total # of unique markers: 5</code></pre>
<pre><code># of markers (): 1</code></pre>
<pre><code>total # of unique markers: 1</code></pre>
<pre><code># of markers (): 0</code></pre>
<pre><code>total # of unique markers: 0</code></pre>
<pre class="r"><code>dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>marker.counts.SV1 &lt;- find_markers(summ_exp, 
                            as.matrix(iasva.sv[,c(1)]), rsq.cutoff = 0.4)</code></pre>
<pre><code># of markers (): 62</code></pre>
<pre><code>total # of unique markers: 62</code></pre>
<pre class="r"><code>marker.counts.SV1.long &lt;- find_markers(summ_exp, 
                              as.matrix(iasva.sv[,c(1)]), rsq.cutoff = 0.3)</code></pre>
<pre><code># of markers (): 119</code></pre>
<pre><code>total # of unique markers: 119</code></pre>
<pre class="r"><code>nrow(marker.counts.SV1) </code></pre>
<pre><code>[1] 62</code></pre>
<pre class="r"><code>nrow(marker.counts.SV1.long)</code></pre>
<pre><code>[1] 119</code></pre>
<pre class="r"><code>anno.col2 &lt;- data.frame(Cluster=Cluster, SV1=iasva.sv[,1])
rownames(anno.col2) &lt;- colnames(marker.counts.SV1)
head(anno.col2)</code></pre>
<pre><code>            Cluster         SV1
SRR1294928 Cluster2  0.03921618
SRR1294930 Cluster2 -0.08861064
SRR1294931 Cluster2 -0.06308121
SRR1294932 Cluster2 -0.07956245
SRR1294934 Cluster2  0.01656383
SRR1294935 Cluster2 -0.04288853</code></pre>
<pre class="r"><code>cluster.col &lt;- color.vec[1:2]
names(cluster.col) &lt;- as.vector(levels(Cluster))
anno.colors &lt;- list(Cluster=cluster.col)
anno.colors</code></pre>
<pre><code>$Cluster
 Cluster1  Cluster2 
&quot;#E41A1C&quot; &quot;#377EB8&quot; </code></pre>
<pre class="r"><code>pheatmap(log(marker.counts.SV1+1), show_colnames =FALSE, 
         clustering_method = &quot;ward.D2&quot;,cutree_cols = 2,annotation_col = anno.col2,
         annotation_colors = anno.colors)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/find_markers_SV1-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pheatmap(log(marker.counts.SV1.long+1), show_colnames =FALSE, 
         clustering_method = &quot;ward.D2&quot;,cutree_cols = 2,annotation_col = anno.col2,
         annotation_colors = anno.colors)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/find_markers_SV1-2.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>gene.list &lt;- rownames(marker.counts.SV1)
write.table(gene.list, file = paste0(&quot;output/CC_genes.short.txt&quot;),
            col.names =F, row.names = F, quote = F)

gene.list &lt;- rownames(marker.counts.SV1.long)
write.table(gene.list, file = paste0(&quot;output/CC_genes.long.txt&quot;),
            col.names =F, row.names = F, quote = F)</code></pre>
<p>Theses marker genes are strongly enriched in cell-cycle related GO terms and KEGG pathways. (See Supplementary Figure 6 in <a href="https://doi.org/10.1101/151217" class="uri">https://doi.org/10.1101/151217</a>)</p>
</div>
<div id="cell-type-assignment-using-scran-r-package" class="section level2">
<h2>Cell type assignment using scran R package</h2>
<pre class="r"><code>ENSG.counts &lt;- counts
rownames(ENSG.counts) &lt;- ENSG.ID
sce &lt;- SingleCellExperiment(list(counts=ENSG.counts))

# load human cell cycle markers
hs.pairs &lt;- readRDS(system.file(&quot;exdata&quot;, 
                                &quot;human_cycle_markers.rds&quot;, package=&quot;scran&quot;))
assigned &lt;- cyclone(sce, pairs=hs.pairs)
head(assigned$scores)</code></pre>
<pre><code>     G1     S   G2M
1 0.432 0.983 0.001
2 0.103 0.926 0.027
3 0.105 0.895 0.009
4 0.032 0.950 0.024
5 0.003 0.692 0.560
6 0.452 0.905 0.000</code></pre>
<pre class="r"><code>table(assigned$phases)</code></pre>
<pre><code>
 G1 G2M   S 
  8  14  52 </code></pre>
<pre class="r"><code>phase &lt;- rep(&quot;S&quot;, ncol(sce))
phase[assigned$scores$G1 &gt; 0.5 &amp; assigned$scores$G2M &lt; 0.5] &lt;- &quot;G1&quot;
phase[assigned$scores$G1 &lt; 0.5 &amp; assigned$scores$G2M &gt; 0.5] &lt;- &quot;G2M&quot;
phase[assigned$scores$G1 &lt; 0.5 &amp; assigned$scores$G2M &lt; 0.5] &lt;- &quot;S&quot;
phase[assigned$scores$G1 &gt; 0.5 &amp; assigned$scores$G2M &gt; 0.5] &lt;- &quot;Unknown&quot;
table(phase)</code></pre>
<pre><code>phase
     G1     G2M       S Unknown 
      7      12      52       3 </code></pre>
<pre class="r"><code>G1 &lt;- iasva.sv[,1][phase==&quot;G1&quot;]
S &lt;- iasva.sv[,1][phase==&quot;S&quot;]
G2M &lt;- iasva.sv[,1][phase==&quot;G2M&quot;]
Unknown &lt;- iasva.sv[,1][phase==&quot;Unknown&quot;]
vioplot(G1, S, G2M, Unknown, names=c(&quot;G1&quot;, &quot;S&quot;, &quot;G2M&quot;, &quot;Unknown&quot;), 
   col=&quot;gold&quot;)
title(xlab=&quot;Cell-cycle stage predictions&quot;, ylab=&quot;IA-SVA factor (SV1)&quot;)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/run_scran-1.png" width="576" style="display: block; margin: auto;" /> ## Run tSNE to detect the hidden heterogeneity. For comparison purposes, we applied tSNE on read counts of all genes to identify the hidden heterogeneity. We used the Rtsne R package with default settings.</p>
<pre class="r"><code>set.seed(43324)
tsne.res &lt;- Rtsne(t(lcounts), dims = 2, perplexity = 20)
plot(tsne.res$Y, main=&quot;tSNE&quot;, xlab=&quot;Dim1&quot;, ylab=&quot;Dim2&quot;, 
     pch=21, col=color.vec[Cluster], bg=color.vec[Cluster], oma=c(4,4,6,12))
legend(&quot;bottomright&quot;, levels(Cluster), border=&quot;white&quot;,
       fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/run_tsne-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>As shown above, tSNE fails to detect the outlier cells that are identified by IA-SVA when all genes are used. Same color coding is used as above.</p>
</div>
<div id="run-tsne-post-ia-sva-analyses-i.e.-run-tsne-on-marker-genes-associated-with-sv1-as-detected-by-ia-sva." class="section level2">
<h2>Run tSNE post IA-SVA analyses, i.e., run tSNE on marker genes associated with SV1 as detected by IA-SVA.</h2>
<p>Here, we apply tSNE on the marker genes for SV1 obtained from IA-SVA</p>
<pre class="r"><code>set.seed(3452)
tsne.res &lt;- Rtsne(unique(t(log(marker.counts.SV1.long+1))),
                  dims = 2, perplexity = 20)

plot(tsne.res$Y, main=&quot;IA-SVA + tSNE&quot;, xlab=&quot;tSNE Dim1&quot;,
     ylab=&quot;tSNE Dim2&quot;, pch=21, col=color.vec[Cluster],
     bg=color.vec[Cluster], oma=c(4,4,6,12))
legend(&quot;topright&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/run_tsne_post_iasva-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="run-principal-component-analysis-pca-to-detect-the-hidden-heterogeneity-sv1." class="section level2">
<h2>Run principal component analysis (PCA) to detect the hidden heterogeneity (SV1).</h2>
<p>Here, we use PCA to detect the cell cycle stage difference (SV1) detected by IA-SVA.</p>
<pre class="r"><code>set.seed(3333)
pca.res = irlba(lcounts - rowMeans(lcounts), 5)$v ## partial SVD

pairs(pca.res[,1:5], main=&quot;PCA&quot;, pch=21, col=color.vec[Cluster],
      bg=color.vec[Cluster],
      oma=c(4,4,6,14))
legend(&quot;right&quot;, levels(Cluster), fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/pca_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(pca.res[,1:2], main=&quot;PCA&quot;, pch=21, xlab=&quot;PC1&quot;, ylab=&quot;PC2&quot;,
     col=color.vec[Cluster], bg=color.vec[Cluster])
legend(&quot;bottomleft&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/pca_plot-2.png" width="672" style="display: block; margin: auto;" /> PCA failed to capture the heterogeneity.</p>
</div>
<div id="run-surrogate-variable-analysis-sva-to-detect-the-hidden-heterogeneity-sv1." class="section level2">
<h2>Run surrogate variable analysis (SVA) to detect the hidden heterogeneity (SV1).</h2>
<p>Here, for comparison purposes we use SVA to detect the hidden heterogeneity in our example data.</p>
<pre class="r"><code>mod1 &lt;- model.matrix(~Geo_Lib)
mod0 &lt;- cbind(mod1[,1])
sva.res = svaseq(counts,mod1,mod0, n.sv=5)$sv</code></pre>
<pre><code>Number of significant surrogate variables is:  5 
Iteration (out of 5 ):1  2  3  4  5  </code></pre>
<pre class="r"><code>pairs(sva.res[,1:5], main=&quot;SVA&quot;, pch=21, col=color.vec[Cluster],
      bg=color.vec[Cluster], oma=c(4,4,6,14))
legend(&quot;right&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/run_sva-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(sva.res[,1:2], main=&quot;SVA&quot;, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;,
     pch=21, col=color.vec[Cluster], bg=color.vec[Cluster])
legend(&quot;topleft&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/run_sva-2.png" width="672" style="display: block; margin: auto;" /> SVA failed to detect the cell cycle stage difference.</p>
</div>
<div id="correlation-between-sv1-and-the-geometric-library-size" class="section level2">
<h2>Correlation between SV1 and the geometric library size</h2>
<pre class="r"><code>cor(Num_Detected_Genes, iasva.sv[,1])</code></pre>
<pre><code>[1] 0.3477167</code></pre>
<pre class="r"><code>cor(Geo_Lib, iasva.sv[,1])</code></pre>
<pre><code>[1] 0.4360903</code></pre>
<p>By allowing correlation between factors, IA-SVA accurately detects the cell cycle stage difference, which is moderately correlated (|r|=0.44) with the geometric library size (the first principal component). Existing methods fail to detect the heterogeneity due to the orthogonality assumption.</p>
<pre class="r"><code>pdf(file=paste0(&quot;output/Patel_Glioblastoma_MGH30_CellCycle_Figure3ABCD.pdf&quot;), width=5, height=8)
layout(matrix(c(1,2,3,4,5,5), nrow=3, ncol=2, byrow=TRUE))
plot(iasva.sv[,1:2], main=&quot;IA-SVA&quot;, pch=21, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;,
     col=color.vec[Cluster], bg=color.vec[Cluster], oma=c(4,4,6,12))
legend(&quot;topright&quot;, levels(Cluster), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)
plot(pca.res[,1:2], main=&quot;PCA&quot;, pch=21, xlab=&quot;PC1&quot;, 
     ylab=&quot;PC2&quot;, col=color.vec[Cluster], bg=color.vec[Cluster])
plot(sva.res[,1:2], main=&quot;USVA&quot;, xlab=&quot;SV1&quot;, ylab=&quot;SV2&quot;,
     pch=21, col=color.vec[Cluster], bg=color.vec[Cluster])
plot(tsne.res$Y, main=&quot;tSNE&quot;, xlab=&quot;Dimension 1&quot;, 
     ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Cluster], bg=color.vec[Cluster])
vioplot(G1, S, G2M, Unknown, names=c(&quot;G1&quot;, &quot;S&quot;, &quot;G2M&quot;, &quot;Unknown&quot;), 
   col=&quot;gold&quot;)
title(xlab=&quot;Cell-cycle stage predictions&quot;, ylab=&quot;IA-SVA factor&quot;)
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>anno.col2 &lt;- data.frame(Cluster=Cluster)
rownames(anno.col2) &lt;- colnames(marker.counts.SV1)
head(anno.col2)</code></pre>
<pre><code>            Cluster
SRR1294928 Cluster2
SRR1294930 Cluster2
SRR1294931 Cluster2
SRR1294932 Cluster2
SRR1294934 Cluster2
SRR1294935 Cluster2</code></pre>
<pre class="r"><code>cluster.col &lt;- color.vec[1:2]
names(cluster.col) &lt;- as.vector(levels(Cluster))
anno.colors &lt;- list(Cluster=cluster.col)
anno.colors</code></pre>
<pre><code>$Cluster
 Cluster1  Cluster2 
&quot;#E41A1C&quot; &quot;#377EB8&quot; </code></pre>
<pre class="r"><code>pheatmap(log(marker.counts.SV1.long+1), show_colnames =FALSE, 
         clustering_method = &quot;ward.D2&quot;,cutree_cols = 2,annotation_col = anno.col2,
         annotation_colors = anno.colors)</code></pre>
<p><img src="figure/hidden_heterogeneity_glioblastoma.Rmd/gen_Figure3F-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pheatmap(log(marker.counts.SV1.long+1), show_colnames =FALSE,
         clustering_method = &quot;ward.D2&quot;,cutree_cols = 2,annotation_col = anno.col2,
         annotation_colors = anno.colors,
         filename=paste0(&quot;output/Patel_Glioblastoma_MGH30_iasva_SV1_genes_rsqcutoff0.3_pheatmap_iasvaV0.95_Figure3F.pdf&quot;),
         width=6, height=16)</code></pre>
<pre class="r"><code>write.table(as.data.frame(rownames(marker.counts.SV1)), 
            file=paste0(&quot;output/Patel_Glioblastoma_MGH30_Cellcycle_SV1_Genes_rsqcutoff0.4.txt&quot;),
            quote=F, row.names=F, col.names=F, sep=&quot; &quot;)

write.table(as.data.frame(rownames(marker.counts.SV1.long)), 
            file=paste0(&quot;output/Patel_Glioblastoma_MGH30_Cellcycle_SV1_Genes_rsqcutoff0.3.txt&quot;), 
            quote=F, row.names=F, col.names=F, sep=&quot; &quot;)</code></pre>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.0 (2018-04-23)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] vioplot_0.2                 sm_2.2-5.4                 
 [3] RColorBrewer_1.1-2          DescTools_0.99.24          
 [5] corrplot_0.84               pheatmap_1.0.10            
 [7] Rtsne_0.13                  scater_1.8.2               
 [9] ggplot2_2.2.1.9000          scran_1.8.2                
[11] SingleCellExperiment_1.2.0  SummarizedExperiment_1.10.1
[13] DelayedArray_0.6.1          matrixStats_0.53.1         
[15] Biobase_2.40.0              GenomicRanges_1.32.3       
[17] GenomeInfoDb_1.16.0         IRanges_2.14.10            
[19] S4Vectors_0.18.3            BiocGenerics_0.26.0        
[21] SCnorm_1.2.1                sva_3.28.0                 
[23] BiocParallel_1.14.2         genefilter_1.62.0          
[25] mgcv_1.8-23                 nlme_3.1-137               
[27] iasvaExamples_1.0.0         iasva_0.99.3               
[29] irlba_2.3.2                 Matrix_1.2-14              
[31] workflowr_1.0.1             rmarkdown_1.9              

loaded via a namespace (and not attached):
 [1] ggbeeswarm_0.6.0         colorspace_1.3-2        
 [3] rjson_0.2.19             dynamicTreeCut_1.63-1   
 [5] rprojroot_1.3-2          XVector_0.20.0          
 [7] MatrixModels_0.4-1       DT_0.4                  
 [9] bit64_0.9-7              manipulate_1.0.1        
[11] mvtnorm_1.0-8            AnnotationDbi_1.42.1    
[13] splines_3.5.0            R.methodsS3_1.7.1       
[15] tximport_1.8.0           knitr_1.20              
[17] annotate_1.58.0          cluster_2.0.7-1         
[19] R.oo_1.22.0              shinydashboard_0.7.0    
[21] shiny_1.1.0              compiler_3.5.0          
[23] backports_1.1.2          assertthat_0.2.0        
[25] lazyeval_0.2.1           limma_3.36.2            
[27] later_0.7.2              htmltools_0.3.6         
[29] quantreg_5.36            tools_3.5.0             
[31] bindrcpp_0.2.2           igraph_1.2.1            
[33] gtable_0.2.0             glue_1.2.0              
[35] GenomeInfoDbData_1.1.0   reshape2_1.4.3          
[37] dplyr_0.7.5              Rcpp_0.12.17            
[39] DelayedMatrixStats_1.2.0 stringr_1.3.1           
[41] mime_0.5                 statmod_1.4.30          
[43] XML_3.98-1.11            edgeR_3.22.3            
[45] MASS_7.3-50              zlibbioc_1.26.0         
[47] scales_0.5.0             promises_1.0.1          
[49] expm_0.999-2             rhdf5_2.24.0            
[51] SparseM_1.77             yaml_2.1.19             
[53] memoise_1.1.0            gridExtra_2.3           
[55] stringi_1.2.2            RSQLite_2.1.1           
[57] boot_1.3-20              rlang_0.2.1             
[59] pkgconfig_2.0.1          moments_0.14            
[61] bitops_1.0-6             evaluate_0.10.1         
[63] lattice_0.20-35          purrr_0.2.5             
[65] Rhdf5lib_1.2.1           bindr_0.1.1             
[67] labeling_0.3             htmlwidgets_1.2         
[69] bit_1.1-14               tidyselect_0.2.4        
[71] plyr_1.8.4               magrittr_1.5            
[73] R6_2.2.2                 DBI_1.0.0               
[75] foreign_0.8-70           pillar_1.2.3            
[77] whisker_0.3-2            withr_2.1.2             
[79] survival_2.42-3          RCurl_1.95-4.10         
[81] tibble_1.4.2             viridis_0.5.1           
[83] locfit_1.5-9.1           grid_3.5.0              
[85] data.table_1.11.4        blob_1.1.1              
[87] FNN_1.1                  git2r_0.21.0            
[89] digest_0.6.15            xtable_1.8-2            
[91] httpuv_1.4.3             R.utils_2.6.0           
[93] munsell_0.4.3            beeswarm_0.2.3          
[95] viridisLite_0.3.0        vipor_0.4.5             </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1
</p>
<hr>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
