<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Donghyung Lee" />

<meta name="date" content="2018-08-27" />

<title>Reviewer 2 Simulation</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">iasvaExamples</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/dleelab/iasvaExamples">GitHub</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Reviewer 2 Simulation</h1>
<h4 class="author"><em>Donghyung Lee</em></h4>
<h4 class="date"><em>2018-08-27</em></h4>

</div>

<div id="TOC">
<ul>
<li><a href="#the-reviewer-2s-simulation-1-with-weak-unknown-factor.">The reviewer 2’s simulation #1 with weak unknown factor.</a></li>
<li><a href="#the-reviewer-2s-simulation-2-with-stronger-effect-size.">The reviewer 2’s simulation #2 with stronger effect size.</a></li>
<li><a href="#several-issues-in-the-simulation-studies-suggested-by-the-reviewer-2">Several issues in the simulation studies suggested by the reviewer #2</a></li>
<li><a href="#corrected-simulation-with-low-correlation-btw-known-and-hidden-factors">Corrected simulation with low correlation btw known and hidden factors</a></li>
<li><a href="#corrected-simulation-with-high-correlation-btw-known-and-hidden-factors-with-stronger-effect-sizes">Corrected Simulation with high correlation btw known and hidden factors with stronger effect sizes</a></li>
<li><a href="#more-challenging-scenario.">More challenging scenario.</a></li>
<li><a href="#session-information">Session information</a></li>
</ul>
</div>

<p><strong>Last updated:</strong> 2018-08-27</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(20180731)</code> </summary></p>
<p>The command <code>set.seed(20180731)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/dleelab/iasvaExamples/tree/492eafb937bf0a4cccd4a4bd2d798bd400d644f7" target="_blank">492eafb</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    docs/.DS_Store
    Ignored:    docs/figure/.DS_Store
    Ignored:    inst/.DS_Store
    Ignored:    inst/doc/.DS_Store
    Ignored:    output/.DS_Store
    Ignored:    vignettes/.DS_Store

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/492eafb937bf0a4cccd4a4bd2d798bd400d644f7/analysis/reviewer2_sim.Rmd" target="_blank">492eafb</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-27
</td>
<td style="text-align:left;">
figure number changed
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/e042edfba8fbe7a06d2b01573c03579ff6d81cd0/analysis/reviewer2_sim.Rmd" target="_blank">e042edf</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-27
</td>
<td style="text-align:left;">
figure updated
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/dleelab/iasvaExamples/5e29c04506c1303bb96adfec18e01215d3c089ff/docs/reviewer2_sim.html" target="_blank">5e29c04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-22
</td>
<td style="text-align:left;">
reviewer2 sim updated.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/17ebb0364f4b76ba44d101be572d065b84338d5f/analysis/reviewer2_sim.Rmd" target="_blank">17ebb03</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-22
</td>
<td style="text-align:left;">
text edit
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/fa4ce46fa17184b7a3e51f999086676f9d5e87c5/analysis/reviewer2_sim.Rmd" target="_blank">fa4ce46</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-21
</td>
<td style="text-align:left;">
figure changed
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/cf73dbbe3747c9cc457d7e9f4ea3c297cabebf04/analysis/reviewer2_sim.Rmd" target="_blank">cf73dbb</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-21
</td>
<td style="text-align:left;">
reviewer2 sim updated
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/c241b04dac899407045051a04b7539237f4d183f/analysis/reviewer2_sim.Rmd" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/dleelab/iasvaExamples/c241b04dac899407045051a04b7539237f4d183f/docs/reviewer2_sim.html" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<div id="the-reviewer-2s-simulation-1-with-weak-unknown-factor." class="section level2">
<h2>The reviewer 2’s simulation #1 with weak unknown factor.</h2>
<pre class="r"><code># Setting up the known and unknown factors. 
set.seed(10000) 
known &lt;- runif(1000) 
unknown &lt;- runif(1000, 0, 0.1) # weak unknown factor 

# Generating Poisson count data from log-normal means. 
ngenes &lt;- 5000 
known.effect &lt;- rnorm(ngenes, sd=2) 
unknown.effect &lt;- rnorm(ngenes, sd=2) 
mat &lt;- outer(known.effect, known) + outer(unknown.effect, unknown) + 2 # to get decent-sized counts 

counts &lt;- matrix(rpois(length(mat), lambda=2^mat), nrow=ngenes) 
library(SummarizedExperiment) 
se &lt;- SummarizedExperiment(list(counts=counts)) 

# Running IA-SVA, version 0.99.3. 
library(iasva) 
design &lt;- model.matrix(~known) 
res &lt;- iasva(se, design, num.sv=5, permute = FALSE) 
# IA-SVA running...
# 
# SV 1 Detected!
# 
# SV 2 Detected!
# 
# SV 3 Detected!
# 
# SV 4 Detected!
# 
# SV 5 Detected!
# 
# # of significant surrogate variables: 5

cor(res$sv[,1], unknown) # these SVs are not the unknown factor (correlation ~= 0) 
# [1] 0.05699341
cor(res$sv[,2], unknown) 
# [1] -0.1317231
cor(res$sv[,3], unknown) 
# [1] 0.06301151
cor(res$sv[,4], unknown) 
# [1] -0.08444932
cor(res$sv[,5], unknown) 
# [1] 0.1239618

cor(res$sv[,1], known) # ... but are instead the known factor (correlation ~= +/-1) 
# [1] -0.9988192
cor(res$sv[,2], known) 
# [1] 0.9971117
cor(res$sv[,3], known) 
# [1] -0.9558875
cor(res$sv[,4], known) 
# [1] 0.9982102
cor(res$sv[,5], known) 
# [1] -0.9934994

## Compare to just naively taking the first PC of the residual matrix. 
library(limma) 
# 
# Attaching package: &#39;limma&#39;
# The following object is masked from &#39;package:BiocGenerics&#39;:
# 
#     plotMA
resid &lt;- removeBatchEffect(log(assay(se)+1), covariates=known) 
pr.out &lt;- prcomp(t(resid), rank.=1) 
abs(cor(pr.out$x[,1], unknown)) # close to 1 
# [1] 0.9766822
abs(cor(pr.out$x[,1], known)) # close to zero. 
# [1] 2.35573e-17

## Compare to SVA
library(sva)
mod1 &lt;- model.matrix(~known)
mod0 &lt;- cbind(mod1[,1])
sva.res = svaseq(counts,mod1,mod0, n.sv=5)$sv
# Number of significant surrogate variables is:  5 
# Iteration (out of 5 ):1  2  3  4  5
abs(cor(sva.res[,1], unknown)) # 0.55
# [1] 0.5522043
abs(cor(sva.res[,1], known)) # close to zero.
# [1] 0.04603016

hf.mat1 &lt;- cbind(known, unknown, res$sv[,1], pr.out$x[,1], sva.res[,1])
colnames(hf.mat1) &lt;- c(&quot;KF&quot;,&quot;HF&quot;,&quot;IASVA&quot;,&quot;PC1_Residual&quot;,&quot;SVA&quot;)
corrplot.mixed(abs(cor(hf.mat1)), tl.pos=&quot;lt&quot;)</code></pre>
<p><img src="figure/reviewer2_sim.Rmd/sim1-1.png" width="576" style="display: block; margin: auto;" /></p>
<details> <summary><em>Expand here to see past versions of sim1-1.png:</em></summary>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/5e29c04506c1303bb96adfec18e01215d3c089ff/docs/figure/reviewer2_sim.Rmd/sim1-1.png" target="_blank">5e29c04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-22
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/c241b04dac899407045051a04b7539237f4d183f/docs/figure/reviewer2_sim.Rmd/sim1-1.png" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
</tr>
</tbody>
</table>
<p></details></p>
</div>
<div id="the-reviewer-2s-simulation-2-with-stronger-effect-size." class="section level2">
<h2>The reviewer 2’s simulation #2 with stronger effect size.</h2>
<pre class="r"><code>set.seed(10000) 

known &lt;- runif(1000) 
unknown &lt;- runif(1000) 

ngenes &lt;- 5000 
known.effect &lt;- rnorm(ngenes, sd=2) 
unknown.effect &lt;- rnorm(ngenes, sd=2) 
mat &lt;- outer(known.effect, known) + outer(unknown.effect, unknown) + 2 # to get decent-sized counts 

counts &lt;- matrix(rpois(length(mat), lambda=2^mat), nrow=ngenes) 
library(SummarizedExperiment) 
se &lt;- SummarizedExperiment(list(counts=counts)) 

library(iasva) 
design &lt;- model.matrix(~known) 
res &lt;- iasva(se, design, num.sv=5, permute = FALSE) 
# IA-SVA running...
# 
# SV 1 Detected!
# 
# SV 2 Detected!
# 
# SV 3 Detected!
# 
# SV 4 Detected!
# 
# SV 5 Detected!
# 
# # of significant surrogate variables: 5

cor(res$sv[,1], unknown) # first SV is the unknown factor. 
# [1] -0.9960557
cor(res$sv[,2], unknown) 
# [1] 0.7637275
cor(res$sv[,3], unknown) 
# [1] 0.7564822
cor(res$sv[,4], unknown) 
# [1] -0.7629373
cor(res$sv[,5], unknown) 
# [1] 0.4760279

cor(res$sv[,1], known) 
# [1] -0.004959888
cor(res$sv[,2], known) # but SVs 2-5 are correlated with the known factor... 
# [1] -0.7014209
cor(res$sv[,3], known) 
# [1] -0.7076734
cor(res$sv[,4], known) 
# [1] -0.476617
cor(res$sv[,5], known) 
# [1] -0.8898334

## Compare to just naively taking the first PC of the residual matrix. 
library(limma) 
resid &lt;- removeBatchEffect(log(assay(se)+1), covariates=known) 
pr.out &lt;- prcomp(t(resid), rank.=1) 
abs(cor(pr.out$x[,1], unknown)) # close to 1 
# [1] 0.9963041
abs(cor(pr.out$x[,1], known)) # close to zero.
# [1] 1.691586e-16

# Compare to SVA
library(sva)
mod1 &lt;- model.matrix(~known)
mod0 &lt;- cbind(mod1[,1])
sva.res = svaseq(counts,mod1,mod0, n.sv=1)$sv
# Number of significant surrogate variables is:  1 
# Iteration (out of 5 ):1  2  3  4  5
abs(cor(sva.res[,1], unknown)) # close to 1 
# [1] 0.9900321
abs(cor(sva.res[,1], known)) # close to zero.
# [1] 0.0007644251

hf.mat2 &lt;- cbind(known, unknown, res$sv[,1], pr.out$x[,1], sva.res[,1])
colnames(hf.mat2) &lt;- c(&quot;KF&quot;,&quot;HF&quot;,&quot;IASVA&quot;,&quot;PC1_Residual&quot;,&quot;SVA&quot;)
corrplot.mixed(abs(cor(hf.mat2)), tl.pos=&quot;lt&quot;)</code></pre>
<p><img src="figure/reviewer2_sim.Rmd/sim2-1.png" width="576" style="display: block; margin: auto;" /></p>
<details> <summary><em>Expand here to see past versions of sim2-1.png:</em></summary>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/5e29c04506c1303bb96adfec18e01215d3c089ff/docs/figure/reviewer2_sim.Rmd/sim2-1.png" target="_blank">5e29c04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-22
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/c241b04dac899407045051a04b7539237f4d183f/docs/figure/reviewer2_sim.Rmd/sim2-1.png" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
</tr>
</tbody>
</table>
<p></details></p>
</div>
<div id="several-issues-in-the-simulation-studies-suggested-by-the-reviewer-2" class="section level2">
<h2>Several issues in the simulation studies suggested by the reviewer #2</h2>
<p>We noticed several issues in these simulations. First, SVA or factor analysis based methods including IA-SVA are designed to estimate hidden variables with distinct levels (i.e., factors) or a transformation of factors. However, instead of simulating factors (e.g., a variable with two levels (1 and -1)) as typically done in such analyses1,2 (e.g., see <a href="http://jtleek.com/svaseq/simulateData.html" class="uri">http://jtleek.com/svaseq/simulateData.html</a> and Method section (Simulated data) of <a href="https://www.biorxiv.org/content/biorxiv/early/2017/10/27/200345.full.pdf" class="uri">https://www.biorxiv.org/content/biorxiv/early/2017/10/27/200345.full.pdf</a>), the reviewer simulated numeric vectors from a uniform distribution (e.g., runif(1000)) and used them as hidden variables. We note that this leads, like IA-SVA, SVA also to perform poorly in the simulation study (r = -0.55) (see <a href="https://dleelab.github.io/iasvaExamples/reviewer2_sim.html" class="uri">https://dleelab.github.io/iasvaExamples/reviewer2_sim.html</a>). Second, IA-SVA does not accept intercept column in the design matrix; therefore instead of design, design[,-1] should have been used in these simulations. Third, the reviewer simulated counts using a Poisson distribution (mean = variance); however, negative binomial or zero-inflated negative binomial is a more appropriate model to simulate high variability (mean &lt;&lt; variance) in bulk or single cell RNA-seq data. Fourth, even though only one hidden factor is simulated in the study, IA-SVA was forced to generate five hidden factors by setting the option ‘num.sv’ = 5. This leads IA-SVA to generate SV2-SV5, which are highly correlated SV1. To avoid this, we recommend users using the permutation procedure (by setting the default setting for option ‘permute’ (i.e, permute = TRUE) in ‘iasva’ function) or ‘fast_iasva’ function to determine the number of meaningful hidden factors. Lastly, to simulate a hidden factor with stronger effect sizes, the standard deviation (sd) in “unknown.effect &lt;- rnorm(ngenes, sd=2)” should be increased (e.g., sd=4); rather increasing the range of simulated factor values from “unknown &lt;- runif(1000, 0, 0.1)” to “unknown &lt;- runif(1000)”. We have repeated the simulation studies by taking into consideration these points below.</p>
</div>
<div id="corrected-simulation-with-low-correlation-btw-known-and-hidden-factors" class="section level2">
<h2>Corrected simulation with low correlation btw known and hidden factors</h2>
<pre class="r"><code># Setting up the known and unknown factors. 
set.seed(10000) 

## to reduce the computational burden, we reduced the sample size (# of cells) and # of genes.
sample.size &lt;- 100 
ngenes &lt;- 2000     

# CORRECTEION: Here, we simulate factors with two levels (-1 or 1).
factor.prop &lt;- 0.5
flip.prob &lt;- 0.5 ## to make no correlation btw known and hidden factors
known &lt;- c(rep(-1,each=sample.size*factor.prop),rep(1,each=sample.size-(sample.size*factor.prop)))
coinflip = rbinom(sample.size,size=1,prob=flip.prob)
unknown = known*coinflip + -known*(1-coinflip)
cor(known, unknown)
# [1] -0.06030227

# Generating Poisson count data from log-normal means. 
# CORRECTION: to make factor with weak effect size, we use 1 as SD here.
known.effect &lt;- rnorm(ngenes, sd=1) 
unknown.effect &lt;- rnorm(ngenes, sd=1)

# To make estimation of the hidden factor more challenging, here we ramdomly select 1500 (75%) and 1900 (95%) genes from each effect size vector and set their effect sizes as 0. That is, only 25% and 5% of genes are affected by the known and hidden factors, respectively. 
known.effect[sample(1:ngenes, 1500, replace=FALSE)] &lt;- 0
unknown.effect[sample(1:ngenes, 1900, replace=FALSE)] &lt;- 0

mat &lt;- outer(known.effect, known) + outer(unknown.effect, unknown) + 2 # to get decent-sized counts 
counts &lt;- matrix(rpois(length(mat), lambda=2^mat), nrow=ngenes)
se &lt;- SummarizedExperiment(assays=counts) 

# Running IA-SVA, version 0.99.3. 
library(iasva) 
design &lt;- model.matrix(~known) 

# CORRECTION: IASVA doesn&#39;t accept the constant term of design matrix, so we use design[,-1] instead of design
res &lt;- iasva(se, as.matrix(design[,-1]), threads=8) # iasva detected only one hidden factor.
# IA-SVA running...
# 
# SV 1 Detected!
# 
# # of significant surrogate variables: 1
# fast_iasva used here.
res.fast &lt;- fast_iasva(se, as.matrix(design[,-1]))  # fast_iasva detected only one hidden factor.
# fast IA-SVA running...
# 
# SV 1 Detected!
# 
# # of obtained surrogate variables: 1

#plot(res$sv[,1], unknown) 
abs(cor(res$sv[,1], unknown)) ## absolute value of cor(SV1, unknown factor) = 0.9995
# [1] 0.9972323
abs(cor(res$sv[,1], known))   ## absolute value of cor(SV1, known factor) close to zero.
# [1] 0.0707345

#plot(res.fast$sv[,1], unknown) 
abs(cor(res.fast$sv[,1], unknown)) ## absolute value of cor(SV1, unknown factor) = 0.9995
# [1] 0.9972323
abs(cor(res.fast$sv[,1], known))   ## absolute value of cor(SV1, known factor) close to zero.
# [1] 0.0707345


# Compare to just naively taking the first PC of the residual matrix. 
library(limma) 
resid &lt;- removeBatchEffect(log(assay(se)+1), covariates=known) 
pr.out &lt;- prcomp(t(resid), rank.=1) 
#plot(pr.out$x[,1], unknown) 
abs(cor(pr.out$x[,1], unknown)) # close to 1 
# [1] 0.9951611
abs(cor(pr.out$x[,1], known)) # close to zero. 
# [1] 4.020405e-16


# Compare to SVA
library(sva)
mod1 &lt;- model.matrix(~known)
mod0 &lt;- cbind(mod1[,1])
sva.res = svaseq(counts,mod1,mod0)$sv
# Number of significant surrogate variables is:  1 
# Iteration (out of 5 ):1  2  3  4  5
#plot(sva.res[,1], unknown)
abs(cor(sva.res[,1], unknown)) # close to 1 
# [1] 0.9967295
abs(cor(sva.res[,1], known)) # close to zero. 
# [1] 0.05326845

hf.mat3 &lt;- cbind(known, unknown, res$sv[,1], pr.out$x[,1], sva.res[,1])
colnames(hf.mat3) &lt;- c(&quot;KF&quot;,&quot;HF&quot;,&quot;IASVA&quot;,&quot;PC1_Residual&quot;,&quot;SVA&quot;)
corrplot.mixed(abs(cor(hf.mat3)), tl.pos=&quot;lt&quot;)</code></pre>
<p><img src="figure/reviewer2_sim.Rmd/sim1_corrected-1.png" width="576" style="display: block; margin: auto;" /></p>
<details> <summary><em>Expand here to see past versions of sim1_corrected-1.png:</em></summary>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/5e29c04506c1303bb96adfec18e01215d3c089ff/docs/figure/reviewer2_sim.Rmd/sim1_corrected-1.png" target="_blank">5e29c04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-22
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/c241b04dac899407045051a04b7539237f4d183f/docs/figure/reviewer2_sim.Rmd/sim1_corrected-1.png" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
</tr>
</tbody>
</table>
<p></details></p>
</div>
<div id="corrected-simulation-with-high-correlation-btw-known-and-hidden-factors-with-stronger-effect-sizes" class="section level2">
<h2>Corrected Simulation with high correlation btw known and hidden factors with stronger effect sizes</h2>
<p>Here, we simulated known and hidden factors are strongly correlated (r=0.7) and affect 25% and 5% of genes respectively. We also simulated stronger effect sizes by increasing sd from 1 to 2 when simulating effect sizes from normal distribution,“unknown.effect &lt;- rnorm(ngenes, sd=2)”.</p>
<pre class="r"><code># Setting up the known and unknown factors. 
set.seed(10000) 

# to reduce the computational burden, we reduced the sample size (# of cells) and # of genes.
sample.size &lt;- 100 
ngenes &lt;- 2000     

# CORRECTEION: Here, we simulate factors with two levels (-1 or 1).
factor.prop &lt;- 0.5
flip.prob &lt;- 0.8 # to simulate high correlation btw known and hidden factors
known &lt;- c(rep(-1,each=sample.size*factor.prop),rep(1,each=sample.size-(sample.size*factor.prop)))
coinflip = rbinom(sample.size,size=1,prob=flip.prob)
unknown = known*coinflip + -known*(1-coinflip)
cor(known, unknown) # correlation between known and unknown factors = 0.7
# [1] 0.6940221
 
# Generating Poisson count data from log-normal means. 
# CORRECTION: to make factor with stronger effect size, we use 2 as SD here.
known.effect &lt;- rnorm(ngenes, sd=2) 
unknown.effect &lt;- rnorm(ngenes, sd=2)

# To make estimation of the hidden factor more challenging, here we ramdomly select 1500 (75%) and 1900 (95%) genes from each effect size vector and set their effect sizes as 0. That is, only 25% and 5% of genes are affected by the known and hidden factors, respectively. 
known.effect[sample(1:ngenes, 1500, replace=FALSE)] &lt;- 0
unknown.effect[sample(1:ngenes, 1900, replace=FALSE)] &lt;- 0

mat &lt;- outer(known.effect, known) + outer(unknown.effect, unknown) + 2 # to get decent-sized counts 
counts &lt;- matrix(rpois(length(mat), lambda=2^mat), nrow=ngenes)
se &lt;- SummarizedExperiment(assays=counts) 

# Running IA-SVA, version 0.99.3. 
library(iasva) 
design &lt;- model.matrix(~known) 

# CORRECTION: IASVA doesn&#39;t accept the constant term of design matrix, so we use design[,-1] instead of design
res &lt;- iasva(se, as.matrix(design[,-1]), threads=8) # iasva detected only one hidden factor.
# IA-SVA running...
# 
# SV 1 Detected!
# 
# # of significant surrogate variables: 1
# fast_iasva used here.
res.fast &lt;- fast_iasva(se, as.matrix(design[,-1]))  # fast_iasva detected only one hidden factor.
# fast IA-SVA running...
# 
# SV 1 Detected!
# 
# # of obtained surrogate variables: 1

#plot(res$sv[,1], unknown) 
abs(cor(res$sv[,1], unknown)) ## absolute value of cor(SV1, unknown factor) = 0.9995
# [1] 0.999519
abs(cor(res$sv[,1], known))   ## absolute value of cor(SV1, known factor) close to 0.7.
# [1] 0.7031551

#plot(res.fast$sv[,1], unknown) 
abs(cor(res.fast$sv[,1], unknown)) ## absolute value of cor(SV1, unknown factor) = 0.9995
# [1] 0.999519
abs(cor(res.fast$sv[,1], known))   ## absolute value of cor(SV1, known factor) close to 0.7.
# [1] 0.7031551


# Compare to just naively taking the first PC of the residual matrix.
library(limma) 
resid &lt;- removeBatchEffect(log(assay(se)+1), covariates=known) 
pr.out &lt;- prcomp(t(resid), rank.=1) 
#plot(pr.out$x[,1], unknown) 
abs(cor(pr.out$x[,1], unknown)) # cor = 0.7 
# [1] 0.7191922
abs(cor(pr.out$x[,1], known)) # close to zero. 
# [1] 3.721348e-16


# Compare to SVA
library(sva)
mod1 &lt;- model.matrix(~known)
mod0 &lt;- cbind(mod1[,1])
sva.res = svaseq(counts,mod1,mod0)$sv
# Number of significant surrogate variables is:  1 
# Iteration (out of 5 ):1  2  3  4  5
#plot(sva.res[,1], unknown)
abs(cor(sva.res[,1], unknown)) # cor = 0.7 
# [1] 0.7306284
abs(cor(sva.res[,1], known)) # close to zero.
# [1] 0.07688679

hf.mat4 &lt;- cbind(known, unknown, res$sv[,1], pr.out$x[,1], sva.res[,1])
colnames(hf.mat4) &lt;- c(&quot;KF&quot;,&quot;HF&quot;,&quot;IASVA&quot;,&quot;PC1_Residual&quot;,&quot;SVA&quot;)
corrplot.mixed(abs(cor(hf.mat4)), tl.pos=&quot;lt&quot;)</code></pre>
<p><img src="figure/reviewer2_sim.Rmd/sim_corrected_high_corr-1.png" width="576" style="display: block; margin: auto;" /></p>
<details> <summary><em>Expand here to see past versions of sim_corrected_high_corr-1.png:</em></summary>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/5e29c04506c1303bb96adfec18e01215d3c089ff/docs/figure/reviewer2_sim.Rmd/sim_corrected_high_corr-1.png" target="_blank">5e29c04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-22
</td>
</tr>
</tbody>
</table>
<p></details></p>
</div>
<div id="more-challenging-scenario." class="section level2">
<h2>More challenging scenario.</h2>
<p>Here, we simulated the known and hidden factors to affect only 100 (5%) and 20 (1%) genes respectively. We simulated the 20 genes affected by the hidden factor to be also affected by the known factor. Similarly, the two factors are highly correlated (r=0.7).</p>
<pre class="r"><code># Setting up the known and unknown factors. 
set.seed(10000) 

# to reduce the computational burden, we reduced the sample size (# of cells) and # of genes.
sample.size &lt;- 100 
ngenes &lt;- 2000     

# CORRECTEION: Here, we simulate factors with two levels (-1 or 1).
factor.prop &lt;- 0.5
flip.prob &lt;- 0.8 # to simulate high correlation btw known and hidden factors
known &lt;- c(rep(-1,each=sample.size*factor.prop),rep(1,each=sample.size-(sample.size*factor.prop)))
coinflip = rbinom(sample.size,size=1,prob=flip.prob)
unknown = known*coinflip + -known*(1-coinflip)
cor(known, unknown) # correlation between known and unknown factors = 0.7
# [1] 0.6940221
 
# Generating Poisson count data from log-normal means. 
# CORRECTION: to make factor with stronger effect size, we use 2 as SD here.
known.effect &lt;- rnorm(ngenes, sd=2) 
unknown.effect &lt;- rnorm(ngenes, sd=2)

# To make estimation of the hidden factor more challenging, here we select 1900 (95%) and 1980 (1%) genes from each effect size vector and set their effect sizes as 0. That is, 5% and 1% of genes are affected by the known and hidden factors, respectively. All 20 genes affected by the hidden factor are affected by the known factor. 

known.effect[1:1900] &lt;- 0
unknown.effect[1:1980] &lt;- 0

mat &lt;- outer(known.effect, known) + outer(unknown.effect, unknown) + 2 # to get decent-sized counts 
counts &lt;- matrix(rpois(length(mat), lambda=2^mat), nrow=ngenes)
se &lt;- SummarizedExperiment(assays=counts) 

# Running IA-SVA, version 0.99.3. 
library(iasva) 
design &lt;- model.matrix(~known) 

# CORRECTION: IASVA doesn&#39;t accept the constant term of design matrix, so we use design[,-1] instead of design
res &lt;- iasva(se, as.matrix(design[,-1]), threads=8) # iasva detected only one hidden factor.
# IA-SVA running...
# 
# SV 1 Detected!
# 
# # of significant surrogate variables: 1
# fast_iasva used here.
res.fast &lt;- fast_iasva(se, as.matrix(design[,-1]), pct.cutoff = 2)  # fast_iasva detected only one hidden factor.
# fast IA-SVA running...
# 
# SV 1 Detected!
# 
# # of obtained surrogate variables: 1

#plot(res$sv[,1], unknown) 
abs(cor(res$sv[,1], unknown)) ## absolute value of cor(SV1, unknown factor) = 0.9995
# [1] 0.9831191
abs(cor(res$sv[,1], known))   ## absolute value of cor(SV1, known factor) close to 0.7.
# [1] 0.8073574

#plot(res.fast$sv[,1], unknown) 
abs(cor(res.fast$sv[,1], unknown)) ## absolute value of cor(SV1, unknown factor) = 0.9995
# [1] 0.9831191
abs(cor(res.fast$sv[,1], known))   ## absolute value of cor(SV1, known factor) close to 0.7.
# [1] 0.8073574


# Compare to just naively taking the first PC of the residual matrix.
library(limma) 
resid &lt;- removeBatchEffect(log(assay(se)+1), covariates=known) 
pr.out &lt;- prcomp(t(resid), rank.=1) 
#plot(pr.out$x[,1], unknown) 
abs(cor(pr.out$x[,1], unknown)) # cor = 0.7 
# [1] 0.7114086
abs(cor(pr.out$x[,1], known)) # close to zero. 
# [1] 3.193674e-16


# Compare to SVA
library(sva)
mod1 &lt;- model.matrix(~known)
mod0 &lt;- cbind(mod1[,1])
sva.res = svaseq(counts,mod1,mod0)$sv
# Number of significant surrogate variables is:  1 
# Iteration (out of 5 ):1  2  3  4  5
#plot(sva.res[,1], unknown)
abs(cor(sva.res[,1], unknown)) # cor = 0.7 
# [1] 0.6006951
abs(cor(sva.res[,1], known)) # close to zero.
# [1] 0.09633978

hf.mat5 &lt;- cbind(known, unknown, res$sv[,1], pr.out$x[,1], sva.res[,1])
colnames(hf.mat5) &lt;- c(&quot;KF&quot;,&quot;HF&quot;,&quot;IASVA&quot;,&quot;PC1_Residual&quot;,&quot;SVA&quot;)
corrplot.mixed(abs(cor(hf.mat5)), tl.pos=&quot;lt&quot;)</code></pre>
<p><img src="figure/reviewer2_sim.Rmd/sim_corrected_high_corr_challenging-1.png" width="576" style="display: block; margin: auto;" /></p>
<details> <summary><em>Expand here to see past versions of sim_corrected_high_corr_challenging-1.png:</em></summary>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/5e29c04506c1303bb96adfec18e01215d3c089ff/docs/figure/reviewer2_sim.Rmd/sim_corrected_high_corr_challenging-1.png" target="_blank">5e29c04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-22
</td>
</tr>
</tbody>
</table>
<p></details></p>
<pre class="r"><code>#par(oma = c(0, 2, 2, 2))
pdf(&quot;output/FigureL2.pdf&quot;, width=6,height=9)
par(mfrow=c(3,2))
corrplot.mixed(abs(cor(hf.mat1)), tl.pos=&quot;lt&quot;)
corrplot.mixed(abs(cor(hf.mat2)), tl.pos=&quot;lt&quot;)
corrplot.mixed(abs(cor(hf.mat3)), tl.pos=&quot;lt&quot;)
corrplot.mixed(abs(cor(hf.mat4)), tl.pos=&quot;lt&quot;)
corrplot.mixed(abs(cor(hf.mat5)), tl.pos=&quot;lt&quot;)
dev.off()
# quartz_off_screen 
#                 2</code></pre>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()
# R version 3.5.0 (2018-04-23)
# Platform: x86_64-apple-darwin15.6.0 (64-bit)
# Running under: macOS Sierra 10.12.6
# 
# Matrix products: default
# BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
# LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# attached base packages:
# [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
# [8] methods   base     
# 
# other attached packages:
#  [1] limma_3.36.2                corrplot_0.84              
#  [3] SummarizedExperiment_1.10.1 DelayedArray_0.6.1         
#  [5] matrixStats_0.53.1          Biobase_2.40.0             
#  [7] GenomicRanges_1.32.3        GenomeInfoDb_1.16.0        
#  [9] IRanges_2.14.10             S4Vectors_0.18.3           
# [11] BiocGenerics_0.26.0         sva_3.28.0                 
# [13] BiocParallel_1.14.2         genefilter_1.62.0          
# [15] mgcv_1.8-23                 nlme_3.1-137               
# [17] iasva_0.99.3                workflowr_1.0.1            
# [19] rmarkdown_1.9              
# 
# loaded via a namespace (and not attached):
#  [1] splines_3.5.0          lattice_0.20-35        htmltools_0.3.6       
#  [4] yaml_2.1.19            blob_1.1.1             XML_3.98-1.11         
#  [7] survival_2.42-3        R.oo_1.22.0            DBI_1.0.0             
# [10] R.utils_2.6.0          bit64_0.9-7            GenomeInfoDbData_1.1.0
# [13] stringr_1.3.1          zlibbioc_1.26.0        R.methodsS3_1.7.1     
# [16] evaluate_0.10.1        memoise_1.1.0          knitr_1.20            
# [19] irlba_2.3.2            AnnotationDbi_1.42.1   Rcpp_0.12.17          
# [22] xtable_1.8-2           backports_1.1.2        annotate_1.58.0       
# [25] XVector_0.20.0         bit_1.1-14             digest_0.6.15         
# [28] stringi_1.2.2          grid_3.5.0             rprojroot_1.3-2       
# [31] tools_3.5.0            bitops_1.0-6           magrittr_1.5          
# [34] RCurl_1.95-4.10        RSQLite_2.1.1          cluster_2.0.7-1       
# [37] whisker_0.3-2          Matrix_1.2-14          git2r_0.21.0          
# [40] compiler_3.5.0</code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1
</p>
<hr>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
