<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Donghyung Lee" />

<meta name="date" content="2018-08-01" />

<title>Accuracy of SVs detected when hidden and known factors affect the overlapping sets of genes</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">iasvaExamples</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/dleelab/iasvaExamples">GitHub</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Accuracy of SVs detected when hidden and known factors affect the overlapping sets of genes</h1>
<h4 class="author"><em>Donghyung Lee</em></h4>
<h4 class="date"><em>2018-08-01</em></h4>

</div>

<div id="TOC">
<ul>
<li><a href="#install-packages">Install packages</a></li>
<li><a href="#load-packages">Load packages</a></li>
<li><a href="#load-the-islet-single-cell-rna-seq-data">Load the islet single cell RNA-Seq data</a></li>
<li><a href="#look-at-mean-variance-relationship">Look at mean variance relationship</a></li>
<li><a href="#estimate-zero-inflated-negative-binomial-parameters-from-the-islet-data">Estimate zero inflated negative binomial parameters from the islet data</a></li>
<li><a href="#generate-a-data-set-affected-by-two-factors-known-and-unknown.">Generate a data set affected by two factors (known and unknown).</a></li>
<li><a href="#estimate-hidden-factor-using-ia-sva">Estimate hidden factor using IA-SVA</a></li>
<li><a href="#r-squared-weights-as-a-function-of-hidden-and-known-factors-effect-size">R-squared (Weights) as a function of hidden and known factors’ effect size</a></li>
<li><a href="#hidden-factor-estimation-using-existing-methods">Hidden factor estimation using existing methods</a></li>
<li><a href="#plot-hidden-factor-estimates-of-ia-sva-and-existing-methods">Plot hidden factor estimates of IA-SVA and existing methods</a></li>
<li><a href="#multiple-simulations-where-hidden-and-known-factors-affect-overlapping-sets-of-genes">Multiple simulations where hidden and known factors affect overlapping sets of genes</a></li>
<li><a href="#estimation-of-hidden-factors-using-ia-sva-and-existing-methods">Estimation of hidden factors using IA-SVA and existing methods</a></li>
<li><a href="#compare-the-accuracy">Compare the accuracy</a></li>
<li><a href="#session-information">Session information</a></li>
</ul>
</div>

<p><strong>Last updated:</strong> 2018-08-01</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(20180731)</code> </summary></p>
<p>The command <code>set.seed(20180731)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/dleelab/iasvaExamples/tree/5a96a443f9928e6e1ce51e2783b44377f84c39d4" target="_blank">5a96a44</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    inst/.DS_Store
    Ignored:    inst/doc/.DS_Store
    Ignored:    vignettes/.DS_Store

Untracked files:
    Untracked:  analysis/Brain_scRNASeq_neuron_vs_oligodendrocyte_single_run.Rmd
    Untracked:  analysis/detecting_hidden_heterogeneity.Rmd
    Untracked:  analysis/hidden_heterogeneity_glioblastoma.Rmd
    Untracked:  analysis/reviewer2_sim.Rmd
    Untracked:  analysis/scRNASeq_simulation.Rmd
    Untracked:  analysis/tSNE_post_IA-SVA_3celltypes.Rmd
    Untracked:  analysis/tSNE_post_IA-SVA_Xin_Islets.Rmd
    Untracked:  docs/figure/
    Untracked:  output/GeneOverlap.Fig1.pdf
    Untracked:  output/GeneOverlap.Fig2.pdf
    Untracked:  output/absCor_gene_overlap_pct.pdf
    Untracked:  output/gene_overlap.pdf

Unstaged changes:
    Modified:   analysis/_site.yml
    Deleted:    analysis/about.Rmd
    Modified:   analysis/index.Rmd
    Modified:   analysis/license.Rmd

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/5a96a443f9928e6e1ce51e2783b44377f84c39d4/analysis/sim_study_KF_HF_genes_overlap.Rmd" target="_blank">5a96a44</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-01
</td>
<td style="text-align:left;">
changed figure sizes and deleted unused codes
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/3f6039eb3e8e9d0fdea1dff65204e56106659b44/analysis/sim_study_KF_HF_genes_overlap.Rmd" target="_blank">3f6039e</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-01
</td>
<td style="text-align:left;">
figure updated
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/6c423dfd1430e583e2ba50660adcccd5d836881b/analysis/sim_study_KF_HF_genes_overlap.Rmd" target="_blank">6c423df</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-01
</td>
<td style="text-align:left;">
sim_KF_HF updated
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/bd4870c4b4d1f721bd13f2678228a11cbb3498c9/analysis/sim_study_KF_HF_genes_overlap.Rmd" target="_blank">bd4870c</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-01
</td>
<td style="text-align:left;">
sim_KF_HF added
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<div id="install-packages" class="section level2">
<h2>Install packages</h2>
<pre class="r"><code>#devtools
library(devtools)
#iasva
devtools::install_github(&quot;UcarLab/IA-SVA&quot;)
#iasvaExamples  
devtools::install_github(&quot;dleelab/iasvaExamples&quot;)</code></pre>
</div>
<div id="load-packages" class="section level2">
<h2>Load packages</h2>
<pre class="r"><code>rm(list=ls())
library(irlba) # partial SVD, the augmented implicitly restarted Lanczos bidiagonalization algorithm
library(iasva)
library(iasvaExamples)
library(sva)
library(polyester)
library(corrplot)
library(DescTools) #pcc i.e., Pearson&#39;s contingency coefficient
library(RColorBrewer)
library(reshape2)
library(ggplot2)
library(SummarizedExperiment)

color.vec &lt;- brewer.pal(9, &quot;Set1&quot;)

iasva.tmp &lt;- function(Y, X, intercept=TRUE, num.sv=NULL, permute=TRUE, num.p=100, sig.cutoff= 0.05, threads=1, num.sv.permtest=NULL, tol=1e-10, verbose=FALSE){
  cat(&quot;IA-SVA running...&quot;)
  sv &lt;- NULL
  pc.stat.obs &lt;- NULL
  pval &lt;- NULL
  wgt.mat &lt;- NULL
  rsq &lt;- NULL
  isv &lt;- 0
  while(TRUE){
    if(!is.null(num.sv)){
      if(isv==num.sv){ 
        break
      }
    }
    iasva.res &lt;- iasva.unit.tmp(Y, X, intercept, permute, num.p, threads, num.sv.permtest, tol, verbose)
    if(iasva.res$pval &lt; sig.cutoff){
      sv &lt;- cbind(sv, iasva.res$sv)
      pc.stat.obs &lt;- cbind(pc.stat.obs, iasva.res$pc.stat.obs)
      pval &lt;- c(pval, iasva.res$pval)
      wgt.mat &lt;- cbind(wgt.mat, iasva.res$wgt)
      X &lt;- cbind(X,iasva.res$sv)
    } else { break }
    isv &lt;- isv+1
    cat(paste0(&quot;\nSV&quot;,isv, &quot; Detected!&quot;))
  }
  if (isv &gt; 0) {
    colnames(sv) &lt;- paste0(&quot;SV&quot;, 1:ncol(sv))
    cat(paste0(&quot;\n# of significant surrogate variables: &quot;,length(pval)))
    return(list(sv=sv, pc.stat.obs=pc.stat.obs, pval=pval, n.sv=length(pval), wgt.mat=wgt.mat))
  } else {
    cat (&quot;\nNo significant surrogate variables&quot;)
  }
}

iasva.unit.tmp &lt;- function(Y, X, intercept=TRUE, permute=TRUE, num.p=100, threads=1, num.sv.permtest=NULL, tol=1e-10, verbose=FALSE){
  if(min(Y)&lt;0){ Y &lt;- Y + abs(min(Y)) }
  lY &lt;- log(Y+1)
  if(intercept){
    fit &lt;- .lm.fit(cbind(1,X), lY)
  } else {
    fit &lt;- .lm.fit(X, lY)
  }
  resid &lt;- resid(fit)
  tresid = t(resid)
  
  if(verbose) {cat(&quot;\n Perform SVD on residuals&quot;)}
  svd_pca &lt;- irlba::irlba(tresid-rowMeans(tresid), 1, tol=tol)
  
  if(verbose) {cat(&quot;\n Regress residuals on PC1&quot;)}
  fit &lt;- .lm.fit(cbind(1,svd_pca$v[,1]), resid)
  
  if(verbose) {cat(&quot;\n Get Rsq&quot;)}
  rsq.vec &lt;- calc.rsq(resid, fit)
  
  if(verbose) {cat(&quot;\n Rsq 0-1 Normalization&quot;)}
  rsq.vec[is.na(rsq.vec)] &lt;- min(rsq.vec, na.rm=TRUE)
  wgt &lt;- (rsq.vec-min(rsq.vec))/(max(rsq.vec)-min(rsq.vec)) #0-1 normalization
  
  if(verbose) {cat(&quot;\n Obtain weighted log-transformed read counts&quot;)}
  tlY = t(lY)*wgt # weigh each row (gene) with respect to its Rsq value.
  
  if(verbose) {cat(&quot;\n Perform SVD on weighted log-transformed read counts&quot;)}
  sv &lt;- irlba::irlba(tlY-rowMeans(tlY), 1, tol=tol)$v[,1]
  
  if(permute==TRUE){
    if(verbose) {cat(&quot;\n Assess the significance of the contribution of SV&quot;)}
    if(is.null(num.sv.permtest)){
      svd.res.obs &lt;- svd(tresid - rowMeans(tresid))
    } else {
      svd.res.obs &lt;- irlba::irlba(tresid-rowMeans(tresid), num.sv.permtest, tol=tol)
    }

    pc.stat.obs &lt;- svd.res.obs$d[1]^2/sum(svd.res.obs$d^2)
    if(verbose) {cat(&quot;\n PC test statistic value:&quot;, pc.stat.obs)}
    
    # Generate an empirical null distribution of the PC test statistic.
    pc.stat.null.vec &lt;- rep(0, num.p)
    permute.svd &lt;- permute.svd.factory(lY, X, num.sv.permtest, tol, verbose)
    if (threads &gt; 1) {
      threads &lt;- min(threads, parallel::detectCores()-1)
      cl &lt;- parallel::makeCluster(threads)
      pc.stat.null.vec &lt;- tryCatch(parallel::parSapply(cl, 1:num.p, permute.svd), error=function(err){parallel::stopCluster(cl); stop(err)})
      parallel::stopCluster(cl)
    } else {
      pc.stat.null.vec &lt;- sapply(1:num.p, permute.svd)
    }
    if(verbose) {cat(&quot;\n Empirical null distribution of the PC statistic:&quot;, sort(pc.stat.null.vec))}
    pval &lt;- sum(pc.stat.obs &lt;= pc.stat.null.vec)/(num.p+1)
    if(verbose) {cat(&quot;\n Permutation p-value:&quot;, pval)}
  } else {
    pc.stat.obs &lt;- -1
    pval &lt;- -1
  }
  return(list(sv=sv, pc.stat.obs=pc.stat.obs, pval=pval, wgt=wgt))
}

calc.rsq &lt;- function(resid, fit) {
  RSS &lt;- colSums(resid(fit)^2)
  TSS &lt;- colSums(t(t(resid) - colSums(resid)/ncol(resid)) ^ 2)  # vectorized
  # TSS &lt;- colSums(sweep(resid, 2, mean, &quot;-&quot;) ^ 2)  # alt-2
  return(1-(RSS/(nrow(resid)-2))/(TSS/(nrow(resid)-1)))
}

permute.svd.factory &lt;- function(lY, X, num.sv.permtest, tol, verbose) {
  
  permute.svd &lt;- function(i) {
    permuted.lY &lt;- apply(t(lY), 1, sample, replace=FALSE)
    tresid.null &lt;- t(resid(.lm.fit(cbind(1,X), permuted.lY)))
    if(is.null(num.sv.permtest)) {
      svd.res.null &lt;- svd(tresid.null)
    } else {
      svd.res.null &lt;- irlba::irlba(tresid.null, num.sv.permtest, tol=tol)
    }
    return(svd.res.null$d[1]^2/sum(svd.res.null$d^2))
  }
  
  return(permute.svd)
}</code></pre>
</div>
<div id="load-the-islet-single-cell-rna-seq-data" class="section level2">
<h2>Load the islet single cell RNA-Seq data</h2>
<pre class="r"><code>data(&quot;Lawlor_Islet_scRNAseq_Read_Counts&quot;)
data(&quot;Lawlor_Islet_scRNAseq_Annotations&quot;)
ls()</code></pre>
<pre><code>[1] &quot;calc.rsq&quot;                          &quot;color.vec&quot;                        
[3] &quot;iasva.tmp&quot;                         &quot;iasva.unit.tmp&quot;                   
[5] &quot;Lawlor_Islet_scRNAseq_Annotations&quot; &quot;Lawlor_Islet_scRNAseq_Read_Counts&quot;
[7] &quot;permute.svd.factory&quot;              </code></pre>
<pre class="r"><code>counts &lt;- Lawlor_Islet_scRNAseq_Read_Counts
anns &lt;- Lawlor_Islet_scRNAseq_Annotations
dim(anns)</code></pre>
<pre><code>[1] 638  26</code></pre>
<pre class="r"><code>dim(counts)</code></pre>
<pre><code>[1] 26542   638</code></pre>
<pre class="r"><code>summary(anns)</code></pre>
<pre><code>     run             cell.type             COL1A1          INS       
 Length:638         Length:638         Min.   :1.00   Min.   :1.000  
 Class :character   Class :character   1st Qu.:1.00   1st Qu.:1.000  
 Mode  :character   Mode  :character   Median :1.00   Median :1.000  
                                       Mean   :1.03   Mean   :1.414  
                                       3rd Qu.:1.00   3rd Qu.:2.000  
                                       Max.   :2.00   Max.   :2.000  
                                                                     
     PRSS1            SST             GCG            KRT19      
 Min.   :1.000   Min.   :1.000   Min.   :1.000   Min.   :1.000  
 1st Qu.:1.000   1st Qu.:1.000   1st Qu.:1.000   1st Qu.:1.000  
 Median :1.000   Median :1.000   Median :1.000   Median :1.000  
 Mean   :1.038   Mean   :1.039   Mean   :1.375   Mean   :1.044  
 3rd Qu.:1.000   3rd Qu.:1.000   3rd Qu.:2.000   3rd Qu.:1.000  
 Max.   :2.000   Max.   :2.000   Max.   :2.000   Max.   :2.000  
                                                                
      PPY          num.genes             Cell_ID        UNOS_ID   
 Min.   :1.000   Min.   :3529   10th_C1_S59  :  1   ACCG268 :136  
 1st Qu.:1.000   1st Qu.:5270   10th_C10_S104:  1   ACJV399 :108  
 Median :1.000   Median :6009   10th_C11_S96 :  1   ACEL337 :103  
 Mean   :1.028   Mean   :5981   10th_C13_S61 :  1   ACIW009 : 93  
 3rd Qu.:1.000   3rd Qu.:6676   10th_C14_S53 :  1   ACCR015A: 57  
 Max.   :2.000   Max.   :8451   10th_C16_S105:  1   ACIB065 : 57  
                                (Other)      :632   (Other) : 84  
      Age        Biomaterial_Provider    Gender              Phenotype  
 Min.   :22.00   IIDP      : 45       Female:303   Non-Diabetic   :380  
 1st Qu.:29.00   Prodo Labs:593       Male  :335   Type 2 Diabetic:258  
 Median :42.00                                                          
 Mean   :39.63                                                          
 3rd Qu.:53.00                                                          
 Max.   :56.00                                                          
                                                                        
               Race          BMI          Cell_Type     Patient_ID 
 African American:175   Min.   :22.00   INS    :264   P1     :136  
 Hispanic        :165   1st Qu.:26.60   GCG    :239   P8     :108  
 White           :298   Median :32.95   KRT19  : 28   P3     :103  
                        Mean   :32.85   SST    : 25   P7     : 93  
                        3rd Qu.:35.80   PRSS1  : 24   P5     : 57  
                        Max.   :55.00   none   : 21   P6     : 57  
                                        (Other): 37   (Other): 84  
 Sequencing_Run Batch       Coverage       Percent_Aligned 
 12t    : 57    B1:193   Min.   :1206135   Min.   :0.8416  
 4th    : 57    B2:148   1st Qu.:2431604   1st Qu.:0.8769  
 9th    : 57    B3:297   Median :3042800   Median :0.8898  
 10t    : 56             Mean   :3160508   Mean   :0.8933  
 7th    : 55             3rd Qu.:3871697   3rd Qu.:0.9067  
 3rd    : 53             Max.   :5931638   Max.   :0.9604  
 (Other):303                                               
 Mitochondrial_Fraction Num_Expressed_Genes
 Min.   :0.003873       Min.   :3529       
 1st Qu.:0.050238       1st Qu.:5270       
 Median :0.091907       Median :6009       
 Mean   :0.108467       Mean   :5981       
 3rd Qu.:0.142791       3rd Qu.:6676       
 Max.   :0.722345       Max.   :8451       
                                           </code></pre>
<pre class="r"><code>ContCoef(table(anns$Gender, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.225969</code></pre>
<pre class="r"><code>ContCoef(table(anns$Phenotype, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.1145096</code></pre>
<pre class="r"><code>ContCoef(table(anns$Race, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.3084146</code></pre>
<pre class="r"><code>ContCoef(table(anns$Patient_ID, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.5232058</code></pre>
<pre class="r"><code>ContCoef(table(anns$Batch, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.3295619</code></pre>
</div>
<div id="look-at-mean-variance-relationship" class="section level2">
<h2>Look at mean variance relationship</h2>
<pre class="r"><code>plot(rowMeans(log(counts+1)),rowVars(log(counts+1)),pch=19,col=color.vec[2])</code></pre>
<p><img src="figure/sim_study_KF_HF_genes_overlap.Rmd/mean_var_relationship-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="estimate-zero-inflated-negative-binomial-parameters-from-the-islet-data" class="section level2">
<h2>Estimate zero inflated negative binomial parameters from the islet data</h2>
<pre class="r"><code>## Estimate the zero inflated negative binomial parameters
params = get_params(counts)</code></pre>
</div>
<div id="generate-a-data-set-affected-by-two-factors-known-and-unknown." class="section level2">
<h2>Generate a data set affected by two factors (known and unknown).</h2>
<p>Here we compare unsupervised SVA, supervised SVA and IA-SVA</p>
<pre class="r"><code>set.seed(1234) #5000
sample.size &lt;- 50
num.genes &lt;- 10000
prop.kfactor.genes &lt;- 0.2    #known factor
prop.hfactor1.genes &lt;- 0.1   #hidden factor1

num.kfactor.genes &lt;- num.genes*prop.kfactor.genes
num.hfactor1.genes &lt;- num.genes*prop.hfactor1.genes

factor.prop &lt;- 0.5
kfactor = c(rep(-1,each=sample.size*factor.prop),rep(1,each=sample.size-(sample.size*factor.prop)))
coinflip = rbinom(sample.size,size=1,prob=0.8)
hfactor1 = kfactor*coinflip + -kfactor*(1-coinflip)

cor(cbind(kfactor,hfactor1))</code></pre>
<pre><code>           kfactor  hfactor1
kfactor  1.0000000 0.6821865
hfactor1 0.6821865 1.0000000</code></pre>
<pre class="r"><code>hfactor.mat &lt;- cbind(hfactor1)
kfcoeffs = c(rnorm(num.kfactor.genes),rep(0,num.genes-num.kfactor.genes))
nullindex= (num.kfactor.genes+1):num.genes

overlap.prop &lt;- 0.99

#hfcoeffs1 = c(rep(0,num.kfactor.genes*(1-overlap.prop)),rnorm(num.hfactor1.genes,sd=1),rep(0,num.genes-num.kfactor.genes*(1-overlap.prop)-num.hfactor1.genes))
hfcoeffs1 = c(rep(0,num.kfactor.genes-num.hfactor1.genes*overlap.prop),rnorm(num.hfactor1.genes,sd=1),rep(0,num.genes-(num.kfactor.genes-num.hfactor1.genes*overlap.prop)-num.hfactor1.genes))

par(mfrow=c(2,1))
plot(kfcoeffs)
plot(hfcoeffs1)</code></pre>
<p><img src="figure/sim_study_KF_HF_genes_overlap.Rmd/simulate_data-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(mfrow=c(1,1))

coeffs = cbind(hfcoeffs1,kfcoeffs)
controls = (hfcoeffs1!=0)&amp;(kfcoeffs==0)
mod = model.matrix(~-1 + hfactor1 + kfactor)

dat0 = create_read_numbers(params$mu,params$fit,
                                     params$p0,beta=coeffs,mod=mod)
sum(dat0==0)/length(dat0)</code></pre>
<pre><code>[1] 0.680858</code></pre>
<pre class="r"><code>filter = apply(dat0, 1, function(x) length(x[x&gt;5])&gt;=2)
dat0 = dat0[filter,]
sum(dat0==0)/length(dat0)</code></pre>
<pre><code>[1] 0.5555512</code></pre>
<pre class="r"><code>controls &lt;- controls[filter]

dim(dat0)</code></pre>
<pre><code>[1] 7067   50</code></pre>
<pre class="r"><code>dim(mod)</code></pre>
<pre><code>[1] 50  2</code></pre>
</div>
<div id="estimate-hidden-factor-using-ia-sva" class="section level2">
<h2>Estimate hidden factor using IA-SVA</h2>
<p>We also plot effect sizes of known and hidden factor and estimated weights (R-squared)</p>
<pre class="r"><code>## Set null and alternative models
mod1 = model.matrix(~kfactor)
mod0 = cbind(mod1[,1])

### iasva 
hfactors_iasva.res &lt;- iasva.tmp(t(dat0), kfactor, num.sv=1, permute=F)</code></pre>
<pre><code>IA-SVA running...
SV1 Detected!
# of significant surrogate variables: 1</code></pre>
<pre class="r"><code>hfactors_iasva &lt;- hfactors_iasva.res$sv
cor(hfactor1,hfactors_iasva)</code></pre>
<pre><code>            SV1
[1,] -0.9871754</code></pre>
<pre class="r"><code>filter2 &lt;- hfcoeffs1[filter]!=0

par(mfrow=c(3,1))
plot(kfcoeffs[filter], xlim=c(0,2500), xlab=&quot;Gene Index&quot;, ylab=&quot;Known Factor Effect Size&quot;, cex=1.5)
plot(hfcoeffs1[filter], xlim=c(0,2500), xlab=&quot;Gene Index&quot;, ylab=&quot;Hidden Factor Effect Size&quot;)
plot(hfactors_iasva.res$wgt.mat, xlim=c(0,2500), xlab=&quot;Gene Index&quot;, ylab=&quot;R squared&quot;)</code></pre>
<p><img src="figure/sim_study_KF_HF_genes_overlap.Rmd/iasva_hidden_factor_estimation-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(mfrow=c(1,1))


pdf(&quot;output/GeneOverlap.Fig1.pdf&quot;, width = 5, height=8)
par(mfrow=c(3,1))
plot(kfcoeffs[filter], xlim=c(0,2500), xlab=&quot;Gene Index&quot;, ylab=&quot;Known Factor Effect Size&quot;,cex.lab=1.5,cex.axis=1.5)
plot(hfcoeffs1[filter], xlim=c(0,2500), xlab=&quot;Gene Index&quot;, ylab=&quot;Hidden Factor Effect Size&quot;,cex.lab=1.5,cex.axis=1.5)
plot(hfactors_iasva.res$wgt.mat, xlim=c(0,2500), xlab=&quot;Gene Index&quot;, ylab=&quot;R squared&quot;,cex.lab=1.5,cex.axis=1.5)
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>par(mfrow=c(1,1))</code></pre>
</div>
<div id="r-squared-weights-as-a-function-of-hidden-and-known-factors-effect-size" class="section level2">
<h2>R-squared (Weights) as a function of hidden and known factors’ effect size</h2>
<pre class="r"><code>overlap.hfcoeffs1 &lt;- hfcoeffs1[filter][filter2]
overlap.kfcoeffs &lt;- kfcoeffs[filter][filter2]
overlap.R2 &lt;- hfactors_iasva.res$wgt.mat[filter2]

summary(lm(overlap.R2~overlap.hfcoeffs1+overlap.kfcoeffs))</code></pre>
<pre><code>
Call:
lm(formula = overlap.R2 ~ overlap.hfcoeffs1 + overlap.kfcoeffs)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.14720 -0.10585 -0.07321  0.02319  0.91561 

Coefficients:
                  Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)       0.117215   0.006611  17.730   &lt;2e-16 ***
overlap.hfcoeffs1 0.012627   0.006434   1.963   0.0501 .  
overlap.kfcoeffs  0.001709   0.006849   0.250   0.8030    
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 0.1768 on 712 degrees of freedom
Multiple R-squared:  0.005446,  Adjusted R-squared:  0.002652 
F-statistic: 1.949 on 2 and 712 DF,  p-value: 0.1431</code></pre>
<pre class="r"><code>par(mfrow=c(2,1))
plot(overlap.hfcoeffs1, overlap.R2, xlab=&quot;Hidden Factor Effect Size&quot;, ylab=&quot;R squared&quot;)
plot(overlap.kfcoeffs, overlap.R2, xlab=&quot;Known Factor Effect Size&quot;, ylab=&quot;R squared&quot;)</code></pre>
<p><img src="figure/sim_study_KF_HF_genes_overlap.Rmd/iasva_results_analysis-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(mfrow=c(1,1))

pdf(&quot;output/GeneOverlap.Fig2.pdf&quot;, width = 5, height=8)
par(mfrow=c(2,1))
plot(overlap.hfcoeffs1, overlap.R2, xlab=&quot;Hidden Factor Effect Size&quot;, ylab=&quot;R squared&quot;)
plot(overlap.kfcoeffs, overlap.R2, xlab=&quot;Known Factor Effect Size&quot;, ylab=&quot;R squared&quot;)
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>par(mfrow=c(1,1))</code></pre>
</div>
<div id="hidden-factor-estimation-using-existing-methods" class="section level2">
<h2>Hidden factor estimation using existing methods</h2>
<pre class="r"><code>## Estimate batch with pca
ldat0 = log(dat0 + 1)
hfactors_pca = svd(ldat0 - rowMeans(ldat0))$v[,1]
cor(hfactor1,hfactors_pca)</code></pre>
<pre><code>[1] -0.8373417</code></pre>
<pre class="r"><code>## Estimate batch with svaseq (unsupervised)
hfactors_usva = svaseq(dat0,mod1,mod0, n.sv=1)$sv</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  </code></pre>
<pre class="r"><code>cor(hfactor1,hfactors_usva)</code></pre>
<pre><code>           [,1]
[1,] -0.9720361</code></pre>
<pre class="r"><code>## Estimate batch with svaseq (supervised)
hfactors_ssva = svaseq(dat0,mod1,mod0,controls=controls, n.sv=1)$sv</code></pre>
<pre><code>sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 </code></pre>
<pre class="r"><code>cor(hfactor1,hfactors_ssva)</code></pre>
<pre><code>           [,1]
[1,] 0.02989441</code></pre>
</div>
<div id="plot-hidden-factor-estimates-of-ia-sva-and-existing-methods" class="section level2">
<h2>Plot hidden factor estimates of IA-SVA and existing methods</h2>
<pre class="r"><code>## Plot the results
par(mfrow=c(5,1))
plot(hfactor1,col=color.vec[1],pch=19,main=&quot;Hidden Factor1&quot;)
plot(hfactors_iasva,pch=19,col=color.vec[2],main=&quot;IA-SVA&quot;)
plot(hfactors_pca,pch=19, col=color.vec[3],main=&quot;PCA&quot;)
plot(hfactors_usva,pch=19,col=color.vec[4],main=&quot;USVA&quot;)
plot(hfactors_ssva,pch=19,col=color.vec[5],main=&quot;SSVA&quot;)</code></pre>
<p><img src="figure/sim_study_KF_HF_genes_overlap.Rmd/comp_figures-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(mfrow=c(1,1))</code></pre>
<p><br></p>
</div>
<div id="multiple-simulations-where-hidden-and-known-factors-affect-overlapping-sets-of-genes" class="section level2">
<h2>Multiple simulations where hidden and known factors affect overlapping sets of genes</h2>
<pre class="r"><code>set.seed(4000)
sample.size &lt;- 50
num.genes &lt;- 10000
prop.kfactor.genes &lt;- 0.2    #known factor
prop.hfactor1.genes &lt;- 0.1   #hidden factor1

num.kfactor.genes &lt;- num.genes*prop.kfactor.genes
num.hfactor1.genes &lt;- num.genes*prop.hfactor1.genes

factor.prop &lt;- 0.5
kfactor = c(rep(-1,each=sample.size*factor.prop),rep(1,each=sample.size-(sample.size*factor.prop)))
coinflip = rbinom(sample.size,size=1,prob=0.9)
hfactor1 = kfactor*coinflip + -kfactor*(1-coinflip)

cor(cbind(kfactor,hfactor1))</code></pre>
<pre><code>         kfactor hfactor1
kfactor     1.00     0.76
hfactor1    0.76     1.00</code></pre>
<pre class="r"><code>overlap.prop.vec &lt;- c(0.99,0.9,0.8,0.7,0.6,0.5,0.4,0.3,0.2,0.1,0)

data.list &lt;- list()
controls.list &lt;- list()

hfactor.mat &lt;- cbind(hfactor1)
kfbetas &lt;- rnorm(num.kfactor.genes)
hfbetas &lt;- rnorm(num.hfactor1.genes)  

kfcoeffs = c(rnorm(num.kfactor.genes),rep(0,num.genes-num.kfactor.genes))
nullindex= (num.kfactor.genes+1):num.genes

pdf(&quot;output/gene_overlap.pdf&quot;, width = 10, height=5)
for(i in 1:length(overlap.prop.vec)){

  hfcoeffs1 = c(rep(0,num.kfactor.genes-num.hfactor1.genes*overlap.prop.vec[i]),hfbetas,rep(0,num.genes-(num.kfactor.genes-num.hfactor1.genes*overlap.prop.vec[i])-num.hfactor1.genes))
  
  coeffs = cbind(hfcoeffs1,kfcoeffs)
  controls = (hfcoeffs1!=0)&amp;(kfcoeffs==0)
  mod = model.matrix(~-1 + hfactor1 + kfactor)

  dat0 = create_read_numbers(params$mu,params$fit,
                                       params$p0,beta=coeffs,mod=mod)
  sum(dat0==0)/length(dat0)
  filter = apply(dat0, 1, function(x) length(x[x&gt;5])&gt;=2)
  dat0 = dat0[filter,]
  sum(dat0==0)/length(dat0)
  controls.list[[i]] &lt;- controls[filter]

  #dim(dat0)
  #dim(mod)
  data.list[[i]] &lt;- dat0
  par(mfrow=c(2,1))
  print(plot(kfcoeffs))
  print(plot(hfcoeffs1))
}</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre><code>NULL</code></pre>
<pre class="r"><code>dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="estimation-of-hidden-factors-using-ia-sva-and-existing-methods" class="section level2">
<h2>Estimation of hidden factors using IA-SVA and existing methods</h2>
<pre class="r"><code>## Set null and alternative models
mod1 = model.matrix(~kfactor)
mod0 = cbind(mod1[,1])
hfactors_iasva_list &lt;- list()
hfactors_pca_list &lt;- list()
hfactors_usva_list &lt;- list()
hfactors_ssva_list &lt;- list()
for(i in 1:length(overlap.prop.vec)){
  dat0 &lt;- data.list[[i]]
  ### iasva 
  summ_exp &lt;- SummarizedExperiment(assays = dat0)
  hfactors_iasva_list[[i]] &lt;- iasva(summ_exp, as.matrix(kfactor), num.sv=1, permute = F)$sv
  ## Estimate batch with pca
  ldat0 = log(dat0 + 1)
  hfactors_pca_list[[i]] = svd(ldat0 - rowMeans(ldat0))$v[,1]
  ## Estimate batch with svaseq (unsupervised)
  hfactors_usva_list[[i]] = svaseq(dat0,mod1,mod0,n.sv=1)$sv[,1]
  ## Estimate batch with svaseq (supervised)
  hfactors_ssva_list[[i]] = svaseq(dat0,mod1,mod0,n.sv=1,controls=controls.list[[i]])$sv[,1]
  cat(i,&quot;\n&quot;)
}</code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
1 </code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
2 </code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
3 </code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
4 </code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
5 </code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
6 </code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
7 </code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
8 </code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
9 </code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
10 </code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 1</code></pre>
<pre><code>Number of significant surrogate variables is:  1 
Iteration (out of 5 ):1  2  3  4  5  sva warning: controls provided so supervised sva is being performed.
Number of significant surrogate variables is:  1 
11 </code></pre>
</div>
<div id="compare-the-accuracy" class="section level2">
<h2>Compare the accuracy</h2>
<pre class="r"><code>iasva.cor.vec &lt;- as.vector(abs(cor(hfactor1, do.call(cbind, hfactors_iasva_list))))
pca.cor.vec &lt;- as.vector(abs(cor(hfactor1, do.call(cbind, hfactors_pca_list))))
usva.cor.vec &lt;- as.vector(abs(cor(hfactor1, do.call(cbind, hfactors_usva_list))))
ssva.cor.vec &lt;- as.vector(abs(cor(hfactor1, do.call(cbind, hfactors_ssva_list))))

cor.df &lt;- data.frame(overlap.pct = 100*overlap.prop.vec,
                     IASVA = iasva.cor.vec,
                     PCA = pca.cor.vec,
                     USVA = usva.cor.vec,
                     SSVA = ssva.cor.vec)

melt.cor.df &lt;- melt(cor.df, id.var=c(&quot;overlap.pct&quot;))

p &lt;- ggplot(melt.cor.df, aes(x=overlap.pct, y=value, group=variable))
p &lt;- p + geom_line(aes(col=variable), size=1)
p &lt;- p + geom_point(aes(col=variable), size=3)
p &lt;- p + xlab(&quot;Gene Set Overlap Percentage&quot;)
p &lt;- p + ylab(&quot;Absolute Correlation Coefficient&quot;)
#p &lt;- p + scale_color_manual(values=color.vec)
p &lt;- p + scale_color_brewer(palette = &quot;Spectral&quot;, name=&quot;Method&quot;)
p &lt;- p + theme_bw()
p &lt;- p + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
               #panel.grid.minor = element_blank(), 
               axis.line = element_line(colour = &quot;black&quot;))
p</code></pre>
<p><img src="figure/sim_study_KF_HF_genes_overlap.Rmd/assess_estimation_accuracy-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggsave(&quot;output/absCor_gene_overlap_pct.pdf&quot;, width=8, height=5)</code></pre>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.0 (2018-04-23)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] SummarizedExperiment_1.10.1 DelayedArray_0.6.1         
 [3] matrixStats_0.53.1          Biobase_2.40.0             
 [5] GenomicRanges_1.32.3        GenomeInfoDb_1.16.0        
 [7] IRanges_2.14.10             S4Vectors_0.18.3           
 [9] BiocGenerics_0.26.0         ggplot2_2.2.1.9000         
[11] reshape2_1.4.3              RColorBrewer_1.1-2         
[13] DescTools_0.99.24           corrplot_0.84              
[15] polyester_1.16.0            sva_3.28.0                 
[17] BiocParallel_1.14.2         genefilter_1.62.0          
[19] mgcv_1.8-23                 nlme_3.1-137               
[21] iasvaExamples_1.0.0         iasva_0.99.3               
[23] irlba_2.3.2                 Matrix_1.2-14              

loaded via a namespace (and not attached):
 [1] bitops_1.0-6           bit64_0.9-7            rprojroot_1.3-2       
 [4] tools_3.5.0            backports_1.1.2        R6_2.2.2              
 [7] DBI_1.0.0              lazyeval_0.2.1         colorspace_1.3-2      
[10] withr_2.1.2            tidyselect_0.2.4       bit_1.1-14            
[13] compiler_3.5.0         git2r_0.21.0           expm_0.999-2          
[16] labeling_0.3           scales_0.5.0           mvtnorm_1.0-8         
[19] stringr_1.3.1          digest_0.6.15          foreign_0.8-70        
[22] rmarkdown_1.9          R.utils_2.6.0          XVector_0.20.0        
[25] pkgconfig_2.0.1        htmltools_0.3.6        manipulate_1.0.1      
[28] limma_3.36.2           rlang_0.2.1            RSQLite_2.1.1         
[31] bindr_0.1.1            dplyr_0.7.5            R.oo_1.22.0           
[34] RCurl_1.95-4.10        magrittr_1.5           GenomeInfoDbData_1.1.0
[37] Rcpp_0.12.17           munsell_0.4.3          R.methodsS3_1.7.1     
[40] stringi_1.2.2          whisker_0.3-2          logspline_2.1.11      
[43] yaml_2.1.19            MASS_7.3-50            zlibbioc_1.26.0       
[46] plyr_1.8.4             grid_3.5.0             blob_1.1.1            
[49] lattice_0.20-35        Biostrings_2.48.0      splines_3.5.0         
[52] annotate_1.58.0        knitr_1.20             pillar_1.2.3          
[55] boot_1.3-20            XML_3.98-1.11          glue_1.2.0            
[58] evaluate_0.10.1        gtable_0.2.0           purrr_0.2.5           
[61] assertthat_0.2.0       xtable_1.8-2           survival_2.42-3       
[64] tibble_1.4.2           AnnotationDbi_1.42.1   memoise_1.1.0         
[67] workflowr_1.0.1        bindrcpp_0.2.2         cluster_2.0.7-1       </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1
</p>
<hr>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
