<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Donghyung Lee" />

<meta name="date" content="2018-08-03" />

<title>IA-SVA based feature selection improves the performance of clustering algorithms [2]</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">iasvaExamples</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/dleelab/iasvaExamples">GitHub</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">IA-SVA based feature selection improves the performance of clustering algorithms [2]</h1>
<h4 class="author"><em>Donghyung Lee</em></h4>
<h4 class="date"><em>2018-08-03</em></h4>

</div>

<div id="TOC">
<ul>
<li><a href="#load-packages">Load packages</a></li>
<li><a href="#load-the-islet-scrna-seq-data">Load the islet scRNA-Seq data</a></li>
<li><a href="#filter-out-low-expressed-genes">Filter out low-expressed genes</a></li>
<li><a href="#calculate-the-number-of-detected-genes">Calculate the number of detected genes</a></li>
<li><a href="#run-tsne-to-cluster-islet-cells.">Run tSNE to cluster islet cells.</a></li>
<li><a href="#run-cellview-to-cluster-islet-cells">Run CellView to cluster islet cells</a></li>
<li><a href="#run-surrogate-variable-analysis-sva.">Run surrogate variable analysis (SVA).</a></li>
<li><a href="#run-ia-sva">Run IA-SVA</a></li>
<li><a href="#correlation-between-svs">Correlation between SVs</a></li>
<li><a href="#clustering-analyses">clustering analyses</a></li>
<li><a href="#find-marker-genes-for-sv1.">Find marker genes for SV1.</a></li>
<li><a href="#find-marker-genes-for-sv3.">Find marker genes for SV3.</a></li>
<li><a href="#run-tsne-post-ia-sva-i.e.-run-tsne-on-marker-genes-obtained-from-ia-sva.">Run tSNE post IA-SVA, i.e., run tSNE on marker genes obtained from IA-SVA.</a></li>
<li><a href="#session-information">Session information</a></li>
</ul>
</div>

<p><strong>Last updated:</strong> 2018-08-03</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(20180731)</code> </summary></p>
<p>The command <code>set.seed(20180731)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/dleelab/iasvaExamples/tree/268983524d3aee232c1bf032e8dd8993bcfb4fc3" target="_blank">2689835</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    inst/.DS_Store
    Ignored:    inst/doc/.DS_Store
    Ignored:    vignettes/.DS_Store

Untracked files:
    Untracked:  Clustering_analyses_figure4_islets_sv1_3.pdf
    Untracked:  docs/figure/hidden_heterogeneity_glioblastoma.Rmd/
    Untracked:  docs/figure/tSNE_post_IA-SVA_3celltypes.Rmd/
    Untracked:  output/CC_genes.long.txt
    Untracked:  output/CC_genes.short.txt
    Untracked:  output/Clustering_analyses_figure3_sv1.pdf
    Untracked:  output/Lawlor_Islets_3Cells_CellView_Seurat_FigS.pdf
    Untracked:  output/Lawlor_Islets_3Cells_IASVA_SV1SV3_rsqcutoff0.3_pheatmap_iasvaV0.95_Figure4_C.pdf
    Untracked:  output/Lawlor_Islets_3Cells_IASVA_SV4_rsqcutoff0.3_pheatmap_iasvaV0.95.pdf
    Untracked:  output/Lawlor_Islets_3Cells_IASVA_pairs4SVs_iasvaV0.95_black_FigS6.pdf
    Untracked:  output/Lawlor_Islets_3Cells_IASVA_pairs4SVs_iasvaV0.95_color_FigS6.pdf
    Untracked:  output/Lawlor_Islets_3Cells_SV1_SV3_Cell_Type_Genes_rsqcutoff0.3.txt
    Untracked:  output/Lawlor_Islets_3Cells_SV4_Genes_rsqcutoff0.3.txt
    Untracked:  output/Lawlor_Islets_3Cells_tSNE_IA-SVA_Fig4AB.pdf
    Untracked:  output/Patel_Glioblastoma_MGH30_CellCycle_Figure3ABCD.pdf
    Untracked:  output/Patel_Glioblastoma_MGH30_Cellcycle_SV1_Genes_rsqcutoff0.3.txt
    Untracked:  output/Patel_Glioblastoma_MGH30_Cellcycle_SV1_Genes_rsqcutoff0.4.txt
    Untracked:  output/Patel_Glioblastoma_MGH30_iasva_SV1_genes_rsqcutoff0.3_pheatmap_iasvaV0.95_Figure3F.pdf

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/e3d1fb211887a2e6342e133cf891368150e4b3cf/analysis/tSNE_post_IA-SVA_Xin_Islets.Rmd" target="_blank">e3d1fb2</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
figure output path updated
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/c241b04dac899407045051a04b7539237f4d183f/analysis/tSNE_post_IA-SVA_Xin_Islets.Rmd" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/dleelab/iasvaExamples/c241b04dac899407045051a04b7539237f4d183f/docs/tSNE_post_IA-SVA_Xin_Islets.html" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<p>Here, we used single cell RNA sequencing (scRNA-Seq) data with strong confounding variables, which is also obtained from human pancreatic islet samples (<a href="http://www.cell.com/cell-metabolism/abstract/S1550-4131(16)30434-X">Xin et. al., 2016</a>). This dataset is included in an R data package (“iasvaExamples”) containing data examples for IA-SVA (<a href="https://github.com/dleelab/iasvaExamples" class="uri">https://github.com/dleelab/iasvaExamples</a>). To install the ‘iasvaExamples’ package, follow the instructions provided in the GitHub page.</p>
<div id="load-packages" class="section level2">
<h2>Load packages</h2>
<pre class="r"><code>rm(list=ls())
library(sva)</code></pre>
<pre><code>Loading required package: mgcv</code></pre>
<pre><code>Loading required package: nlme</code></pre>
<pre><code>This is mgcv 1.8-23. For overview type &#39;help(&quot;mgcv-package&quot;)&#39;.</code></pre>
<pre><code>Loading required package: genefilter</code></pre>
<pre><code>Loading required package: BiocParallel</code></pre>
<pre class="r"><code>library(Seurat)</code></pre>
<pre><code>Loading required package: ggplot2</code></pre>
<pre><code>Loading required package: cowplot</code></pre>
<pre><code>
Attaching package: &#39;cowplot&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:ggplot2&#39;:

    ggsave</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre class="r"><code>library(iasva)
library(iasvaExamples)
library(dbscan)
library(irlba)
library(Rtsne)
library(pheatmap)
library(corrplot)</code></pre>
<pre><code>corrplot 0.84 loaded</code></pre>
<pre class="r"><code>library(DescTools) #pcc i.e., Pearson&#39;s contingency coefficient</code></pre>
<pre><code>
Attaching package: &#39;DescTools&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:genefilter&#39;:

    AUC</code></pre>
<pre class="r"><code>library(RColorBrewer)
library(cluster)
library(SummarizedExperiment)</code></pre>
<pre><code>Loading required package: GenomicRanges</code></pre>
<pre><code>Loading required package: stats4</code></pre>
<pre><code>Loading required package: BiocGenerics</code></pre>
<pre><code>Loading required package: parallel</code></pre>
<pre><code>
Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:parallel&#39;:

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>The following objects are masked from &#39;package:Matrix&#39;:

    colMeans, colSums, rowMeans, rowSums, which</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, append, as.data.frame, basename, cbind,
    colMeans, colnames, colSums, dirname, do.call, duplicated,
    eval, evalq, Filter, Find, get, grep, grepl, intersect,
    is.unsorted, lapply, lengths, Map, mapply, match, mget, order,
    paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,
    Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which, which.max,
    which.min</code></pre>
<pre><code>Loading required package: S4Vectors</code></pre>
<pre><code>
Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:Matrix&#39;:

    expand</code></pre>
<pre><code>The following object is masked from &#39;package:base&#39;:

    expand.grid</code></pre>
<pre><code>Loading required package: IRanges</code></pre>
<pre><code>
Attaching package: &#39;IRanges&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:nlme&#39;:

    collapse</code></pre>
<pre><code>Loading required package: GenomeInfoDb</code></pre>
<pre><code>Loading required package: Biobase</code></pre>
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>Loading required package: DelayedArray</code></pre>
<pre><code>Loading required package: matrixStats</code></pre>
<pre><code>
Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:Biobase&#39;:

    anyMissing, rowMedians</code></pre>
<pre><code>The following objects are masked from &#39;package:genefilter&#39;:

    rowSds, rowVars</code></pre>
<pre><code>
Attaching package: &#39;DelayedArray&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:matrixStats&#39;:

    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    aperm, apply</code></pre>
<pre class="r"><code>color.vec &lt;- brewer.pal(8, &quot;Set1&quot;)
tol21rainbow= c(&quot;#771155&quot;, &quot;#AA4488&quot;, &quot;#CC99BB&quot;, &quot;#114477&quot;, &quot;#4477AA&quot;,
                &quot;#77AADD&quot;, &quot;#117777&quot;, &quot;#44AAAA&quot;, &quot;#77CCCC&quot;, &quot;#117744&quot;,
                &quot;#44AA77&quot;, &quot;#88CCAA&quot;, &quot;#777711&quot;, &quot;#AAAA44&quot;, &quot;#DDDD77&quot;,
                &quot;#774411&quot;, &quot;#AA7744&quot;, &quot;#DDAA77&quot;, &quot;#771122&quot;, &quot;#AA4455&quot;,
                &quot;#DD7788&quot;)

# Normalization.
normalize &lt;- function(counts) 
{
    normfactor &lt;- colSums(counts)
    return(t(t(counts)/normfactor)*median(normfactor))
}</code></pre>
</div>
<div id="load-the-islet-scrna-seq-data" class="section level2">
<h2>Load the islet scRNA-Seq data</h2>
<pre class="r"><code>data(&quot;Xin_Islet_scRNAseq_Read_Counts&quot;)
data(&quot;Xin_Islet_scRNAseq_Annotations&quot;)
ls()</code></pre>
<pre><code>[1] &quot;color.vec&quot;                      &quot;normalize&quot;                     
[3] &quot;tol21rainbow&quot;                   &quot;Xin_Islet_scRNAseq_Annotations&quot;
[5] &quot;Xin_Islet_scRNAseq_Read_Counts&quot;</code></pre>
<pre class="r"><code>counts &lt;- Xin_Islet_scRNAseq_Read_Counts
anns &lt;- Xin_Islet_scRNAseq_Annotations
dim(anns)</code></pre>
<pre><code>[1] 1600    9</code></pre>
<pre class="r"><code>dim(counts)</code></pre>
<pre><code>[1] 26542  1600</code></pre>
<pre class="r"><code>Lib_Size &lt;- colSums(counts)
plot(sort(Lib_Size))</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/load_data-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>hist(Lib_Size)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/load_data-2.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>summary(Lib_Size)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 199678  773969 1002044 1104395 1288619 7369192 </code></pre>
<p>The annotations describing the islet samples and experimental settings are stored in “anns” and the read counts information is stored in “counts”.</p>
</div>
<div id="filter-out-low-expressed-genes" class="section level2">
<h2>Filter out low-expressed genes</h2>
<pre class="r"><code>dim(counts)</code></pre>
<p>[1] 26542 1600</p>
<pre class="r"><code>dim(anns)</code></pre>
<p>[1] 1600 9</p>
<pre class="r"><code>anns &lt;- droplevels(anns)
prop.zeros &lt;- sum(counts==0)/length(counts)
prop.zeros</code></pre>
<p>[1] 0.7040829</p>
<pre class="r"><code>filter = apply(counts, 1, function(x) length(x[x&gt;5])&gt;=3)
counts = counts[filter,]
dim(counts)</code></pre>
<p>[1] 19169 1600</p>
<pre class="r"><code>prop.zeros &lt;- sum(counts==0)/length(counts)
prop.zeros</code></pre>
<p>[1] 0.5913386</p>
<pre class="r"><code>summary(anns)</code></pre>
<pre><code>      X             Age        Cell_Type          Condition  </code></pre>
<p>SRR3541303: 1 Min. :23.00 alpha:946 non-diabetic:651<br />
SRR3541304: 1 1st Qu.:37.00 beta :503 T2D :949<br />
SRR3541305: 1 Median :42.00 delta: 58<br />
SRR3541306: 1 Mean :43.89 PP : 93<br />
SRR3541307: 1 3rd Qu.:55.00<br />
SRR3541308: 1 Max. :68.00<br />
(Other) :1594<br />
Donor_ID Ethnicity Gender Num_Expressed_Genes T2D 4 :306 AA:369 female:658 Min. :1745<br />
T2D 6 :224 AI: 45 male :942 1st Qu.:4412<br />
T2D 5 :176 C :593 Median :5158<br />
Non T2D 12:144 H :593 Mean :5185<br />
T2D 1 : 91 3rd Qu.:5902<br />
T2D 2 : 83 Max. :9731<br />
(Other) :576<br />
Mitochondrial_Fraction Min. : 0.2812<br />
1st Qu.: 4.1449<br />
Median : 6.3422<br />
Mean : 7.4164<br />
3rd Qu.:10.1241<br />
Max. :22.8479</p>
<pre class="r"><code>Patient_ID &lt;- anns$Donor_ID
Gender &lt;- anns$Gender
Age &lt;- anns$Age
Cell_Type &lt;- anns$Cell_Type
Phenotype &lt;- anns$Condition
Ethnicity &lt;- anns$Ethnicity
Mito_Frac &lt;- anns$Mitochondrial_Fraction
table(Cell_Type, Patient_ID)</code></pre>
<pre><code>     Patient_ID</code></pre>
<p>Cell_Type Non T2D 1 Non T2D 10 Non T2D 11 Non T2D 12 Non T2D 2 Non T2D 3 alpha 56 20 49 92 4 29 beta 13 26 12 50 13 10 delta 5 2 1 2 2 3 PP 3 1 1 0 5 2 Patient_ID Cell_Type Non T2D 4 Non T2D 5 Non T2D 6 Non T2D 7 Non T2D 8 Non T2D 9 alpha 15 26 38 8 14 26 beta 11 24 7 17 17 7 delta 0 5 4 0 3 1 PP 5 2 1 4 11 4 Patient_ID Cell_Type T2D 1 T2D 2 T2D 3 T2D 4 T2D 5 T2D 6 alpha 39 50 33 152 161 134 beta 12 18 29 144 12 81 delta 4 5 4 8 1 8 PP 36 10 3 2 2 1</p>
<pre class="r"><code>ContCoef(table(Cell_Type, Patient_ID))</code></pre>
<p>[1] 0.4832611</p>
<pre class="r"><code>table(Cell_Type, Gender)</code></pre>
<pre><code>     Gender</code></pre>
<p>Cell_Type female male alpha 358 588 beta 244 259 delta 24 34 PP 32 61</p>
<pre class="r"><code>ContCoef(table(Cell_Type, Gender))</code></pre>
<p>[1] 0.1033314</p>
<pre class="r"><code>table(Cell_Type, Age)</code></pre>
<pre><code>     Age</code></pre>
<p>Cell_Type 23 24 27 29 31 32 37 41 42 43 51 55 56 57 60 68 alpha 85 26 26 8 49 4 50 152 161 20 134 33 107 39 14 38 beta 23 7 24 17 12 13 18 144 12 26 81 29 61 12 17 7 delta 8 1 5 0 1 2 5 8 1 2 8 4 2 4 3 4 PP 5 4 2 4 1 5 10 2 2 1 1 3 5 36 11 1</p>
<pre class="r"><code>ContCoef(table(Cell_Type, Age))</code></pre>
<p>[1] 0.4782088</p>
<pre class="r"><code>table(Cell_Type, Phenotype)</code></pre>
<pre><code>     Phenotype</code></pre>
<p>Cell_Type non-diabetic T2D alpha 377 569 beta 207 296 delta 28 30 PP 39 54</p>
<pre class="r"><code>ContCoef(table(Cell_Type, Phenotype))</code></pre>
<p>[1] 0.03317408</p>
<pre class="r"><code>table(Cell_Type, Ethnicity)</code></pre>
<pre><code>     Ethnicity</code></pre>
<p>Cell_Type AA AI C H alpha 213 14 384 335 beta 99 17 150 237 delta 16 3 22 17 PP 41 11 37 4</p>
<pre class="r"><code>ContCoef(table(Cell_Type, Ethnicity))</code></pre>
<p>[1] 0.2517959</p>
<pre class="r"><code>ContCoef(table(Patient_ID, Gender))</code></pre>
<p>[1] 0.7071068</p>
<pre class="r"><code>ContCoef(table(Patient_ID, Age))</code></pre>
<p>[1] 0.9682458</p>
<pre class="r"><code>ContCoef(table(Patient_ID, Phenotype))</code></pre>
<p>[1] 0.7071068</p>
<pre class="r"><code>ContCoef(table(Patient_ID, Ethnicity))</code></pre>
<p>[1] 0.8660254</p>
<pre class="r"><code>ContCoef(table(Gender, Age))</code></pre>
<p>[1] 0.6803731</p>
<pre class="r"><code>ContCoef(table(Gender, Phenotype))</code></pre>
<p>[1] 0.1724809</p>
<pre class="r"><code>ContCoef(table(Gender, Ethnicity))</code></pre>
<p>[1] 0.4526994</p>
<pre class="r"><code>ContCoef(table(Age, Phenotype))</code></pre>
<p>[1] 0.7071068</p>
<pre class="r"><code>ContCoef(table(Age, Ethnicity))</code></pre>
<p>[1] 0.8569589</p>
<pre class="r"><code>ContCoef(table(Phenotype, Ethnicity))</code></pre>
<p>[1] 0.4785346</p>
<pre class="r"><code>raw.counts &lt;- counts
summary(colSums(counts))</code></pre>
<p>Min. 1st Qu. Median Mean 3rd Qu. Max. 199628 773880 1001875 1104286 1288589 7369004</p>
<pre class="r"><code>counts &lt;- normalize(counts)
summary(colSums(counts))</code></pre>
<p>Min. 1st Qu. Median Mean 3rd Qu. Max. 1001875 1001875 1001875 1001875 1001875 1001875</p>
<p>Note that the orignial cell assignments are highly correlated with known factors.</p>
</div>
<div id="calculate-the-number-of-detected-genes" class="section level2">
<h2>Calculate the number of detected genes</h2>
<p>It is well known that the number of detected genes in each cell explains a very large portion of variability in scRNA-Seq data (<a href="http://biorxiv.org/content/early/2015/08/25/025528">Hicks et. al. 2015 BioRxiv</a>, <a href="http://www.nature.com/nbt/journal/v34/n6/full/nbt.3498.html">McDavid et. al. 2016 Nature Biotechnology</a>). Frequently, the first principal component of log-transformed scRNA-Seq read counts is highly correlated with the number of detected genes (e.g., r &gt; 0.9). Here, we calculate the number of detected genes for islet cells, which will be used as an known factor in the IA-SVA analyses.</p>
<pre class="r"><code>Num_Detected_Genes &lt;- colSums(counts&gt;0)
summary(Num_Detected_Genes)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   3261    6890    7832    7834    8766   12067 </code></pre>
<pre class="r"><code>barplot(Num_Detected_Genes, xlab=&quot;Cell&quot;, las=2, ylab = &quot;Number of detected genes&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/num_detected_genes-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>lcounts &lt;- log(counts + 1)
# PCA
pca.res = irlba(lcounts - rowMeans(lcounts), 5)$v
cor(Num_Detected_Genes, pca.res[,1])</code></pre>
<pre><code>[1] 0.8906152</code></pre>
<pre class="r"><code>dim(lcounts)</code></pre>
<pre><code>[1] 19169  1600</code></pre>
</div>
<div id="run-tsne-to-cluster-islet-cells." class="section level2">
<h2>Run tSNE to cluster islet cells.</h2>
<p>For comparison purposes,we applied tSNE on read counts of all genes. We used the Rtsne R package with default settings. Genes are color coded wrt their expression of marker genes.</p>
<pre class="r"><code>set.seed(34544532)
tsne.res.all &lt;- Rtsne(t(lcounts), dims = 2)
plot(tsne.res.all$Y, main=&quot;tSNE&quot;, xlab=&quot;tSNE Dim1&quot;, 
     ylab=&quot;tSNE Dim2&quot;,pch=21, col=color.vec[Cell_Type], 
     bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend(&quot;topright&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_tsne-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(mfrow=c(2,2))
plot(tsne.res.all$Y, main=&quot;Gender&quot;, xlab=&quot;Dimension 1&quot;, 
     ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Gender], 
     bg=color.vec[Gender], oma=c(4,4,6,12))
legend(&quot;topleft&quot;, levels(Gender), border=&quot;white&quot;, 
       fill=color.vec, bty=&quot;n&quot;)
plot(tsne.res.all$Y, main=&quot;Patient ID&quot;, xlab=&quot;Dimension 1&quot;,
     ylab=&quot;Dimension 2&quot;, pch=21, col=tol21rainbow[Patient_ID],
     bg=tol21rainbow[Patient_ID], oma=c(4,4,6,12))
legend(&quot;topleft&quot;, levels(Patient_ID), border=&quot;white&quot;, 
       fill=tol21rainbow, bty=&quot;n&quot;, cex=0.5)
plot(tsne.res.all$Y, main=&quot;Ethnicity&quot;, xlab=&quot;Dimension 1&quot;,
     ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Ethnicity],
     bg=color.vec[Ethnicity], oma=c(4,4,6,12))
legend(&quot;topleft&quot;, levels(Ethnicity), border=&quot;white&quot;,
       fill=color.vec, bty=&quot;n&quot;)
plot(tsne.res.all$Y, main=&quot;Phenotype&quot;, xlab=&quot;Dimension 1&quot;,
     ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Phenotype], 
     bg=color.vec[Phenotype], oma=c(4,4,6,12))
legend(&quot;topleft&quot;, levels(Phenotype), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_tsne-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(mfrow=c(1,1))</code></pre>
<p>Known factors deteriorate the performance of t-SNE.</p>
</div>
<div id="run-cellview-to-cluster-islet-cells" class="section level2">
<h2>Run CellView to cluster islet cells</h2>
<pre class="r"><code># specify gene number to select for
gene_num &lt;- 1000

# calcuclate dispersion
row.var &lt;- apply(lcounts,1,sd)**2
row.mean &lt;- apply(lcounts,1,mean)
dispersion &lt;- row.var/row.mean

# generate sequence of bins
bins &lt;- seq(from = min(row.mean), to = max(row.mean), length.out = 20)

# sort mean expression data into the bins
bin.exp &lt;- row.mean
# sort the values
bin.sort &lt;- sort(bin.exp, decreasing = FALSE)
# vector of bin assignment
cuts &lt;- cut(x = bin.exp, breaks = bins, labels = FALSE)
# find which are NA and change to zero
na.ids &lt;- which(is.na(cuts) == TRUE)
cuts[na.ids] &lt;- 0

# create an empty vector for overdispersion
overdispersion &lt;- NULL

# for each gene and bin index, calculate median, mad, and then normalized dispersion
# first loop through length of bins found
for (k in 1:length(names(table(cuts)))) {
  # find index of bins
  bin.id &lt;- which(cuts == names(table(cuts))[k])
  # median of all genes in the bin
  median.bin &lt;- median(dispersion[bin.id], na.rm = TRUE)
  # calculate mad (median absolute deviation)
  mad.bin &lt;- mad(dispersion[bin.id])
  # calculate norm dispersion for each gene
  for (m in 1:length(bin.id)) {
    norm.dispersion &lt;- abs(dispersion[bin.id[m]] - median.bin)/mad.bin
    overdispersion &lt;- c(overdispersion, norm.dispersion) 
  }
}

# remove nans 
overdis.na &lt;- which(is.na(overdispersion) == TRUE)
if (length(overdis.na) &gt; 0) {
  overdis.filt &lt;- overdispersion[-overdis.na]
} else {
  overdis.filt &lt;- overdispersion
}


# plot mean expression vs overdisperssion
ids &lt;- which(names(overdis.filt) %in% names(row.mean))
plot(row.mean[ids], overdis.filt)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_CellView-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Do t-sne using top over-dispersed genes (apply mean expression filter too)
rank.ov &lt;- order(overdis.filt, decreasing = TRUE)
ov.genes &lt;- names(overdis.filt[rank.ov[1:gene_num]])
log.sel &lt;- lcounts[ov.genes,]

all1 &lt;- t(log.sel)
# Remove groups that are all zeros
df &lt;- all1[,apply(all1, 2, var, na.rm=TRUE) != 0]

set.seed(45344)
rtsne_out &lt;- Rtsne(as.matrix(df), dims = 3)

# Set rownames of matrix to tsne matrix
rownames(rtsne_out$Y) &lt;- rownames(df)

plot(rtsne_out$Y[,c(1,2)], main=&quot;CellView&quot;, xlab=&quot;Dimension 1&quot;, 
     ylab=&quot;Dimension 2&quot;,pch=21, col=color.vec[Cell_Type], 
     bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend(&quot;topright&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_CellView-2.png" width="672" style="display: block; margin: auto;" /> ## Run Seurat to cluster islet cells</p>
<pre class="r"><code>set.seed(12344)
seurat.obj &lt;- CreateSeuratObject(raw.data=counts, 
                                 min.cells=3, min.genes=200, project=&quot;Seurat_Comp&quot;)

names(Patient_ID) &lt;- rownames(seurat.obj@meta.data)
seurat.obj &lt;- AddMetaData(object = seurat.obj, 
                          metadata = Patient_ID, col.name = &quot;patient.id&quot;)

# Normalizing the data
seurat.obj &lt;- NormalizeData(object = seurat.obj, normalization.method = &quot;LogNormalize&quot;, 
    scale.factor = 10000)

# Detection of variable genes across the single cells
seurat.obj &lt;- FindVariableGenes(object = seurat.obj, mean.function = ExpMean, dispersion.function = LogVMR, 
    x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_Seurat-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>length(x = seurat.obj@var.genes) #6639</code></pre>
<pre><code>[1] 6639</code></pre>
<pre class="r"><code># Scaling the data and removing unwanted sources of variation
seurat.obj &lt;- ScaleData(object = seurat.obj, vars.to.regress = c(&quot;nGene&quot;, &quot;patient.id&quot;))</code></pre>
<pre><code>Regressing out: nGene, patient.id</code></pre>
<pre><code>
Time Elapsed:  31.3347609043121 secs</code></pre>
<pre><code>Scaling data matrix</code></pre>
<pre class="r"><code># Perform linear dimensional reduction
seurat.obj &lt;- RunPCA(object = seurat.obj, pc.genes = seurat.obj@var.genes, do.print = TRUE, pcs.print = 1:5, 
    genes.print = 5)</code></pre>
<pre><code>[1] &quot;PC1&quot;
[1] &quot;TM4SF4&quot; &quot;GC&quot;     &quot;GLS&quot;    &quot;FAP&quot;    &quot;PAPPA2&quot;
[1] &quot;&quot;
[1] &quot;HADH&quot;    &quot;TIMP2&quot;   &quot;SYNE2&quot;   &quot;ADCYAP1&quot; &quot;PCSK1&quot;  
[1] &quot;&quot;
[1] &quot;&quot;
[1] &quot;PC2&quot;
[1] &quot;ADCYAP1&quot; &quot;HADH&quot;    &quot;ERO1LB&quot;  &quot;PFKFB2&quot;  &quot;NPTX2&quot;  
[1] &quot;&quot;
[1] &quot;IFITM3&quot; &quot;SPARC&quot;  &quot;LGALS1&quot; &quot;COL1A2&quot; &quot;COL1A1&quot;
[1] &quot;&quot;
[1] &quot;&quot;
[1] &quot;PC3&quot;
[1] &quot;COL1A2&quot;  &quot;COL3A1&quot;  &quot;BGN&quot;     &quot;SPARC&quot;   &quot;COL15A1&quot;
[1] &quot;&quot;
[1] &quot;CFTR&quot;     &quot;MMP7&quot;     &quot;LCN2&quot;     &quot;KRT19&quot;    &quot;PDZK1IP1&quot;
[1] &quot;&quot;
[1] &quot;&quot;
[1] &quot;PC4&quot;
[1] &quot;SLITRK6&quot; &quot;CALCRL&quot;  &quot;THSD7A&quot;  &quot;CALB1&quot;   &quot;ETV1&quot;   
[1] &quot;&quot;
[1] &quot;PFKFB2&quot; &quot;ANXA2&quot;  &quot;ROBO2&quot;  &quot;RIN2&quot;   &quot;ENO1&quot;  
[1] &quot;&quot;
[1] &quot;&quot;
[1] &quot;PC5&quot;
[1] &quot;ETV1&quot;    &quot;CALB1&quot;   &quot;SLITRK6&quot; &quot;SPOCK1&quot;  &quot;THSD7A&quot; 
[1] &quot;&quot;
[1] &quot;PODXL&quot; &quot;CD93&quot;  &quot;FLT1&quot;  &quot;PLVAP&quot; &quot;ESAM&quot; 
[1] &quot;&quot;
[1] &quot;&quot;</code></pre>
<pre class="r"><code># Determine statistically significant principal components
if(FALSE){
seurat.obj &lt;- JackStraw(object = seurat.obj, num.replicate = 100, do.print = FALSE)
JackStrawPlot(object = seurat.obj, PCs = 1:20)
PCElbowPlot(object = seurat.obj)
# Cluster the cells
seurat.obj &lt;- FindClusters(object = seurat.obj, reduction.type = &quot;pca&quot;, dims.use = 1:10, 
    resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = seurat.obj)
}

set.seed(12344454)
seurat.obj &lt;- RunTSNE(object = seurat.obj, dims.use = 1:10, do.fast = TRUE)
TSNEPlot(object = seurat.obj)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_Seurat-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># with true cell types
plot(seurat.obj@dr$tsne@cell.embeddings[,c(1,2)], 
     main=&quot;Spectral tSNE (Seurat)&quot;, xlab=&quot;Dimension 1&quot;, 
     ylab=&quot;Dimension 2&quot;,pch=21, col=color.vec[Cell_Type], 
     bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend(&quot;topright&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_Seurat-3.png" width="672" style="display: block; margin: auto;" /> ## Run PCA to cluster islet cells. Here, we applied PCA on read counts of all genes. Genes are color coded with their expression of marker genes.</p>
<pre class="r"><code>pairs(pca.res[,1:4], main=&quot;PCA&quot;, pch=21, col=color.vec[Cell_Type],
      bg=color.vec[Cell_Type], cex=0.8, oma=c(4,4,6,12))
legend(&quot;right&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_pca-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="run-surrogate-variable-analysis-sva." class="section level2">
<h2>Run surrogate variable analysis (SVA).</h2>
<p>Here, for comparison purposes we conduct SVA on our example data while adjusting for Patient_ID and Geo_Lib_Size and obtained 4 hidden factors.</p>
<pre class="r"><code>set.seed(34544455)
mod1 &lt;- model.matrix(~Patient_ID+Num_Detected_Genes)
mod0 &lt;- cbind(mod1[,1])
sva.res = svaseq(counts,mod1,mod0, n.sv=4)$sv</code></pre>
<pre><code>Number of significant surrogate variables is:  4 
Iteration (out of 5 ):1  2  3  4  5  </code></pre>
<pre class="r"><code>pairs(sva.res[,1:4], main=&quot;SVA&quot;, pch=21, col=color.vec[Cell_Type], 
      bg=color.vec[Cell_Type], cex=0.8, oma=c(4,4,6,12))
legend(&quot;right&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_sva-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="run-ia-sva" class="section level2">
<h2>Run IA-SVA</h2>
<p>Here, we run IA-SVA using Patient_ID and Geo_Lib_Size as known factors and identify 4 hidden factors.</p>
<pre class="r"><code>set.seed(345466666)
mod &lt;- model.matrix(~Patient_ID+Num_Detected_Genes)
summ_exp &lt;- SummarizedExperiment(assays = counts)
iasva.res&lt;- iasva(summ_exp, mod[,-1],verbose=FALSE, permute=FALSE, num.sv=4)</code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
SV 2 Detected!</code></pre>
<pre><code>
SV 3 Detected!</code></pre>
<pre><code>
SV 4 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 4</code></pre>
<pre class="r"><code>iasva.sv &lt;- iasva.res$sv

## no color
pairs(iasva.sv[,1:4], pch=21, col=&quot;black&quot;, bg=&quot;black&quot;, cex=0.8)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_iasva-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## with color-coding
pairs(iasva.sv[,1:4], main=&quot;IA-SVA&quot;, pch=21, col=color.vec[Cell_Type],
      bg=color.vec[Cell_Type], cex=0.8, oma=c(4,4,6,12))
legend(&quot;right&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_iasva-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="correlation-between-svs" class="section level2">
<h2>Correlation between SVs</h2>
<pre class="r"><code>cor(iasva.sv)</code></pre>
<pre><code>           SV1         SV2         SV3        SV4
SV1 1.00000000  0.03808617  0.06200799  0.2507764
SV2 0.03808617  1.00000000 -0.15724552 -0.1830763
SV3 0.06200799 -0.15724552  1.00000000  0.2474547
SV4 0.25077642 -0.18307630  0.24745468  1.0000000</code></pre>
<pre class="r"><code>corrplot(cor(iasva.sv))</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/corr_btw_SVs-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="clustering-analyses" class="section level2">
<h2>clustering analyses</h2>
<p>Here, try different R2 cutoffs</p>
<pre class="r"><code>no.clusters &lt;- 4
C.scores &lt;- matrix(0,0,0)
Number.of.genes &lt;- matrix(0,0,0)
for (i in seq(0.1,0.9,0.05)){
  marker.counts &lt;- find_markers(summ_exp, as.matrix(iasva.sv[,2]), rsq.cutoff = i)
  no.genes &lt;- dim(marker.counts)[1]
  if(no.genes == 0){
    break
  }
  else{
    my.dist &lt;- dist(t(log(marker.counts+1)))
    my.clustering &lt;- hclust(my.dist, method = &quot;ward.D2&quot;)
    my.silhoutte &lt;-silhouette(cutree(my.clustering,no.clusters),my.dist)
    C1 &lt;- mean(my.silhoutte[my.silhoutte[,1]==1,3])
    C2 &lt;- mean(my.silhoutte[my.silhoutte[,1]==2,3])
    average.C &lt;- (C1+C2)/2
    C.scores &lt;- c(C.scores, average.C)
    Number.of.genes &lt;- c(Number.of.genes,no.genes)
  }
}</code></pre>
<pre><code># of markers (): 2204</code></pre>
<pre><code>total # of unique markers: 2204</code></pre>
<pre><code># of markers (): 1326</code></pre>
<pre><code>total # of unique markers: 1326</code></pre>
<pre><code># of markers (): 839</code></pre>
<pre><code>total # of unique markers: 839</code></pre>
<pre><code># of markers (): 528</code></pre>
<pre><code>total # of unique markers: 528</code></pre>
<pre><code># of markers (): 329</code></pre>
<pre><code>total # of unique markers: 329</code></pre>
<pre><code># of markers (): 190</code></pre>
<pre><code>total # of unique markers: 190</code></pre>
<pre><code># of markers (): 111</code></pre>
<pre><code>total # of unique markers: 111</code></pre>
<pre><code># of markers (): 57</code></pre>
<pre><code>total # of unique markers: 57</code></pre>
<pre><code># of markers (): 30</code></pre>
<pre><code>total # of unique markers: 30</code></pre>
<pre><code># of markers (): 15</code></pre>
<pre><code>total # of unique markers: 15</code></pre>
<pre><code># of markers (): 2</code></pre>
<pre><code>total # of unique markers: 2</code></pre>
<pre><code># of markers (): 0</code></pre>
<pre><code>total # of unique markers: 0</code></pre>
<pre class="r"><code>output.matrix &lt;- rbind(C.scores, Number.of.genes)


pdf(&quot;output/Clustering_analyses_figure4_Xin.pdf&quot;)
par(mar=c(5,5,5,5))
end.point &lt;- (length(C.scores)-1)*0.05+0.1
plot(Number.of.genes,  xlab = &quot;R^2&quot;, ylab = &quot;Number genes selected&quot;, xaxt=&quot;n&quot;, main = &quot;Number of selected genes vs. Cluster quality&quot;, pch = 18, col =&quot;blue&quot;,type=&quot;b&quot;, lty=2, cex=2)
Axis(1,at=seq(1,length(Number.of.genes)), side = 1, labels=seq(0.1,end.point,0.05),las = 2)
par(new=T)
plot(C.scores, xlab=&#39;&#39;, ylab=&#39;&#39;, axes=F, pch = 18 , col =&quot;red&quot;,type=&quot;b&quot;, lty=2, cex=2)
Axis(side=4)
mtext(side = 4, line = 2, &#39;Average Silhoutte Score&#39;, col = &quot;red&quot;)
par(new=F)
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="find-marker-genes-for-sv1." class="section level2">
<h2>Find marker genes for SV1.</h2>
<pre class="r"><code># try different R2 thresholds
pdf(&quot;output/Clustering_analyses_figure4_Xin_sv1.pdf&quot;)
r2.results &lt;- study_R2(summ_exp, iasva.sv,selected.svs=1, no.clusters=2)</code></pre>
<pre><code># of markers (): 369</code></pre>
<pre><code>total # of unique markers: 369</code></pre>
<pre><code># of markers (): 198</code></pre>
<pre><code>total # of unique markers: 198</code></pre>
<pre><code># of markers (): 123</code></pre>
<pre><code>total # of unique markers: 123</code></pre>
<pre><code># of markers (): 87</code></pre>
<pre><code>total # of unique markers: 87</code></pre>
<pre><code># of markers (): 57</code></pre>
<pre><code>total # of unique markers: 57</code></pre>
<pre><code># of markers (): 44</code></pre>
<pre><code>total # of unique markers: 44</code></pre>
<pre><code># of markers (): 37</code></pre>
<pre><code>total # of unique markers: 37</code></pre>
<pre><code># of markers (): 26</code></pre>
<pre><code>total # of unique markers: 26</code></pre>
<pre><code># of markers (): 20</code></pre>
<pre><code>total # of unique markers: 20</code></pre>
<pre><code># of markers (): 14</code></pre>
<pre><code>total # of unique markers: 14</code></pre>
<pre><code># of markers (): 12</code></pre>
<pre><code>total # of unique markers: 12</code></pre>
<pre><code># of markers (): 8</code></pre>
<pre><code>total # of unique markers: 8</code></pre>
<pre><code># of markers (): 7</code></pre>
<pre><code>total # of unique markers: 7</code></pre>
<pre><code># of markers (): 5</code></pre>
<pre><code>total # of unique markers: 5</code></pre>
<pre><code># of markers (): 4</code></pre>
<pre><code>total # of unique markers: 4</code></pre>
<pre><code># of markers (): 1</code></pre>
<pre><code>total # of unique markers: 1</code></pre>
<pre><code># of markers (): 0</code></pre>
<pre><code>total # of unique markers: 0</code></pre>
<pre class="r"><code>dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>marker.counts &lt;- find_markers(summ_exp, as.matrix(iasva.sv[,c(1)]),  rsq.cutoff = 0.3)</code></pre>
<pre><code># of markers (): 57</code></pre>
<pre><code>total # of unique markers: 57</code></pre>
<pre class="r"><code>nrow(marker.counts)</code></pre>
<pre><code>[1] 57</code></pre>
<pre class="r"><code>marker.counts.long &lt;- find_markers(summ_exp, as.matrix(iasva.sv[,c(1)]),  rsq.cutoff = 0.2)</code></pre>
<pre><code># of markers (): 123</code></pre>
<pre><code>total # of unique markers: 123</code></pre>
<pre class="r"><code>nrow(marker.counts.long) </code></pre>
<pre><code>[1] 123</code></pre>
<pre class="r"><code>anno.col &lt;- data.frame(Cell_Type=Cell_Type)
rownames(anno.col) &lt;- colnames(marker.counts)
head(anno.col)</code></pre>
<pre><code>           Cell_Type
SRR3541303      beta
SRR3541304      beta
SRR3541305      beta
SRR3541306      beta
SRR3541307      beta
SRR3541308      beta</code></pre>
<pre class="r"><code>cell.type.col &lt;- color.vec[1:4]
names(cell.type.col) &lt;- c(&quot;alpha&quot;,&quot;beta&quot;,&quot;delta&quot;,&quot;PP&quot;)
anno.colors &lt;- list(Cell_Type=cell.type.col)

pheatmap(log(marker.counts+1), show_colnames =FALSE, show_rownames = TRUE,
         clustering_method = &quot;ward.D2&quot;,cutree_cols = 4, 
         annotation_col = anno.col, annotation_colors=anno.colors)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/find_markers_SV1-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="find-marker-genes-for-sv3." class="section level2">
<h2>Find marker genes for SV3.</h2>
<pre class="r"><code>marker.counts.SV3 &lt;- find_markers(summ_exp, as.matrix(iasva.sv[,c(3)]),  rsq.cutoff = 0.3)</code></pre>
<pre><code># of markers (): 54</code></pre>
<pre><code>total # of unique markers: 54</code></pre>
<pre class="r"><code>nrow(marker.counts.SV3)</code></pre>
<pre><code>[1] 54</code></pre>
<pre class="r"><code>marker.counts.SV3.long &lt;- find_markers(summ_exp, as.matrix(iasva.sv[,c(3)]),  rsq.cutoff = 0.2)</code></pre>
<pre><code># of markers (): 138</code></pre>
<pre><code>total # of unique markers: 138</code></pre>
<pre class="r"><code>nrow(marker.counts.SV3.long)</code></pre>
<pre><code>[1] 138</code></pre>
<pre class="r"><code>anno.col &lt;- data.frame(Cell_Type=Cell_Type, SV3=iasva.sv[,3])
rownames(anno.col) &lt;- colnames(marker.counts.SV3)
head(anno.col)</code></pre>
<pre><code>           Cell_Type           SV3
SRR3541303      beta  0.0070956379
SRR3541304      beta -0.0006460813
SRR3541305      beta -0.0116911796
SRR3541306      beta -0.0044453589
SRR3541307      beta  0.0008762206
SRR3541308      beta -0.0080419417</code></pre>
<pre class="r"><code>cell.type.col &lt;- color.vec[1:4]
names(cell.type.col) &lt;- c(&quot;alpha&quot;,&quot;beta&quot;,&quot;delta&quot;,&quot;PP&quot;)
anno.colors &lt;- list(Cell_Type=cell.type.col)

pheatmap(log(marker.counts.SV3+1), show_colnames =FALSE, show_rownames = TRUE, clustering_method = &quot;ward.D2&quot;,cutree_cols = 2,annotation_col = anno.col, annotation_colors=anno.colors)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/find_markers_SV3-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="run-tsne-post-ia-sva-i.e.-run-tsne-on-marker-genes-obtained-from-ia-sva." class="section level2">
<h2>Run tSNE post IA-SVA, i.e., run tSNE on marker genes obtained from IA-SVA.</h2>
<pre class="r"><code>set.seed(75458456)
tsne.res.iasva &lt;- Rtsne(unique(t(log(marker.counts+1))), dims = 2)

plot(tsne.res.iasva$Y[,1:2], main=&quot;IA-SVA&quot;, xlab=&quot;tSNE Dim1&quot;,
     ylab=&quot;tSNE Dim2&quot;, pch=21, col=color.vec[Cell_Type], 
     bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend(&quot;topright&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_Xin_Islets.Rmd/run_tsne_post_iasva-1.png" width="672" style="display: block; margin: auto;" /> tSNE performed on marker genes selected via IA-SVA performs better than original tSNE analyses using all genes. This example again reiterates the importance of gene selection using IA-SVA for effective clustering of single-cell datasets.</p>
<pre class="r"><code>pdf(file=&quot;output/Xin_Islets_All_demensionality_reduction_Figure4DEFG.pdf&quot;, width=8, height=9)
par(mfrow=c(2,2))
plot(tsne.res.all$Y, main=&quot;tSNE&quot;, xlab=&quot;Dimension 1&quot;, ylab=&quot;Dimension 2&quot;,pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type])
legend(&quot;topleft&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)
plot(tsne.res.iasva$Y, main=&quot;IA-SVA + tSNE&quot;, xlab=&quot;Dimension 1&quot;, ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type])
plot(rtsne_out$Y[,c(1,2)], main=&quot;CellView&quot;, xlab=&quot;Dimension 1&quot;, ylab=&quot;Dimension 2&quot;,pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type])
plot(seurat.obj@dr$tsne@cell.embeddings[,c(1,2)], main=&quot;Spectral tSNE (Seurat)&quot;, xlab=&quot;Dim1&quot;, ylab=&quot;Dim2&quot;,pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type])
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>pdf(file=&quot;output/Xin_Islets_AllCells_tSNEByKnownFactors_FigureS9.pdf&quot;, width=10, height=11)
par(mfrow=c(2,2))
plot(tsne.res.all$Y, main=&quot;Gender&quot;, xlab=&quot;Dimension 1&quot;, ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Gender], bg=color.vec[Gender], oma=c(4,4,6,12))
legend(&quot;topleft&quot;, levels(Gender), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)
plot(tsne.res.all$Y, main=&quot;Patient ID&quot;, xlab=&quot;Dimension 1&quot;, ylab=&quot;Dimension 2&quot;, pch=21, col=tol21rainbow[Patient_ID], bg=tol21rainbow[Patient_ID], oma=c(4,4,6,12))
legend(&quot;topleft&quot;, levels(Patient_ID), border=&quot;white&quot;, fill=tol21rainbow, bty=&quot;n&quot;, cex=0.5)
plot(tsne.res.all$Y, main=&quot;Ethnicity&quot;, xlab=&quot;Dimension 1&quot;, ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Ethnicity], bg=color.vec[Ethnicity], oma=c(4,4,6,12))
legend(&quot;topleft&quot;, levels(Ethnicity), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)
plot(tsne.res.all$Y, main=&quot;Phenotype&quot;, xlab=&quot;Dimension 1&quot;, ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Phenotype], bg=color.vec[Phenotype], oma=c(4,4,6,12))
legend(&quot;topleft&quot;, levels(Phenotype), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)
par(mfrow=c(1,1))
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>pdf(file=&quot;output/Xin_Islets_AllCells_IASVA_nocolor.pdf&quot;, width=4, height=4)
pairs(iasva.sv[,1:4], pch=21, col=&quot;black&quot;, bg=&quot;black&quot;, cex=0.3)
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>pdf(file=&quot;output/Xin_Islets_AllCells_IASVA.pdf&quot;, width=4, height=4)
pairs(iasva.sv[,1:4], pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], cex=0.3)
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>pdf(file=&quot;output/Xin_Islets_AllCells_PCA.pdf&quot;, width=4, height=4)
pairs(pca.res[,1:4], pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], cex=0.3)#
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>colnames(sva.res) &lt;- paste0(&quot;SV&quot;,1:4)
pdf(file=&quot;output/Xin_Islets_AllCells_USVA.pdf&quot;, width=4, height=4)
pairs(sva.res[,1:4], pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], cex=0.3) #4,4,6,12
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>anno.col &lt;- data.frame(Cell_Type=Cell_Type)
rownames(anno.col) &lt;- colnames(marker.counts)
head(anno.col)</code></pre>
<pre><code>           Cell_Type
SRR3541303      beta
SRR3541304      beta
SRR3541305      beta
SRR3541306      beta
SRR3541307      beta
SRR3541308      beta</code></pre>
<pre class="r"><code>cell.type.col &lt;- color.vec[1:4]
names(cell.type.col) &lt;- c(&quot;alpha&quot;,&quot;beta&quot;,&quot;delta&quot;,&quot;PP&quot;)
anno.colors &lt;- list(Cell_Type=cell.type.col)

pheatmap(log(marker.counts+1), show_colnames =FALSE, show_rownames = TRUE,
         clustering_method = &quot;ward.D2&quot;,cutree_cols = 4,annotation_col = anno.col, annotation_colors=anno.colors,
         filename=&quot;output/FigureS11_Xin_Islets_AllCells_IASVA_Markers_pheatmap.pdf&quot;, width=10, height=10)</code></pre>
<pre class="r"><code>write.table(as.data.frame(rownames(marker.counts)), file=&quot;output/Xin_Islets_AllCells_SV1_Genes_rsqcutoff0.3.txt&quot;, quote=F,
              row.names=F, col.names=F, sep=&quot; &quot;)

write.table(as.data.frame(rownames(marker.counts.long)), file=&quot;output/Xin_Islets_AllCells_SV1_Genes_rsqcutoff0.2.txt&quot;, quote=F,
              row.names=F, col.names=F, sep=&quot; &quot;)</code></pre>
<pre class="r"><code>write.table(as.data.frame(rownames(marker.counts.SV3)), file=&quot;output/Xin_Islets_AllCells_SV3_Genes_rsqcutoff0.3.txt&quot;, quote=F,
              row.names=F, col.names=F, sep=&quot; &quot;)

write.table(as.data.frame(rownames(marker.counts.SV3.long)), file=&quot;output/Xin_Islets_AllCells_SV3_Genes_rsqcutoff0.2.txt&quot;, quote=F,
              row.names=F, col.names=F, sep=&quot; &quot;)</code></pre>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.0 (2018-04-23)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] SummarizedExperiment_1.10.1 DelayedArray_0.6.1         
 [3] matrixStats_0.53.1          Biobase_2.40.0             
 [5] GenomicRanges_1.32.3        GenomeInfoDb_1.16.0        
 [7] IRanges_2.14.10             S4Vectors_0.18.3           
 [9] BiocGenerics_0.26.0         cluster_2.0.7-1            
[11] RColorBrewer_1.1-2          DescTools_0.99.24          
[13] corrplot_0.84               pheatmap_1.0.10            
[15] Rtsne_0.13                  irlba_2.3.2                
[17] dbscan_1.1-2                iasvaExamples_1.0.0        
[19] iasva_0.99.3                Seurat_2.3.2               
[21] Matrix_1.2-14               cowplot_0.9.2              
[23] ggplot2_2.2.1.9000          sva_3.28.0                 
[25] BiocParallel_1.14.2         genefilter_1.62.0          
[27] mgcv_1.8-23                 nlme_3.1-137               

loaded via a namespace (and not attached):
  [1] reticulate_1.8         R.utils_2.6.0          tidyselect_0.2.4      
  [4] RSQLite_2.1.1          AnnotationDbi_1.42.1   htmlwidgets_1.2       
  [7] grid_3.5.0             trimcluster_0.1-2      ranger_0.10.1         
 [10] munsell_0.4.3          codetools_0.2-15       ica_1.0-2             
 [13] withr_2.1.2            colorspace_1.3-2       knitr_1.20            
 [16] rstudioapi_0.7         geometry_0.3-6         ROCR_1.0-7            
 [19] robustbase_0.93-0      dtw_1.20-1             dimRed_0.1.0          
 [22] labeling_0.3           lars_1.2               git2r_0.21.0          
 [25] GenomeInfoDbData_1.1.0 mnormt_1.5-5           bit64_0.9-7           
 [28] rprojroot_1.3-2        ipred_0.9-6            diptest_0.75-7        
 [31] R6_2.2.2               VGAM_1.0-5             hdf5r_1.0.0           
 [34] flexmix_2.3-14         DRR_0.0.3              bitops_1.0-6          
 [37] assertthat_0.2.0       SDMTools_1.1-221       scales_0.5.0          
 [40] nnet_7.3-12            gtable_0.2.0           ddalpha_1.3.3         
 [43] workflowr_1.0.1        timeDate_3043.102      rlang_0.2.1           
 [46] CVST_0.2-2             scatterplot3d_0.3-41   RcppRoll_0.3.0        
 [49] splines_3.5.0          lazyeval_0.2.1         ModelMetrics_1.1.0    
 [52] acepack_1.4.1          broom_0.4.4            checkmate_1.8.5       
 [55] yaml_2.1.19            reshape2_1.4.3         abind_1.4-5           
 [58] backports_1.1.2        Hmisc_4.1-1            caret_6.0-80          
 [61] tools_3.5.0            lava_1.6.1             psych_1.8.4           
 [64] gplots_3.0.1           proxy_0.4-22           ggridges_0.5.0        
 [67] Rcpp_0.12.17           plyr_1.8.4             base64enc_0.1-3       
 [70] zlibbioc_1.26.0        purrr_0.2.5            RCurl_1.95-4.10       
 [73] rpart_4.1-13           pbapply_1.3-4          zoo_1.8-2             
 [76] sfsmisc_1.1-2          magrittr_1.5           data.table_1.11.4     
 [79] manipulate_1.0.1       lmtest_0.9-36          RANN_2.5.1            
 [82] mvtnorm_1.0-8          whisker_0.3-2          fitdistrplus_1.0-9    
 [85] evaluate_0.10.1        xtable_1.8-2           XML_3.98-1.11         
 [88] mclust_5.4             gridExtra_2.3          compiler_3.5.0        
 [91] tibble_1.4.2           KernSmooth_2.23-15     R.oo_1.22.0           
 [94] htmltools_0.3.6        segmented_0.5-3.0      Formula_1.2-3         
 [97] snow_0.4-2             tidyr_0.8.1            expm_0.999-2          
[100] tclust_1.4-1           lubridate_1.7.4        DBI_1.0.0             
[103] diffusionMap_1.1-0     magic_1.5-8            MASS_7.3-50           
[106] fpc_2.1-11             boot_1.3-20            R.methodsS3_1.7.1     
[109] gdata_2.18.0           metap_0.9              bindr_0.1.1           
[112] gower_0.1.2            igraph_1.2.1           pkgconfig_2.0.1       
[115] foreign_0.8-70         recipes_0.1.3          foreach_1.4.4         
[118] annotate_1.58.0        XVector_0.20.0         prodlim_2018.04.18    
[121] stringr_1.3.1          digest_0.6.15          pls_2.6-0             
[124] tsne_0.1-3             rmarkdown_1.9          htmlTable_1.12        
[127] kernlab_0.9-26         gtools_3.5.0           modeltools_0.2-21     
[130] jsonlite_1.5           bindrcpp_0.2.2         limma_3.36.2          
[133] pillar_1.2.3           lattice_0.20-35        DEoptimR_1.0-8        
[136] survival_2.42-3        glue_1.2.0             FNN_1.1               
[139] png_0.1-7              prabclus_2.2-6         iterators_1.0.9       
[142] bit_1.1-14             class_7.3-14           stringi_1.2.2         
[145] mixtools_1.1.0         blob_1.1.1             doSNOW_1.0.16         
[148] latticeExtra_0.6-28    caTools_1.17.1         memoise_1.1.0         
[151] dplyr_0.7.5            ape_5.1               </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1
</p>
<hr>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
