<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Donghyung Lee" />

<meta name="date" content="2018-08-03" />

<title>IA-SVA based feature selection improves the performance of clustering algorithms [1]</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">iasvaExamples</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/dleelab/iasvaExamples">GitHub</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">IA-SVA based feature selection improves the performance of clustering algorithms [1]</h1>
<h4 class="author"><em>Donghyung Lee</em></h4>
<h4 class="date"><em>2018-08-03</em></h4>

</div>

<div id="TOC">
<ul>
<li><a href="#load-packages">Load packages</a></li>
<li><a href="#load-the-islet-single-cell-rna-seq-data">Load the islet single cell RNA-Seq data</a></li>
<li><a href="#extract-three-cell-types-gcg-alpha-ins-beta-krt19-ductal-expressing-cells-from-healthy-i.e.-non-diabetic-subjects-and-filter-out-low-expressed-genes">Extract three cell types (GCG (alpha), INS (beta), KRT19 (ductal) expressing cells) from healthy (i.e., non-diabetic) subjects and filter out low-expressed genes</a></li>
<li><a href="#calculate-the-number-of-detected-genes">Calculate the number of detected genes</a></li>
<li><a href="#run-tsne-to-cluster-islet-cells.">Run tSNE to cluster islet cells.</a></li>
<li><a href="#run-ia-sva">Run IA-SVA</a></li>
<li><a href="#find-marker-genes-for-sv1-and-sv3.">Find marker genes for SV1 and SV3.</a></li>
<li><a href="#find-marker-genes-for-sv4.">Find marker genes for SV4.</a></li>
<li><a href="#run-tsne-post-ia-sva-i.e.-run-tsne-on-marker-genes-for-sv1-and-sv2-obtained-from-ia-sva.">Run tSNE post IA-SVA, i.e., run tSNE on marker genes for SV1 and SV2 obtained from IA-SVA.</a></li>
<li><a href="#run-cellview-algorithm-to-visualize-the-data.">Run CellView algorithm to visualize the data.</a></li>
<li><a href="#run-seurat-to-reduce-dimensionality-and-visualize-islet-cells">Run Seurat to reduce dimensionality and visualize islet cells</a></li>
<li><a href="#session-information">Session information</a></li>
</ul>
</div>

<p><strong>Last updated:</strong> 2018-08-03</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(20180731)</code> </summary></p>
<p>The command <code>set.seed(20180731)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/dleelab/iasvaExamples/tree/268983524d3aee232c1bf032e8dd8993bcfb4fc3" target="_blank">2689835</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    inst/.DS_Store
    Ignored:    inst/doc/.DS_Store
    Ignored:    vignettes/.DS_Store

Untracked files:
    Untracked:  docs/figure/hidden_heterogeneity_glioblastoma.Rmd/
    Untracked:  output/CC_genes.long.txt
    Untracked:  output/CC_genes.short.txt
    Untracked:  output/Clustering_analyses_figure3_sv1.pdf
    Untracked:  output/Patel_Glioblastoma_MGH30_CellCycle_Figure3ABCD.pdf
    Untracked:  output/Patel_Glioblastoma_MGH30_Cellcycle_SV1_Genes_rsqcutoff0.3.txt
    Untracked:  output/Patel_Glioblastoma_MGH30_Cellcycle_SV1_Genes_rsqcutoff0.4.txt
    Untracked:  output/Patel_Glioblastoma_MGH30_iasva_SV1_genes_rsqcutoff0.3_pheatmap_iasvaV0.95_Figure3F.pdf

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/dleelab/iasvaExamples/blob/c241b04dac899407045051a04b7539237f4d183f/analysis/tSNE_post_IA-SVA_3celltypes.Rmd" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/dleelab/iasvaExamples/c241b04dac899407045051a04b7539237f4d183f/docs/tSNE_post_IA-SVA_3celltypes.html" target="_blank">c241b04</a>
</td>
<td style="text-align:left;">
dleelab
</td>
<td style="text-align:left;">
2018-08-03
</td>
<td style="text-align:left;">
added
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<p>The IA-SVA based feature selection can significantly improve the performance and utility of clustering algorithms (e.g., tSNE, hierarchical clustering). To illustrate how the IA-SVA method can be used to improve the performance of clustering algorithms, we used real-world single cell RNA sequencing (scRNA-Seq) data obtained from human pancreatic islet samples (<a href="http://genome.cshlp.org/content/early/2017/01/16/gr.212720.116">Lawlor et. al., 2016</a>). This dataset is included in a R data package (“iasvaExamples”) containing data examples for IA-SVA (<a href="https://github.com/dleelab/iasvaExamples" class="uri">https://github.com/dleelab/iasvaExamples</a>). To install the ‘iasvaExamples’ package, follow the instruction provided in the GitHub page.</p>
<div id="load-packages" class="section level2">
<h2>Load packages</h2>
<pre class="r"><code>rm(list=ls())
library(irlba)</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre class="r"><code>library(iasva)
library(iasvaExamples)
library(Seurat)</code></pre>
<pre><code>Loading required package: ggplot2</code></pre>
<pre><code>Loading required package: cowplot</code></pre>
<pre><code>
Attaching package: &#39;cowplot&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:ggplot2&#39;:

    ggsave</code></pre>
<pre class="r"><code>library(dbscan)
library(Rtsne)
library(pheatmap)
library(corrplot)</code></pre>
<pre><code>corrplot 0.84 loaded</code></pre>
<pre class="r"><code>library(DescTools) #pcc i.e., Pearson&#39;s contingency coefficient
library(RColorBrewer)
library(SummarizedExperiment)</code></pre>
<pre><code>Loading required package: GenomicRanges</code></pre>
<pre><code>Loading required package: stats4</code></pre>
<pre><code>Loading required package: BiocGenerics</code></pre>
<pre><code>Loading required package: parallel</code></pre>
<pre><code>
Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:parallel&#39;:

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>The following objects are masked from &#39;package:Matrix&#39;:

    colMeans, colSums, rowMeans, rowSums, which</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, append, as.data.frame, basename, cbind,
    colMeans, colnames, colSums, dirname, do.call, duplicated,
    eval, evalq, Filter, Find, get, grep, grepl, intersect,
    is.unsorted, lapply, lengths, Map, mapply, match, mget, order,
    paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,
    Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which, which.max,
    which.min</code></pre>
<pre><code>Loading required package: S4Vectors</code></pre>
<pre><code>
Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:Matrix&#39;:

    expand</code></pre>
<pre><code>The following object is masked from &#39;package:base&#39;:

    expand.grid</code></pre>
<pre><code>Loading required package: IRanges</code></pre>
<pre><code>Loading required package: GenomeInfoDb</code></pre>
<pre><code>Loading required package: Biobase</code></pre>
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>Loading required package: DelayedArray</code></pre>
<pre><code>Loading required package: matrixStats</code></pre>
<pre><code>
Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:Biobase&#39;:

    anyMissing, rowMedians</code></pre>
<pre><code>Loading required package: BiocParallel</code></pre>
<pre><code>
Attaching package: &#39;DelayedArray&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:matrixStats&#39;:

    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    aperm, apply</code></pre>
<pre class="r"><code>color.vec &lt;- brewer.pal(8, &quot;Set1&quot;)
# Normalization.
normalize &lt;- function (counts) 
{
    normfactor &lt;- colSums(counts)
    return(t(t(counts)/normfactor)*median(normfactor))
}</code></pre>
</div>
<div id="load-the-islet-single-cell-rna-seq-data" class="section level2">
<h2>Load the islet single cell RNA-Seq data</h2>
<pre class="r"><code>data(&quot;Lawlor_Islet_scRNAseq_Read_Counts&quot;)
data(&quot;Lawlor_Islet_scRNAseq_Annotations&quot;)
ls()</code></pre>
<pre><code>[1] &quot;color.vec&quot;                         &quot;Lawlor_Islet_scRNAseq_Annotations&quot;
[3] &quot;Lawlor_Islet_scRNAseq_Read_Counts&quot; &quot;normalize&quot;                        </code></pre>
<pre class="r"><code>counts &lt;- Lawlor_Islet_scRNAseq_Read_Counts
anns &lt;- Lawlor_Islet_scRNAseq_Annotations
dim(anns)</code></pre>
<pre><code>[1] 638  26</code></pre>
<pre class="r"><code>dim(counts)</code></pre>
<pre><code>[1] 26542   638</code></pre>
<pre class="r"><code>summary(anns)</code></pre>
<pre><code>     run             cell.type             COL1A1          INS       
 Length:638         Length:638         Min.   :1.00   Min.   :1.000  
 Class :character   Class :character   1st Qu.:1.00   1st Qu.:1.000  
 Mode  :character   Mode  :character   Median :1.00   Median :1.000  
                                       Mean   :1.03   Mean   :1.414  
                                       3rd Qu.:1.00   3rd Qu.:2.000  
                                       Max.   :2.00   Max.   :2.000  
                                                                     
     PRSS1            SST             GCG            KRT19      
 Min.   :1.000   Min.   :1.000   Min.   :1.000   Min.   :1.000  
 1st Qu.:1.000   1st Qu.:1.000   1st Qu.:1.000   1st Qu.:1.000  
 Median :1.000   Median :1.000   Median :1.000   Median :1.000  
 Mean   :1.038   Mean   :1.039   Mean   :1.375   Mean   :1.044  
 3rd Qu.:1.000   3rd Qu.:1.000   3rd Qu.:2.000   3rd Qu.:1.000  
 Max.   :2.000   Max.   :2.000   Max.   :2.000   Max.   :2.000  
                                                                
      PPY          num.genes             Cell_ID        UNOS_ID   
 Min.   :1.000   Min.   :3529   10th_C1_S59  :  1   ACCG268 :136  
 1st Qu.:1.000   1st Qu.:5270   10th_C10_S104:  1   ACJV399 :108  
 Median :1.000   Median :6009   10th_C11_S96 :  1   ACEL337 :103  
 Mean   :1.028   Mean   :5981   10th_C13_S61 :  1   ACIW009 : 93  
 3rd Qu.:1.000   3rd Qu.:6676   10th_C14_S53 :  1   ACCR015A: 57  
 Max.   :2.000   Max.   :8451   10th_C16_S105:  1   ACIB065 : 57  
                                (Other)      :632   (Other) : 84  
      Age        Biomaterial_Provider    Gender              Phenotype  
 Min.   :22.00   IIDP      : 45       Female:303   Non-Diabetic   :380  
 1st Qu.:29.00   Prodo Labs:593       Male  :335   Type 2 Diabetic:258  
 Median :42.00                                                          
 Mean   :39.63                                                          
 3rd Qu.:53.00                                                          
 Max.   :56.00                                                          
                                                                        
               Race          BMI          Cell_Type     Patient_ID 
 African American:175   Min.   :22.00   INS    :264   P1     :136  
 Hispanic        :165   1st Qu.:26.60   GCG    :239   P8     :108  
 White           :298   Median :32.95   KRT19  : 28   P3     :103  
                        Mean   :32.85   SST    : 25   P7     : 93  
                        3rd Qu.:35.80   PRSS1  : 24   P5     : 57  
                        Max.   :55.00   none   : 21   P6     : 57  
                                        (Other): 37   (Other): 84  
 Sequencing_Run Batch       Coverage       Percent_Aligned 
 12t    : 57    B1:193   Min.   :1206135   Min.   :0.8416  
 4th    : 57    B2:148   1st Qu.:2431604   1st Qu.:0.8769  
 9th    : 57    B3:297   Median :3042800   Median :0.8898  
 10t    : 56             Mean   :3160508   Mean   :0.8933  
 7th    : 55             3rd Qu.:3871697   3rd Qu.:0.9067  
 3rd    : 53             Max.   :5931638   Max.   :0.9604  
 (Other):303                                               
 Mitochondrial_Fraction Num_Expressed_Genes
 Min.   :0.003873       Min.   :3529       
 1st Qu.:0.050238       1st Qu.:5270       
 Median :0.091907       Median :6009       
 Mean   :0.108467       Mean   :5981       
 3rd Qu.:0.142791       3rd Qu.:6676       
 Max.   :0.722345       Max.   :8451       
                                           </code></pre>
<pre class="r"><code>ContCoef(table(anns$Gender, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.225969</code></pre>
<pre class="r"><code>ContCoef(table(anns$Phenotype, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.1145096</code></pre>
<pre class="r"><code>ContCoef(table(anns$Race, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.3084146</code></pre>
<pre class="r"><code>ContCoef(table(anns$Patient_ID, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.5232058</code></pre>
<pre class="r"><code>ContCoef(table(anns$Batch, anns$Cell_Type))</code></pre>
<pre><code>[1] 0.3295619</code></pre>
<p>The annotations describing the islet samples and experimental settings are stored in “anns” and the read counts information is stored in “counts”.</p>
</div>
<div id="extract-three-cell-types-gcg-alpha-ins-beta-krt19-ductal-expressing-cells-from-healthy-i.e.-non-diabetic-subjects-and-filter-out-low-expressed-genes" class="section level2">
<h2>Extract three cell types (GCG (alpha), INS (beta), KRT19 (ductal) expressing cells) from healthy (i.e., non-diabetic) subjects and filter out low-expressed genes</h2>
<pre class="r"><code>counts &lt;- counts[, (anns$Phenotype!=&quot;Non-Diabetic&quot;)&amp;
                    ((anns$Cell_Type==&quot;GCG&quot;)|
                      (anns$Cell_Type==&quot;INS&quot;)|
                        (anns$Cell_Type==&quot;KRT19&quot;))]
anns &lt;- subset(anns, (Phenotype!=&quot;Non-Diabetic&quot;)&amp; 
                      ((Cell_Type==&quot;GCG&quot;)|
                        (Cell_Type==&quot;INS&quot;)|
                          (Cell_Type==&quot;KRT19&quot;)))
dim(counts)</code></pre>
<p>[1] 26542 213</p>
<pre class="r"><code>dim(anns)</code></pre>
<p>[1] 213 26</p>
<pre class="r"><code>anns &lt;- droplevels(anns)
prop.zeros &lt;- sum(counts==0)/length(counts)
prop.zeros</code></pre>
<p>[1] 0.6917875</p>
<pre class="r"><code>filter = apply(counts, 1, function(x) length(x[x&gt;5])&gt;=3)
counts = counts[filter,]
dim(counts)</code></pre>
<p>[1] 16005 213</p>
<pre class="r"><code>prop.zeros &lt;- sum(counts==0)/length(counts)
prop.zeros</code></pre>
<p>[1] 0.4956072</p>
<pre class="r"><code>Patient_ID &lt;- anns$Patient_ID
Cell_Type &lt;- anns$Cell_Type
levels(Cell_Type) &lt;- c(&quot;alpha&quot;, &quot;beta&quot;, &quot;ductal&quot;)
Batch &lt;- anns$Batch
table(Cell_Type)</code></pre>
<p>Cell_Type alpha beta ductal 101 96 16</p>
<pre class="r"><code>raw.counts &lt;- counts
summary(colSums(counts))</code></pre>
<p>Min. 1st Qu. Median Mean 3rd Qu. Max. 473182 1014530 1263975 1310323 1492478 3090459</p>
<pre class="r"><code>counts &lt;- normalize(counts)
summary(colSums(counts))</code></pre>
<p>Min. 1st Qu. Median Mean 3rd Qu. Max. 1263975 1263975 1263975 1263975 1263975 1263975</p>
</div>
<div id="calculate-the-number-of-detected-genes" class="section level2">
<h2>Calculate the number of detected genes</h2>
<p>It is well known that the number of detected genes in each cell explains a very large portion of variability in scRNA-Seq data (<a href="http://biorxiv.org/content/early/2015/08/25/025528">Hicks et. al. 2015 BioRxiv</a>, <a href="http://www.nature.com/nbt/journal/v34/n6/full/nbt.3498.html">McDavid et. al. 2016 Nature Biotechnology</a>). Frequently, the first principal component of log-transformed scRNA-Seq read counts is highly correlated with the number of detected genes (e.g., r &gt; 0.9). Here, we calculate the number of detected genes for islet cells, which will be used as an known factor in the IA-SVA analyses.</p>
<pre class="r"><code>Num_Detected_Genes &lt;- colSums(counts&gt;0)
Geo_Lib &lt;- colSums(log(counts+1))
summary(Geo_Lib)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  16331   21429   25600   25262   28659   34613 </code></pre>
<pre class="r"><code>barplot(Geo_Lib, xlab=&quot;Cell&quot;, las=2, ylab = &quot;Geometric Library Size&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/num_detected_genes-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>lcounts &lt;- log(counts + 1)
# PC1 and Geometric library size correlation
pc1 = irlba(lcounts - rowMeans(lcounts), 1)$v[,1] ## partial SVD
cor(Geo_Lib, pc1)</code></pre>
<pre><code>[1] -0.9796015</code></pre>
</div>
<div id="run-tsne-to-cluster-islet-cells." class="section level2">
<h2>Run tSNE to cluster islet cells.</h2>
<p>For comparison purposes, we applied tSNE on read counts of all genes. We used the Rtsne R package with default settings for this analyses. Genes are colored with respect to the expression of marker genes.</p>
<pre class="r"><code>set.seed(32354388)
tsne.res &lt;- Rtsne(t(lcounts), dims = 2)
plot(tsne.res$Y, main=&quot;tSNE&quot;, xlab=&quot;Dim1&quot;, ylab=&quot;Dim2&quot;, pch=21, 
     col=color.vec[Cell_Type], bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend(&quot;bottomright&quot;, levels(Cell_Type), border=&quot;white&quot;,fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/run_tsne-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="run-ia-sva" class="section level2">
<h2>Run IA-SVA</h2>
<p>Here, we first run IA-SVA using Patient_ID, Batch and Geo_Lib_Size as known factors and identify 5 hidden factors. Since cell type is not used as a known factor in this analyses, IA-SVA will detect the heterogeneity associated with the cell types.</p>
<pre class="r"><code>mod &lt;- model.matrix(~Patient_ID+Batch+Geo_Lib)
summ_exp &lt;- SummarizedExperiment(assays = counts)
iasva.res&lt;- iasva(summ_exp, mod[,-1],verbose=FALSE, permute=FALSE, num.sv=5)</code></pre>
<pre><code>IA-SVA running...</code></pre>
<pre><code>
SV 1 Detected!</code></pre>
<pre><code>
SV 2 Detected!</code></pre>
<pre><code>
SV 3 Detected!</code></pre>
<pre><code>
SV 4 Detected!</code></pre>
<pre><code>
SV 5 Detected!</code></pre>
<pre><code>
# of significant surrogate variables: 5</code></pre>
<pre class="r"><code>iasva.sv &lt;- iasva.res$sv

#with color-coding based on true cell-type
pairs(iasva.sv, main=&quot;IA-SVA&quot;, pch=21, col=color.vec[Cell_Type],
      bg=color.vec[Cell_Type], oma=c(4,4,6,14))
legend(&quot;right&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/run_iasva-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>cor(Num_Detected_Genes, iasva.sv)</code></pre>
<pre><code>             SV1        SV2         SV3        SV4       SV5
[1,] -0.06492242 -0.6767266 -0.09183223 -0.1677585 0.7053465</code></pre>
<pre class="r"><code>cor(Geo_Lib, iasva.sv)</code></pre>
<pre><code>            SV1        SV2       SV3        SV4       SV5
[1,] -0.2044236 -0.7212868 -0.275179 -0.1841835 0.7725583</code></pre>
<pre class="r"><code>corrplot(cor(iasva.sv))</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/run_iasva-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="find-marker-genes-for-sv1-and-sv3." class="section level2">
<h2>Find marker genes for SV1 and SV3.</h2>
<p>Here, using the find_markers() function we find marker genes significantly associated with SV1 and SV3 (multiple testing adjusted p-value &lt; 0.05, default significance cutoff, a high R-squared value: R-squared &gt; 0.4).</p>
<pre class="r"><code># try different R2 thresholds
pdf(&quot;Clustering_analyses_figure4_islets_sv1_3.pdf&quot;)
r2.results &lt;- study_R2(summ_exp, iasva.sv,selected.svs=c(1,3), no.clusters=3)</code></pre>
<pre><code># of markers (SV1): 682</code></pre>
<pre><code># of markers (SV3): 534</code></pre>
<pre><code>total # of unique markers: 979</code></pre>
<pre><code># of markers (SV1): 357</code></pre>
<pre><code># of markers (SV3): 281</code></pre>
<pre><code>total # of unique markers: 519</code></pre>
<pre><code># of markers (SV1): 207</code></pre>
<pre><code># of markers (SV3): 161</code></pre>
<pre><code>total # of unique markers: 309</code></pre>
<pre><code># of markers (SV1): 123</code></pre>
<pre><code># of markers (SV3): 99</code></pre>
<pre><code>total # of unique markers: 192</code></pre>
<pre><code># of markers (SV1): 78</code></pre>
<pre><code># of markers (SV3): 62</code></pre>
<pre><code>total # of unique markers: 129</code></pre>
<pre><code># of markers (SV1): 44</code></pre>
<pre><code># of markers (SV3): 37</code></pre>
<pre><code>total # of unique markers: 79</code></pre>
<pre><code># of markers (SV1): 34</code></pre>
<pre><code># of markers (SV3): 26</code></pre>
<pre><code>total # of unique markers: 58</code></pre>
<pre><code># of markers (SV1): 25</code></pre>
<pre><code># of markers (SV3): 20</code></pre>
<pre><code>total # of unique markers: 44</code></pre>
<pre><code># of markers (SV1): 19</code></pre>
<pre><code># of markers (SV3): 14</code></pre>
<pre><code>total # of unique markers: 33</code></pre>
<pre><code># of markers (SV1): 11</code></pre>
<pre><code># of markers (SV3): 9</code></pre>
<pre><code>total # of unique markers: 20</code></pre>
<pre><code># of markers (SV1): 8</code></pre>
<pre><code># of markers (SV3): 7</code></pre>
<pre><code>total # of unique markers: 15</code></pre>
<pre><code># of markers (SV1): 4</code></pre>
<pre><code># of markers (SV3): 7</code></pre>
<pre><code>total # of unique markers: 11</code></pre>
<pre><code># of markers (SV1): 2</code></pre>
<pre><code># of markers (SV3): 5</code></pre>
<pre><code>total # of unique markers: 7</code></pre>
<pre><code># of markers (SV1): 2</code></pre>
<pre><code># of markers (SV3): 3</code></pre>
<pre><code>total # of unique markers: 5</code></pre>
<pre><code># of markers (SV1): 1</code></pre>
<pre><code># of markers (SV3): 3</code></pre>
<pre><code>total # of unique markers: 4</code></pre>
<pre><code># of markers (SV1): 0</code></pre>
<pre><code># of markers (SV3): 1</code></pre>
<pre><code>total # of unique markers: 1</code></pre>
<pre><code># of markers (SV1): 0</code></pre>
<pre><code># of markers (SV3): 0</code></pre>
<pre><code>total # of unique markers: 0</code></pre>
<pre class="r"><code>dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>marker.counts &lt;- find_markers(summ_exp, as.matrix(iasva.sv[,c(1,3)])
                              , rsq.cutoff = 0.4)</code></pre>
<pre><code># of markers (SV1): 34</code></pre>
<pre><code># of markers (SV3): 26</code></pre>
<pre><code>total # of unique markers: 58</code></pre>
<pre class="r"><code>nrow(marker.counts)</code></pre>
<pre><code>[1] 58</code></pre>
<pre class="r"><code>anno.col &lt;- data.frame(Cell_Type=Cell_Type)
rownames(anno.col) &lt;- colnames(marker.counts)
head(anno.col)</code></pre>
<pre><code>            Cell_Type
4th-C63_S30     alpha
4th-C66_S36     alpha
4th-C7_S15       beta
4th-C18_S31     alpha
4th-C5_S8        beta
4th-C57_S18     alpha</code></pre>
<pre class="r"><code>cell.type.col &lt;- color.vec[1:3]
names(cell.type.col) &lt;- c(&quot;alpha&quot;,&quot;beta&quot;,&quot;ductal&quot;)
anno.colors &lt;- list(Cell_Type=cell.type.col)

pheatmap(log(marker.counts+1), show_colnames =FALSE, 
         clustering_method = &quot;ward.D2&quot;, cutree_cols = 3, 
         annotation_col = anno.col, annotation_colors = anno.colors)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/find_markers_SV1_SV3-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>In the case of islet cells, marker genes are well established and IA-SVA did an excellent job of redefining these markers along with some other highly informative genes. Therefore, IA-SVA can be effectively used to uncover heterogeneity associated with cell types and can reveal genes that are expressed in a cell-specific manner.</p>
</div>
<div id="find-marker-genes-for-sv4." class="section level2">
<h2>Find marker genes for SV4.</h2>
<p>Here, using the find_markers() function we find marker genes significantly associated with SV4 (multiple testing adjusted p-value &lt; 0.05, default significance cutoff, a high R-squared value: R-squared &gt; 0.3).</p>
<pre class="r"><code>marker.counts.SV4 &lt;- find_markers(summ_exp, as.matrix(iasva.sv[,c(4)]),
                                  rsq.cutoff = 0.3)</code></pre>
<pre><code># of markers (): 94</code></pre>
<pre><code>total # of unique markers: 94</code></pre>
<pre class="r"><code>nrow(marker.counts.SV4)</code></pre>
<pre><code>[1] 94</code></pre>
<pre class="r"><code>anno.col &lt;- data.frame(SV4=iasva.sv[,4])
rownames(anno.col) &lt;- colnames(marker.counts)
head(anno.col)</code></pre>
<pre><code>                     SV4
4th-C63_S30  0.022045049
4th-C66_S36  0.006979656
4th-C7_S15   0.014475352
4th-C18_S31  0.012836294
4th-C5_S8    0.009518166
4th-C57_S18 -0.126329593</code></pre>
<pre class="r"><code>pheatmap(log(marker.counts.SV4+1), show_colnames =FALSE, 
         clustering_method = &quot;ward.D2&quot;, cutree_cols = 2,
         annotation_col = anno.col)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/find_markers-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="run-tsne-post-ia-sva-i.e.-run-tsne-on-marker-genes-for-sv1-and-sv2-obtained-from-ia-sva." class="section level2">
<h2>Run tSNE post IA-SVA, i.e., run tSNE on marker genes for SV1 and SV2 obtained from IA-SVA.</h2>
<p>Here, we apply tSNE on the marker genes for SV1 and SV2</p>
<pre class="r"><code>set.seed(344588)
tsne.res.iasva &lt;- Rtsne(unique(t(log(marker.counts+1))), dims = 2)
plot(tsne.res.iasva$Y, main=&quot;IA-SVA + tSNE&quot;, xlab=&quot;Dimension 1&quot;,
     ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Cell_Type],
     bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend(&quot;topright&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/run_tsne_post_iasva-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>tSNE conducted on genes selected via IA-SVA very clearly seperates cells into their corresponding cell types. Moreover, this analyses also revealed one cell (green cell clustered together with blue cells) that is potentially mislabeled in the original analyses.</p>
</div>
<div id="run-cellview-algorithm-to-visualize-the-data." class="section level2">
<h2>Run CellView algorithm to visualize the data.</h2>
<pre class="r"><code># specify gene number to select for
gene_num &lt;- 1000
# calcuclate dispersion
row.var &lt;- apply(lcounts,1,sd)**2
row.mean &lt;- apply(lcounts,1,mean)
dispersion &lt;- row.var/row.mean

# generate sequence of bins
bins &lt;- seq(from = min(row.mean), to = max(row.mean), length.out = 20)

# sort mean expression data into the bins
bin.exp &lt;- row.mean
# sort the values
bin.sort &lt;- sort(bin.exp, decreasing = FALSE)
# vector of bin assignment
cuts &lt;- cut(x = bin.exp, breaks = bins, labels = FALSE)
# find which are NA and change to zero
na.ids &lt;- which(is.na(cuts) == TRUE)
cuts[na.ids] &lt;- 0

# create an empty vector for overdispersion
overdispersion &lt;- NULL

# for each gene and bin index, calculate median, mad, and then normalized dispersion
# first loop through length of bins found
for (k in 1:length(names(table(cuts)))) {
  # find index of bins
  bin.id &lt;- which(cuts == names(table(cuts))[k])
  # median of all genes in the bin
  median.bin &lt;- median(dispersion[bin.id], na.rm = TRUE)
  # calculate mad (median absolute deviation)
  mad.bin &lt;- mad(dispersion[bin.id])
  # calculate norm dispersion for each gene
  for (m in 1:length(bin.id)) {
    norm.dispersion &lt;- abs(dispersion[bin.id[m]] - median.bin)/mad.bin
    overdispersion &lt;- c(overdispersion, norm.dispersion) 
  }
}

# remove nans 
overdis.na &lt;- which(is.na(overdispersion) == TRUE)
if (length(overdis.na) &gt; 0) {
  overdis.filt &lt;- overdispersion[-overdis.na]
} else {
  overdis.filt &lt;- overdispersion
}

# plot mean expression vs overdisperssion
ids &lt;- which(names(overdis.filt) %in% names(row.mean))
plot(row.mean[ids], overdis.filt)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/run_CellView-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Do t-sne using top over-dispersed genes (apply mean expression filter too)
rank.ov &lt;- order(overdis.filt, decreasing = TRUE)
ov.genes &lt;- names(overdis.filt[rank.ov[1:gene_num]])
log.sel &lt;- lcounts[ov.genes,]

all1 &lt;- t(log.sel)
# Remove groups that are all zeros
df &lt;- all1[,apply(all1, 2, var, na.rm=TRUE) != 0]

set.seed(34544532)
rtsne_out &lt;- Rtsne(as.matrix(df), dims = 3)

# Set rownames of matrix to tsne matrix
rownames(rtsne_out$Y) &lt;- rownames(df)

tsne.cellview &lt;- rtsne_out$Y

plot(tsne.cellview[,c(1,2)], main=&quot;CellView&quot;, xlab=&quot;Dimension 1&quot;, 
     ylab=&quot;Dimension 2&quot;,pch=21, col=color.vec[Cell_Type], 
     bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend(&quot;topright&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/run_CellView-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="run-seurat-to-reduce-dimensionality-and-visualize-islet-cells" class="section level2">
<h2>Run Seurat to reduce dimensionality and visualize islet cells</h2>
<pre class="r"><code>set.seed(12344)
seurat.obj &lt;- CreateSeuratObject(raw.data=raw.counts, 
                                 min.cells=3, min.genes=200, project=&quot;Seurat_Comp&quot;)

names(Patient_ID) &lt;- rownames(seurat.obj@meta.data)
seurat.obj &lt;- AddMetaData(object = seurat.obj, 
                          metadata = Patient_ID, col.name = &quot;patient.id&quot;)
names(Batch) &lt;- rownames(seurat.obj@meta.data)
seurat.obj &lt;- AddMetaData(object = seurat.obj, 
                          metadata = Batch, col.name = &quot;batch&quot;)
names(Geo_Lib) &lt;- rownames(seurat.obj@meta.data)
seurat.obj &lt;- AddMetaData(object = seurat.obj, 
                          metadata = Geo_Lib, col.name = &quot;geo.lib&quot;)

# Normalizing the data
seurat.obj &lt;- NormalizeData(object = seurat.obj,
                            normalization.method = &quot;LogNormalize&quot;, 
                            scale.factor = median(colSums(raw.counts)))

# Detection of variable genes across the single cells
seurat.obj &lt;- FindVariableGenes(object = seurat.obj, 
                                mean.function = ExpMean, dispersion.function = LogVMR, 
    x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/run_Seurat-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>length(x = seurat.obj@var.genes)</code></pre>
<pre><code>[1] 2933</code></pre>
<pre class="r"><code># Scaling the data and removing unwanted sources of variation
seurat.obj &lt;- ScaleData(object = seurat.obj, 
                        vars.to.regress = c(&quot;patient.id&quot;, &quot;batch&quot;, &quot;geo.lib&quot;))</code></pre>
<pre><code>Regressing out: patient.id, batch, geo.lib</code></pre>
<pre><code>
Time Elapsed:  9.59692406654358 secs</code></pre>
<pre><code>Scaling data matrix</code></pre>
<pre class="r"><code># Perform linear dimensional reduction
seurat.obj &lt;- RunPCA(object = seurat.obj, 
                     pc.genes = seurat.obj@var.genes, 
                     do.print = TRUE, pcs.print = 1:5, 
                     genes.print = 5)</code></pre>
<pre><code>[1] &quot;PC1&quot;
[1] &quot;CCDC108&quot;       &quot;SGIP1&quot;         &quot;MBD5&quot;          &quot;RP11-222A11.1&quot;
[5] &quot;GLRA3&quot;        
[1] &quot;&quot;
[1] &quot;DEFB1&quot; &quot;VTCN1&quot; &quot;PKHD1&quot; &quot;ITGB6&quot; &quot;PROM1&quot;
[1] &quot;&quot;
[1] &quot;&quot;
[1] &quot;PC2&quot;
[1] &quot;ELTD1&quot;  &quot;GPR116&quot; &quot;ERG&quot;    &quot;VWF&quot;    &quot;DYSF&quot;  
[1] &quot;&quot;
[1] &quot;VTCN1&quot;  &quot;DEFB1&quot;  &quot;PKHD1&quot;  &quot;ITGB6&quot;  &quot;LGALS4&quot;
[1] &quot;&quot;
[1] &quot;&quot;
[1] &quot;PC3&quot;
[1] &quot;PAQR4&quot;   &quot;SLITRK6&quot; &quot;BUB1B&quot;   &quot;MTFR2&quot;   &quot;TFAP2A&quot; 
[1] &quot;&quot;
[1] &quot;GCKR&quot;         &quot;LYPD6B&quot;       &quot;LGALS4&quot;       &quot;C6orf222&quot;    
[5] &quot;RP4-534N18.2&quot;
[1] &quot;&quot;
[1] &quot;&quot;
[1] &quot;PC4&quot;
[1] &quot;BUB1&quot;   &quot;ASPM&quot;   &quot;KIFC1&quot;  &quot;RAD51&quot;  &quot;DEPDC1&quot;
[1] &quot;&quot;
[1] &quot;SLIT2&quot;  &quot;COL5A1&quot; &quot;TFAP2A&quot; &quot;RUNX2&quot;  &quot;SFRP1&quot; 
[1] &quot;&quot;
[1] &quot;&quot;
[1] &quot;PC5&quot;
[1] &quot;FHL3&quot;   &quot;FGG&quot;    &quot;MAMLD1&quot; &quot;PRKCG&quot;  &quot;KRT222&quot;
[1] &quot;&quot;
[1] &quot;GLRA1&quot;   &quot;EPB41L2&quot; &quot;ADRA2A&quot;  &quot;TET1&quot;    &quot;DACH2&quot;  
[1] &quot;&quot;
[1] &quot;&quot;</code></pre>
<pre class="r"><code># Run tSNE (Spectral tSNE)
set.seed(8883)
seurat.obj &lt;- RunTSNE(object = seurat.obj, dims.use = 1:5, do.fast = TRUE)

# tSNE plot with color-coding of true cell types
plot(seurat.obj@dr$tsne@cell.embeddings[,c(1,2)], 
     main=&quot;Spectral tSNE (Seurat)&quot;, xlab=&quot;Dimension 1&quot;,
     ylab=&quot;Dimension 2&quot;,pch=21, col=color.vec[Cell_Type], 
     bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend(&quot;topleft&quot;, levels(Cell_Type), border=&quot;white&quot;, fill=color.vec, bty=&quot;n&quot;)</code></pre>
<p><img src="figure/tSNE_post_IA-SVA_3celltypes.Rmd/run_Seurat-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pdf(file=&quot;output/Lawlor_Islets_3Cells_tSNE_IA-SVA_Fig4AB.pdf&quot;, width=9, height=5)
layout(matrix(c(1,2), nrow=1, ncol=2, byrow=TRUE))
plot(tsne.res$Y, main=&quot;tSNE&quot;, xlab=&quot;Dimension 1&quot;, 
     ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Cell_Type],
     bg=color.vec[Cell_Type])
legend(&quot;topleft&quot;, levels(Cell_Type), border=&quot;white&quot;,
       fill=color.vec, bty=&quot;n&quot;)
plot(tsne.res.iasva$Y, main=&quot;IA-SVA + tSNE&quot;, xlab=&quot;Dimension 1&quot;, 
     ylab=&quot;Dimension 2&quot;, pch=21, col=color.vec[Cell_Type], 
     bg=color.vec[Cell_Type])
legend(&quot;topright&quot;, levels(Cell_Type), border=&quot;white&quot;, 
       fill=color.vec, bty=&quot;n&quot;)
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>anno.col &lt;- data.frame(Cell_Type=Cell_Type)
rownames(anno.col) &lt;- colnames(marker.counts)
head(anno.col)</code></pre>
<pre><code>            Cell_Type
4th-C63_S30     alpha
4th-C66_S36     alpha
4th-C7_S15       beta
4th-C18_S31     alpha
4th-C5_S8        beta
4th-C57_S18     alpha</code></pre>
<pre class="r"><code>cell.type.col &lt;- color.vec[1:3]
names(cell.type.col) &lt;- c(&quot;alpha&quot;,&quot;beta&quot;,&quot;ductal&quot;)
anno.colors &lt;- list(Cell_Type=cell.type.col)

pheatmap(log(marker.counts+1), show_colnames =FALSE, 
         clustering_method = &quot;ward.D2&quot;, cutree_cols = 3, 
         annotation_col = anno.col, annotation_colors = anno.colors,
         filename=&quot;output/Lawlor_Islets_3Cells_IASVA_SV1SV3_rsqcutoff0.3_pheatmap_iasvaV0.95_Figure4_C.pdf&quot;,
         width=6, height=17)</code></pre>
<pre class="r"><code>pdf(file=&quot;output/Lawlor_Islets_3Cells_IASVA_pairs4SVs_iasvaV0.95_black_FigS6.pdf&quot;,
    width=4, height=4)
pairs(iasva.sv[,1:4], pch=21, col=&quot;black&quot;, bg=&quot;black&quot;)
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>pdf(file=&quot;output/Lawlor_Islets_3Cells_IASVA_pairs4SVs_iasvaV0.95_color_FigS6.pdf&quot;, 
    width=4, height=4)
pairs(iasva.sv[,1:4], pch=21, col=color.vec[Cell_Type], 
      bg=color.vec[Cell_Type])
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>## 1,2
pdf(file=&quot;output/Lawlor_Islets_3Cells_CellView_Seurat_FigS.pdf&quot;, width=9, height=5)
layout(matrix(c(1,2), nrow=1, ncol=2, byrow=TRUE))
plot(tsne.cellview[,c(1,2)], main=&quot;CellView&quot;, xlab=&quot;Dimension 1&quot;,
     ylab=&quot;Dimension 2&quot;,pch=21, col=color.vec[Cell_Type],
     bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend(&quot;topright&quot;, levels(Cell_Type), border=&quot;white&quot;,
       fill=color.vec, bty=&quot;n&quot;)
plot(seurat.obj@dr$tsne@cell.embeddings[,c(1,2)], 
     main=&quot;Spectral tSNE (Seurat)&quot;, xlab=&quot;Dimension 1&quot;, 
     ylab=&quot;Dimension 2&quot;,pch=21, col=color.vec[Cell_Type], 
     bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend(&quot;topleft&quot;, levels(Cell_Type), border=&quot;white&quot;, 
       fill=color.vec, bty=&quot;n&quot;)
dev.off()</code></pre>
<pre><code>quartz_off_screen 
                2 </code></pre>
<pre class="r"><code>write.table(as.data.frame(rownames(marker.counts)), 
            file=&quot;output/Lawlor_Islets_3Cells_SV1_SV3_Cell_Type_Genes_rsqcutoff0.3.txt&quot;,
            quote=F, row.names=F, col.names=F, sep=&quot; &quot;)</code></pre>
<pre class="r"><code>write.table(as.data.frame(rownames(marker.counts.SV4)), 
            file=&quot;output/Lawlor_Islets_3Cells_SV4_Genes_rsqcutoff0.3.txt&quot;, quote=F,
            row.names=F, col.names=F, sep=&quot; &quot;)

anno.col &lt;- data.frame(SV4=iasva.sv[,4])
rownames(anno.col) &lt;- colnames(marker.counts)
head(anno.col)</code></pre>
<pre><code>                     SV4
4th-C63_S30  0.022045049
4th-C66_S36  0.006979656
4th-C7_S15   0.014475352
4th-C18_S31  0.012836294
4th-C5_S8    0.009518166
4th-C57_S18 -0.126329593</code></pre>
<pre class="r"><code>pheatmap(log(marker.counts.SV4+1), show_colnames =FALSE, 
         clustering_method = &quot;ward.D2&quot;, cutree_cols = 2, 
         annotation_col = anno.col,
         filename=&quot;output/Lawlor_Islets_3Cells_IASVA_SV4_rsqcutoff0.3_pheatmap_iasvaV0.95.pdf&quot;,
         width=8, height=14)</code></pre>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.0 (2018-04-23)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] SummarizedExperiment_1.10.1 DelayedArray_0.6.1         
 [3] BiocParallel_1.14.2         matrixStats_0.53.1         
 [5] Biobase_2.40.0              GenomicRanges_1.32.3       
 [7] GenomeInfoDb_1.16.0         IRanges_2.14.10            
 [9] S4Vectors_0.18.3            BiocGenerics_0.26.0        
[11] RColorBrewer_1.1-2          DescTools_0.99.24          
[13] corrplot_0.84               pheatmap_1.0.10            
[15] Rtsne_0.13                  dbscan_1.1-2               
[17] Seurat_2.3.2                cowplot_0.9.2              
[19] ggplot2_2.2.1.9000          iasvaExamples_1.0.0        
[21] iasva_0.99.3                irlba_2.3.2                
[23] Matrix_1.2-14              

loaded via a namespace (and not attached):
  [1] snow_0.4-2             backports_1.1.2        Hmisc_4.1-1           
  [4] workflowr_1.0.1        VGAM_1.0-5             plyr_1.8.4            
  [7] igraph_1.2.1           lazyeval_0.2.1         splines_3.5.0         
 [10] digest_0.6.15          foreach_1.4.4          htmltools_0.3.6       
 [13] lars_1.2               gdata_2.18.0           magrittr_1.5          
 [16] checkmate_1.8.5        cluster_2.0.7-1        mixtools_1.1.0        
 [19] ROCR_1.0-7             sfsmisc_1.1-2          recipes_0.1.3         
 [22] gower_0.1.2            dimRed_0.1.0           R.utils_2.6.0         
 [25] colorspace_1.3-2       dplyr_0.7.5            jsonlite_1.5          
 [28] RCurl_1.95-4.10        bindr_0.1.1            zoo_1.8-2             
 [31] survival_2.42-3        iterators_1.0.9        ape_5.1               
 [34] glue_1.2.0             DRR_0.0.3              gtable_0.2.0          
 [37] ipred_0.9-6            zlibbioc_1.26.0        XVector_0.20.0        
 [40] kernlab_0.9-26         ddalpha_1.3.3          prabclus_2.2-6        
 [43] DEoptimR_1.0-8         abind_1.4-5            scales_0.5.0          
 [46] mvtnorm_1.0-8          Rcpp_0.12.17           dtw_1.20-1            
 [49] metap_0.9              htmlTable_1.12         tclust_1.4-1          
 [52] magic_1.5-8            reticulate_1.8         bit_1.1-14            
 [55] proxy_0.4-22           foreign_0.8-70         mclust_5.4            
 [58] SDMTools_1.1-221       Formula_1.2-3          tsne_0.1-3            
 [61] lava_1.6.1             prodlim_2018.04.18     htmlwidgets_1.2       
 [64] FNN_1.1                gplots_3.0.1           fpc_2.1-11            
 [67] acepack_1.4.1          modeltools_0.2-21      ica_1.0-2             
 [70] manipulate_1.0.1       pkgconfig_2.0.1        R.methodsS3_1.7.1     
 [73] flexmix_2.3-14         nnet_7.3-12            caret_6.0-80          
 [76] tidyselect_0.2.4       rlang_0.2.1            reshape2_1.4.3        
 [79] munsell_0.4.3          tools_3.5.0            ranger_0.10.1         
 [82] pls_2.6-0              ggridges_0.5.0         broom_0.4.4           
 [85] evaluate_0.10.1        geometry_0.3-6         stringr_1.3.1         
 [88] yaml_2.1.19            bit64_0.9-7            ModelMetrics_1.1.0    
 [91] knitr_1.20             fitdistrplus_1.0-9     robustbase_0.93-0     
 [94] caTools_1.17.1         purrr_0.2.5            RANN_2.5.1            
 [97] bindrcpp_0.2.2         pbapply_1.3-4          nlme_3.1-137          
[100] whisker_0.3-2          R.oo_1.22.0            RcppRoll_0.3.0        
[103] hdf5r_1.0.0            compiler_3.5.0         rstudioapi_0.7        
[106] png_0.1-7              tibble_1.4.2           stringi_1.2.2         
[109] lattice_0.20-35        trimcluster_0.1-2      psych_1.8.4           
[112] diffusionMap_1.1-0     pillar_1.2.3           lmtest_0.9-36         
[115] data.table_1.11.4      bitops_1.0-6           R6_2.2.2              
[118] latticeExtra_0.6-28    KernSmooth_2.23-15     gridExtra_2.3         
[121] codetools_0.2-15       boot_1.3-20            MASS_7.3-50           
[124] gtools_3.5.0           assertthat_0.2.0       CVST_0.2-2            
[127] rprojroot_1.3-2        withr_2.1.2            mnormt_1.5-5          
[130] GenomeInfoDbData_1.1.0 expm_0.999-2           diptest_0.75-7        
[133] doSNOW_1.0.16          grid_3.5.0             rpart_4.1-13          
[136] timeDate_3043.102      tidyr_0.8.1            class_7.3-14          
[139] rmarkdown_1.9          segmented_0.5-3.0      git2r_0.21.0          
[142] lubridate_1.7.4        scatterplot3d_0.3-41   base64enc_0.1-3       </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1
</p>
<hr>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
